// @pulumi/gcp@8.41.1/storage/controlFolderIntelligenceConfig.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ControlFolderIntelligenceConfig = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * ## Example Usage
 *
 * ### Storage Control Folder Intelligence Config Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const example = new gcp.storage.ControlFolderIntelligenceConfig("example", {
 *     name: "12345678",
 *     editionConfig: "STANDARD",
 *     filter: {
 *         includedCloudStorageBuckets: {
 *             bucketIdRegexes: [
 *                 "test-id-1*",
 *                 "test-id-2*",
 *             ],
 *         },
 *         excludedCloudStorageLocations: {
 *             locations: [
 *                 "test-location-1*",
 *                 "test-locations-2*",
 *             ],
 *         },
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * FolderIntelligenceConfig can be imported using any of these accepted formats:
 *
 * * `folders/{{name}}/locations/global/intelligenceConfig`
 *
 * * `{{name}}`
 *
 * When using the `pulumi import` command, FolderIntelligenceConfig can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:storage/controlFolderIntelligenceConfig:ControlFolderIntelligenceConfig default folders/{{name}}/locations/global/intelligenceConfig
 * ```
 *
 * ```sh
 * $ pulumi import gcp:storage/controlFolderIntelligenceConfig:ControlFolderIntelligenceConfig default {{name}}
 * ```
 */
class ControlFolderIntelligenceConfig extends pulumi.CustomResource {
    /**
     * Get an existing ControlFolderIntelligenceConfig resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ControlFolderIntelligenceConfig(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ControlFolderIntelligenceConfig.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ControlFolderIntelligenceConfig.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["editionConfig"] = state ? state.editionConfig : undefined;
            resourceInputs["effectiveIntelligenceConfigs"] = state ? state.effectiveIntelligenceConfigs : undefined;
            resourceInputs["filter"] = state ? state.filter : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["trialConfigs"] = state ? state.trialConfigs : undefined;
            resourceInputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            resourceInputs["editionConfig"] = args ? args.editionConfig : undefined;
            resourceInputs["filter"] = args ? args.filter : undefined;
            resourceInputs["name"] = args ? args.name : undefined;
            resourceInputs["effectiveIntelligenceConfigs"] = undefined /*out*/;
            resourceInputs["trialConfigs"] = undefined /*out*/;
            resourceInputs["updateTime"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(ControlFolderIntelligenceConfig.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ControlFolderIntelligenceConfig = ControlFolderIntelligenceConfig;
/** @internal */
ControlFolderIntelligenceConfig.__pulumiType = 'gcp:storage/controlFolderIntelligenceConfig:ControlFolderIntelligenceConfig';
