// @pulumi/gcp@8.41.1/dataplex/glossaryTerm.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlossaryTerm = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Represents a collection of terms within a Glossary that are related to each other.
 *
 * ## Example Usage
 *
 * ### Dataplex Glossary Term Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const termTestId = new gcp.dataplex.Glossary("term_test_id", {
 *     glossaryId: "tf-test-glossary_88722",
 *     location: "us-central1",
 * });
 * const termTestIdGlossaryTerm = new gcp.dataplex.GlossaryTerm("term_test_id", {
 *     parent: pulumi.interpolate`projects/${termTestId.project}/locations/us-central1/glossaries/${termTestId.glossaryId}`,
 *     glossaryId: termTestId.glossaryId,
 *     location: "us-central1",
 *     termId: "tf-test-term-basic_39249",
 * });
 * ```
 * ### Dataplex Glossary Term Full
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const termTestIdFull = new gcp.dataplex.Glossary("term_test_id_full", {
 *     glossaryId: "tf-test-glossary_74391",
 *     location: "us-central1",
 * });
 * const termTestIdFullGlossaryTerm = new gcp.dataplex.GlossaryTerm("term_test_id_full", {
 *     parent: pulumi.interpolate`projects/${termTestIdFull.project}/locations/us-central1/glossaries/${termTestIdFull.glossaryId}`,
 *     glossaryId: termTestIdFull.glossaryId,
 *     location: "us-central1",
 *     termId: "tf-test-term-full_16511",
 *     labels: {
 *         tag: "test-tf",
 *     },
 *     displayName: "terraform term",
 *     description: "term created by Terraform",
 * });
 * ```
 *
 * ## Import
 *
 * GlossaryTerm can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/locations/{{location}}/glossaries/{{glossary_id}}/terms/{{term_id}}`
 *
 * * `{{project}}/{{location}}/{{glossary_id}}/{{term_id}}`
 *
 * * `{{location}}/{{glossary_id}}/{{term_id}}`
 *
 * When using the `pulumi import` command, GlossaryTerm can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:dataplex/glossaryTerm:GlossaryTerm default projects/{{project}}/locations/{{location}}/glossaries/{{glossary_id}}/terms/{{term_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:dataplex/glossaryTerm:GlossaryTerm default {{project}}/{{location}}/{{glossary_id}}/{{term_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:dataplex/glossaryTerm:GlossaryTerm default {{location}}/{{glossary_id}}/{{term_id}}
 * ```
 */
class GlossaryTerm extends pulumi.CustomResource {
    /**
     * Get an existing GlossaryTerm resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new GlossaryTerm(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of GlossaryTerm.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === GlossaryTerm.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["createTime"] = state ? state.createTime : undefined;
            resourceInputs["description"] = state ? state.description : undefined;
            resourceInputs["displayName"] = state ? state.displayName : undefined;
            resourceInputs["effectiveLabels"] = state ? state.effectiveLabels : undefined;
            resourceInputs["glossaryId"] = state ? state.glossaryId : undefined;
            resourceInputs["labels"] = state ? state.labels : undefined;
            resourceInputs["location"] = state ? state.location : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["parent"] = state ? state.parent : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["pulumiLabels"] = state ? state.pulumiLabels : undefined;
            resourceInputs["termId"] = state ? state.termId : undefined;
            resourceInputs["uid"] = state ? state.uid : undefined;
            resourceInputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.location === undefined) && !opts.urn) {
                throw new Error("Missing required property 'location'");
            }
            if ((!args || args.parent === undefined) && !opts.urn) {
                throw new Error("Missing required property 'parent'");
            }
            resourceInputs["description"] = args ? args.description : undefined;
            resourceInputs["displayName"] = args ? args.displayName : undefined;
            resourceInputs["glossaryId"] = args ? args.glossaryId : undefined;
            resourceInputs["labels"] = args ? args.labels : undefined;
            resourceInputs["location"] = args ? args.location : undefined;
            resourceInputs["parent"] = args ? args.parent : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["termId"] = args ? args.termId : undefined;
            resourceInputs["createTime"] = undefined /*out*/;
            resourceInputs["effectiveLabels"] = undefined /*out*/;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["pulumiLabels"] = undefined /*out*/;
            resourceInputs["uid"] = undefined /*out*/;
            resourceInputs["updateTime"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const secretOpts = { additionalSecretOutputs: ["effectiveLabels", "pulumiLabels"] };
        opts = pulumi.mergeOptions(opts, secretOpts);
        super(GlossaryTerm.__pulumiType, name, resourceInputs, opts);
    }
}
exports.GlossaryTerm = GlossaryTerm;
/** @internal */
GlossaryTerm.__pulumiType = 'gcp:dataplex/glossaryTerm:GlossaryTerm';
