// @pulumi/gcp@8.41.1/artifactregistry/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcscConfig = exports.RepositoryIamPolicy = exports.RepositoryIamMember = exports.RepositoryIamBinding = exports.Repository = exports.getVersionOutput = exports.getVersion = exports.getTagsOutput = exports.getTags = exports.getTagOutput = exports.getTag = exports.getRepositoryIamPolicyOutput = exports.getRepositoryIamPolicy = exports.getRepositoryOutput = exports.getRepository = exports.getRepositoriesOutput = exports.getRepositories = exports.getPackageOutput = exports.getPackage = exports.getLocationsOutput = exports.getLocations = exports.getDockerImagesOutput = exports.getDockerImages = exports.getDockerImageOutput = exports.getDockerImage = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.getDockerImage = null;
exports.getDockerImageOutput = null;
utilities.lazyLoad(exports, ["getDockerImage", "getDockerImageOutput"], () => require("./getDockerImage"));
exports.getDockerImages = null;
exports.getDockerImagesOutput = null;
utilities.lazyLoad(exports, ["getDockerImages", "getDockerImagesOutput"], () => require("./getDockerImages"));
exports.getLocations = null;
exports.getLocationsOutput = null;
utilities.lazyLoad(exports, ["getLocations", "getLocationsOutput"], () => require("./getLocations"));
exports.getPackage = null;
exports.getPackageOutput = null;
utilities.lazyLoad(exports, ["getPackage", "getPackageOutput"], () => require("./getPackage"));
exports.getRepositories = null;
exports.getRepositoriesOutput = null;
utilities.lazyLoad(exports, ["getRepositories", "getRepositoriesOutput"], () => require("./getRepositories"));
exports.getRepository = null;
exports.getRepositoryOutput = null;
utilities.lazyLoad(exports, ["getRepository", "getRepositoryOutput"], () => require("./getRepository"));
exports.getRepositoryIamPolicy = null;
exports.getRepositoryIamPolicyOutput = null;
utilities.lazyLoad(exports, ["getRepositoryIamPolicy", "getRepositoryIamPolicyOutput"], () => require("./getRepositoryIamPolicy"));
exports.getTag = null;
exports.getTagOutput = null;
utilities.lazyLoad(exports, ["getTag", "getTagOutput"], () => require("./getTag"));
exports.getTags = null;
exports.getTagsOutput = null;
utilities.lazyLoad(exports, ["getTags", "getTagsOutput"], () => require("./getTags"));
exports.getVersion = null;
exports.getVersionOutput = null;
utilities.lazyLoad(exports, ["getVersion", "getVersionOutput"], () => require("./getVersion"));
exports.Repository = null;
utilities.lazyLoad(exports, ["Repository"], () => require("./repository"));
exports.RepositoryIamBinding = null;
utilities.lazyLoad(exports, ["RepositoryIamBinding"], () => require("./repositoryIamBinding"));
exports.RepositoryIamMember = null;
utilities.lazyLoad(exports, ["RepositoryIamMember"], () => require("./repositoryIamMember"));
exports.RepositoryIamPolicy = null;
utilities.lazyLoad(exports, ["RepositoryIamPolicy"], () => require("./repositoryIamPolicy"));
exports.VpcscConfig = null;
utilities.lazyLoad(exports, ["VpcscConfig"], () => require("./vpcscConfig"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "gcp:artifactregistry/repository:Repository":
                return new exports.Repository(name, undefined, { urn });
            case "gcp:artifactregistry/repositoryIamBinding:RepositoryIamBinding":
                return new exports.RepositoryIamBinding(name, undefined, { urn });
            case "gcp:artifactregistry/repositoryIamMember:RepositoryIamMember":
                return new exports.RepositoryIamMember(name, undefined, { urn });
            case "gcp:artifactregistry/repositoryIamPolicy:RepositoryIamPolicy":
                return new exports.RepositoryIamPolicy(name, undefined, { urn });
            case "gcp:artifactregistry/vpcscConfig:VpcscConfig":
                return new exports.VpcscConfig(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("gcp", "artifactregistry/repository", _module);
pulumi.runtime.registerResourceModule("gcp", "artifactregistry/repositoryIamBinding", _module);
pulumi.runtime.registerResourceModule("gcp", "artifactregistry/repositoryIamMember", _module);
pulumi.runtime.registerResourceModule("gcp", "artifactregistry/repositoryIamPolicy", _module);
pulumi.runtime.registerResourceModule("gcp", "artifactregistry/vpcscConfig", _module);
