// @pulumi/gcp@8.41.1/chronicle/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Watchlist = exports.RuleDeployment = exports.Rule = exports.Retrohunt = exports.ReferenceList = exports.DataAccessScope = exports.DataAccessLabel = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.DataAccessLabel = null;
utilities.lazyLoad(exports, ["DataAccessLabel"], () => require("./dataAccessLabel"));
exports.DataAccessScope = null;
utilities.lazyLoad(exports, ["DataAccessScope"], () => require("./dataAccessScope"));
exports.ReferenceList = null;
utilities.lazyLoad(exports, ["ReferenceList"], () => require("./referenceList"));
exports.Retrohunt = null;
utilities.lazyLoad(exports, ["Retrohunt"], () => require("./retrohunt"));
exports.Rule = null;
utilities.lazyLoad(exports, ["Rule"], () => require("./rule"));
exports.RuleDeployment = null;
utilities.lazyLoad(exports, ["RuleDeployment"], () => require("./ruleDeployment"));
exports.Watchlist = null;
utilities.lazyLoad(exports, ["Watchlist"], () => require("./watchlist"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "gcp:chronicle/dataAccessLabel:DataAccessLabel":
                return new exports.DataAccessLabel(name, undefined, { urn });
            case "gcp:chronicle/dataAccessScope:DataAccessScope":
                return new exports.DataAccessScope(name, undefined, { urn });
            case "gcp:chronicle/referenceList:ReferenceList":
                return new exports.ReferenceList(name, undefined, { urn });
            case "gcp:chronicle/retrohunt:Retrohunt":
                return new exports.Retrohunt(name, undefined, { urn });
            case "gcp:chronicle/rule:Rule":
                return new exports.Rule(name, undefined, { urn });
            case "gcp:chronicle/ruleDeployment:RuleDeployment":
                return new exports.RuleDeployment(name, undefined, { urn });
            case "gcp:chronicle/watchlist:Watchlist":
                return new exports.Watchlist(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("gcp", "chronicle/dataAccessLabel", _module);
pulumi.runtime.registerResourceModule("gcp", "chronicle/dataAccessScope", _module);
pulumi.runtime.registerResourceModule("gcp", "chronicle/referenceList", _module);
pulumi.runtime.registerResourceModule("gcp", "chronicle/retrohunt", _module);
pulumi.runtime.registerResourceModule("gcp", "chronicle/rule", _module);
pulumi.runtime.registerResourceModule("gcp", "chronicle/ruleDeployment", _module);
pulumi.runtime.registerResourceModule("gcp", "chronicle/watchlist", _module);
