// @pulumi/gcp@8.41.1/chronicle/ruleDeployment.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuleDeployment = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * The RuleDeployment resource represents the deployment state of a Rule.
 *
 * To get more information about RuleDeployment, see:
 *
 * * [API documentation](https://cloud.google.com/chronicle/docs/reference/rest/v1/RuleDeployment)
 * * How-to Guides
 *     * [Google SecOps Guides](https://cloud.google.com/chronicle/docs/secops/secops-overview)
 *
 * ## Example Usage
 *
 * ## Import
 *
 * RuleDeployment can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/locations/{{location}}/instances/{{instance}}/rules/{{rule}}/deployment`
 *
 * * `{{project}}/{{location}}/{{instance}}/{{rule}}`
 *
 * * `{{location}}/{{instance}}/{{rule}}`
 *
 * When using the `pulumi import` command, RuleDeployment can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:chronicle/ruleDeployment:RuleDeployment default projects/{{project}}/locations/{{location}}/instances/{{instance}}/rules/{{rule}}/deployment
 * ```
 *
 * ```sh
 * $ pulumi import gcp:chronicle/ruleDeployment:RuleDeployment default {{project}}/{{location}}/{{instance}}/{{rule}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:chronicle/ruleDeployment:RuleDeployment default {{location}}/{{instance}}/{{rule}}
 * ```
 */
class RuleDeployment extends pulumi.CustomResource {
    /**
     * Get an existing RuleDeployment resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new RuleDeployment(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of RuleDeployment.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === RuleDeployment.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["alerting"] = state ? state.alerting : undefined;
            resourceInputs["archiveTime"] = state ? state.archiveTime : undefined;
            resourceInputs["archived"] = state ? state.archived : undefined;
            resourceInputs["consumerRules"] = state ? state.consumerRules : undefined;
            resourceInputs["enabled"] = state ? state.enabled : undefined;
            resourceInputs["executionState"] = state ? state.executionState : undefined;
            resourceInputs["instance"] = state ? state.instance : undefined;
            resourceInputs["lastAlertStatusChangeTime"] = state ? state.lastAlertStatusChangeTime : undefined;
            resourceInputs["location"] = state ? state.location : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["producerRules"] = state ? state.producerRules : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["rule"] = state ? state.rule : undefined;
            resourceInputs["runFrequency"] = state ? state.runFrequency : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.instance === undefined) && !opts.urn) {
                throw new Error("Missing required property 'instance'");
            }
            if ((!args || args.location === undefined) && !opts.urn) {
                throw new Error("Missing required property 'location'");
            }
            if ((!args || args.rule === undefined) && !opts.urn) {
                throw new Error("Missing required property 'rule'");
            }
            resourceInputs["alerting"] = args ? args.alerting : undefined;
            resourceInputs["archived"] = args ? args.archived : undefined;
            resourceInputs["enabled"] = args ? args.enabled : undefined;
            resourceInputs["instance"] = args ? args.instance : undefined;
            resourceInputs["location"] = args ? args.location : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["rule"] = args ? args.rule : undefined;
            resourceInputs["runFrequency"] = args ? args.runFrequency : undefined;
            resourceInputs["archiveTime"] = undefined /*out*/;
            resourceInputs["consumerRules"] = undefined /*out*/;
            resourceInputs["executionState"] = undefined /*out*/;
            resourceInputs["lastAlertStatusChangeTime"] = undefined /*out*/;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["producerRules"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(RuleDeployment.__pulumiType, name, resourceInputs, opts);
    }
}
exports.RuleDeployment = RuleDeployment;
/** @internal */
RuleDeployment.__pulumiType = 'gcp:chronicle/ruleDeployment:RuleDeployment';
