// @pulumi/gcp@8.41.1/firebase/dataConnectService.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataConnectService = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A Firebase Data Connect service.
 *
 * To get more information about Service, see:
 *
 * * [API documentation](https://firebase.google.com/docs/reference/data-connect/rest)
 * * How-to Guides
 *     * [Official Documentation](https://firebase.google.com/docs/data-connect)
 *
 * ## Example Usage
 *
 * ### Firebasedataconnect Service Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * // Enable Firebase Data Connect API
 * const fdc = new gcp.projects.Service("fdc", {
 *     project: "my-project-name",
 *     service: "firebasedataconnect.googleapis.com",
 *     disableOnDestroy: false,
 * });
 * // Create a Firebase Data Connect service
 * const _default = new gcp.firebase.DataConnectService("default", {
 *     project: "my-project-name",
 *     location: "us-central1",
 *     serviceId: "example-service",
 *     deletionPolicy: "DEFAULT",
 *     labels: {
 *         label: "my-label",
 *     },
 *     annotations: {
 *         key1: "value1",
 *         key2: "value2",
 *     },
 * }, {
 *     dependsOn: [fdc],
 * });
 * ```
 * ### Firebasedataconnect Service With Force Deletion
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * // Enable Firebase Data Connect API
 * const fdc = new gcp.projects.Service("fdc", {
 *     project: "my-project-name",
 *     service: "firebasedataconnect.googleapis.com",
 *     disableOnDestroy: false,
 * });
 * // Create a Firebase Data Connect service
 * const _default = new gcp.firebase.DataConnectService("default", {
 *     project: "my-project-name",
 *     location: "us-central1",
 *     serviceId: "example-service",
 *     deletionPolicy: "FORCE",
 * }, {
 *     dependsOn: [fdc],
 * });
 * ```
 *
 * ## Import
 *
 * Service can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/locations/{{location}}/services/{{service_id}}`
 *
 * * `{{project}}/{{location}}/{{service_id}}`
 *
 * * `{{location}}/{{service_id}}`
 *
 * When using the `pulumi import` command, Service can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:firebase/dataConnectService:DataConnectService default projects/{{project}}/locations/{{location}}/services/{{service_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:firebase/dataConnectService:DataConnectService default {{project}}/{{location}}/{{service_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:firebase/dataConnectService:DataConnectService default {{location}}/{{service_id}}
 * ```
 */
class DataConnectService extends pulumi.CustomResource {
    /**
     * Get an existing DataConnectService resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new DataConnectService(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of DataConnectService.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === DataConnectService.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["annotations"] = state ? state.annotations : undefined;
            resourceInputs["createTime"] = state ? state.createTime : undefined;
            resourceInputs["deletionPolicy"] = state ? state.deletionPolicy : undefined;
            resourceInputs["displayName"] = state ? state.displayName : undefined;
            resourceInputs["effectiveAnnotations"] = state ? state.effectiveAnnotations : undefined;
            resourceInputs["effectiveLabels"] = state ? state.effectiveLabels : undefined;
            resourceInputs["etag"] = state ? state.etag : undefined;
            resourceInputs["labels"] = state ? state.labels : undefined;
            resourceInputs["location"] = state ? state.location : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["pulumiLabels"] = state ? state.pulumiLabels : undefined;
            resourceInputs["reconciling"] = state ? state.reconciling : undefined;
            resourceInputs["serviceId"] = state ? state.serviceId : undefined;
            resourceInputs["uid"] = state ? state.uid : undefined;
            resourceInputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.location === undefined) && !opts.urn) {
                throw new Error("Missing required property 'location'");
            }
            if ((!args || args.serviceId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'serviceId'");
            }
            resourceInputs["annotations"] = args ? args.annotations : undefined;
            resourceInputs["deletionPolicy"] = args ? args.deletionPolicy : undefined;
            resourceInputs["displayName"] = args ? args.displayName : undefined;
            resourceInputs["labels"] = args ? args.labels : undefined;
            resourceInputs["location"] = args ? args.location : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["serviceId"] = args ? args.serviceId : undefined;
            resourceInputs["createTime"] = undefined /*out*/;
            resourceInputs["effectiveAnnotations"] = undefined /*out*/;
            resourceInputs["effectiveLabels"] = undefined /*out*/;
            resourceInputs["etag"] = undefined /*out*/;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["pulumiLabels"] = undefined /*out*/;
            resourceInputs["reconciling"] = undefined /*out*/;
            resourceInputs["uid"] = undefined /*out*/;
            resourceInputs["updateTime"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const secretOpts = { additionalSecretOutputs: ["effectiveLabels", "pulumiLabels"] };
        opts = pulumi.mergeOptions(opts, secretOpts);
        super(DataConnectService.__pulumiType, name, resourceInputs, opts);
    }
}
exports.DataConnectService = DataConnectService;
/** @internal */
DataConnectService.__pulumiType = 'gcp:firebase/dataConnectService:DataConnectService';
