// @pulumi/gcp@8.41.1/siteverification/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebResource = exports.Owner = exports.getTokenOutput = exports.getToken = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.getToken = null;
exports.getTokenOutput = null;
utilities.lazyLoad(exports, ["getToken", "getTokenOutput"], () => require("./getToken"));
exports.Owner = null;
utilities.lazyLoad(exports, ["Owner"], () => require("./owner"));
exports.WebResource = null;
utilities.lazyLoad(exports, ["WebResource"], () => require("./webResource"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "gcp:siteverification/owner:Owner":
                return new exports.Owner(name, undefined, { urn });
            case "gcp:siteverification/webResource:WebResource":
                return new exports.WebResource(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("gcp", "siteverification/owner", _module);
pulumi.runtime.registerResourceModule("gcp", "siteverification/webResource", _module);
