// @pulumi/gcp@8.41.1/siteverification/webResource.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebResource = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A web resource is a website or domain with verified ownership. Once your
 * ownership is verified you will be able to manage your website in the
 * [Google Search Console](https://www.google.com/webmasters/tools/).
 *
 * > **Note:** The verification data (DNS `TXT` record, HTML file, `meta` tag, etc.)
 * must already exist before the web resource is created, and must be deleted before
 * the web resource is destroyed. The Google Site Verification API checks that the
 * verification data exists at creation time and does not exist at destruction time
 * and will fail if the required condition is not met.
 *
 * To get more information about WebResource, see:
 *
 * * [API documentation](https://developers.google.com/site-verification/v1)
 * * How-to Guides
 *     * [Getting Started](https://developers.google.com/site-verification/v1/getting_started)
 *
 * ## Example Usage
 *
 * ### Site Verification Domain Record
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const token = gcp.siteverification.getToken({
 *     type: "INET_DOMAIN",
 *     identifier: "www.example.com",
 *     verificationMethod: "DNS_TXT",
 * });
 * const example = new gcp.dns.RecordSet("example", {
 *     managedZone: "example.com",
 *     name: "www.example.com.",
 *     type: "TXT",
 *     rrdatas: [token.then(token => token.token)],
 *     ttl: 86400,
 * });
 * const exampleWebResource = new gcp.siteverification.WebResource("example", {
 *     site: {
 *         type: token.then(token => token.type),
 *         identifier: token.then(token => token.identifier),
 *     },
 *     verificationMethod: token.then(token => token.verificationMethod),
 * }, {
 *     dependsOn: [example],
 * });
 * ```
 *
 * ## Import
 *
 * WebResource can be imported using any of these accepted formats:
 *
 * * `webResource/{{web_resource_id}}`
 *
 * * `{{web_resource_id}}`
 *
 * When using the `pulumi import` command, WebResource can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:siteverification/webResource:WebResource default webResource/{{web_resource_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:siteverification/webResource:WebResource default {{web_resource_id}}
 * ```
 */
class WebResource extends pulumi.CustomResource {
    /**
     * Get an existing WebResource resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new WebResource(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of WebResource.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === WebResource.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["owners"] = state ? state.owners : undefined;
            resourceInputs["site"] = state ? state.site : undefined;
            resourceInputs["verificationMethod"] = state ? state.verificationMethod : undefined;
            resourceInputs["webResourceId"] = state ? state.webResourceId : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.site === undefined) && !opts.urn) {
                throw new Error("Missing required property 'site'");
            }
            if ((!args || args.verificationMethod === undefined) && !opts.urn) {
                throw new Error("Missing required property 'verificationMethod'");
            }
            resourceInputs["site"] = args ? args.site : undefined;
            resourceInputs["verificationMethod"] = args ? args.verificationMethod : undefined;
            resourceInputs["owners"] = undefined /*out*/;
            resourceInputs["webResourceId"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(WebResource.__pulumiType, name, resourceInputs, opts);
    }
}
exports.WebResource = WebResource;
/** @internal */
WebResource.__pulumiType = 'gcp:siteverification/webResource:WebResource';
