// @pulumi/gcp@8.41.1/vertex/aiRagEngineConfig.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.AiRagEngineConfig = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Vertex AI RAG Engine lets you scale your RagManagedDb instance based on your usage and performance requirements using a choice of two tiers, and optionally, lets you delete your Vertex AI RAG Engine data using a third tier. The tier is a project-level setting that's available in the RagEngineConfig resource that impacts all RAG corpora using RagManagedDb. The following tiers are available in RagEngineConfig: Basic, Scaled and Unprovisioned.
 *
 * To get more information about RagEngineConfig, see:
 *
 * * [API documentation](https://cloud.google.com/vertex-ai/generative-ai/docs/reference/rest/v1/RagEngineConfig)
 * * How-to Guides
 *     * [Official Documentation](https://cloud.google.com/vertex-ai/generative-ai/docs/rag-engine/understanding-ragmanageddb)
 *
 * ## Example Usage
 *
 * ## Import
 *
 * RagEngineConfig can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/locations/{{region}}/ragEngineConfig`
 *
 * * `{{project}}/{{region}}`
 *
 * * `{{region}}`
 *
 * When using the `pulumi import` command, RagEngineConfig can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:vertex/aiRagEngineConfig:AiRagEngineConfig default projects/{{project}}/locations/{{region}}/ragEngineConfig
 * ```
 *
 * ```sh
 * $ pulumi import gcp:vertex/aiRagEngineConfig:AiRagEngineConfig default {{project}}/{{region}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:vertex/aiRagEngineConfig:AiRagEngineConfig default {{region}}
 * ```
 */
class AiRagEngineConfig extends pulumi.CustomResource {
    /**
     * Get an existing AiRagEngineConfig resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new AiRagEngineConfig(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of AiRagEngineConfig.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === AiRagEngineConfig.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["ragManagedDbConfig"] = state ? state.ragManagedDbConfig : undefined;
            resourceInputs["region"] = state ? state.region : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.ragManagedDbConfig === undefined) && !opts.urn) {
                throw new Error("Missing required property 'ragManagedDbConfig'");
            }
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["ragManagedDbConfig"] = args ? args.ragManagedDbConfig : undefined;
            resourceInputs["region"] = args ? args.region : undefined;
            resourceInputs["name"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(AiRagEngineConfig.__pulumiType, name, resourceInputs, opts);
    }
}
exports.AiRagEngineConfig = AiRagEngineConfig;
/** @internal */
AiRagEngineConfig.__pulumiType = 'gcp:vertex/aiRagEngineConfig:AiRagEngineConfig';
