// @pulumi/gcp@8.41.1/folder/serviceIdentity.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceIdentity = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * ## Example Usage
 *
 * ### Folder Service Identity Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const myFolder = new gcp.organizations.Folder("my_folder", {
 *     parent: "organizations/1234567",
 *     displayName: "my-folder",
 * });
 * const osconfigSa = new gcp.folder.ServiceIdentity("osconfig_sa", {
 *     folder: myFolder.folderId,
 *     service: "osconfig.googleapis.com",
 * });
 * const admin = new gcp.folder.IAMMember("admin", {
 *     folder: myFolder.name,
 *     role: "roles/osconfig.serviceAgent",
 *     member: osconfigSa.member,
 * });
 * ```
 *
 * ## Import
 *
 * This resource does not support import.
 */
class ServiceIdentity extends pulumi.CustomResource {
    /**
     * Get an existing ServiceIdentity resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ServiceIdentity(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ServiceIdentity.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ServiceIdentity.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["email"] = state ? state.email : undefined;
            resourceInputs["folder"] = state ? state.folder : undefined;
            resourceInputs["member"] = state ? state.member : undefined;
            resourceInputs["service"] = state ? state.service : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.folder === undefined) && !opts.urn) {
                throw new Error("Missing required property 'folder'");
            }
            if ((!args || args.service === undefined) && !opts.urn) {
                throw new Error("Missing required property 'service'");
            }
            resourceInputs["folder"] = args ? args.folder : undefined;
            resourceInputs["service"] = args ? args.service : undefined;
            resourceInputs["email"] = undefined /*out*/;
            resourceInputs["member"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(ServiceIdentity.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ServiceIdentity = ServiceIdentity;
/** @internal */
ServiceIdentity.__pulumiType = 'gcp:folder/serviceIdentity:ServiceIdentity';
