// @pulumi/gcp@8.41.1/cloudquota/squotaAdjusterSettings.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQuotaAdjusterSettings = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * ## Example Usage
 *
 * ### Cloudquotas Quota Adjuster Settings Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const adjusterSettings = new gcp.cloudquota.SQuotaAdjusterSettings("adjuster_settings", {
 *     parent: "projects/104740170505",
 *     enablement: "ENABLED",
 * });
 * ```
 *
 * ## Import
 *
 * QuotaAdjusterSettings can be imported using any of these accepted formats:
 *
 * * `{{parent}}/locations/global/quotaAdjusterSettings`
 *
 * When using the `pulumi import` command, QuotaAdjusterSettings can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:cloudquota/sQuotaAdjusterSettings:SQuotaAdjusterSettings default {{parent}}/locations/global/quotaAdjusterSettings
 * ```
 */
class SQuotaAdjusterSettings extends pulumi.CustomResource {
    /**
     * Get an existing SQuotaAdjusterSettings resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new SQuotaAdjusterSettings(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of SQuotaAdjusterSettings.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === SQuotaAdjusterSettings.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["effectiveContainer"] = state ? state.effectiveContainer : undefined;
            resourceInputs["effectiveEnablement"] = state ? state.effectiveEnablement : undefined;
            resourceInputs["enablement"] = state ? state.enablement : undefined;
            resourceInputs["inherited"] = state ? state.inherited : undefined;
            resourceInputs["inheritedFrom"] = state ? state.inheritedFrom : undefined;
            resourceInputs["parent"] = state ? state.parent : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.enablement === undefined) && !opts.urn) {
                throw new Error("Missing required property 'enablement'");
            }
            resourceInputs["enablement"] = args ? args.enablement : undefined;
            resourceInputs["parent"] = args ? args.parent : undefined;
            resourceInputs["effectiveContainer"] = undefined /*out*/;
            resourceInputs["effectiveEnablement"] = undefined /*out*/;
            resourceInputs["inherited"] = undefined /*out*/;
            resourceInputs["inheritedFrom"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(SQuotaAdjusterSettings.__pulumiType, name, resourceInputs, opts);
    }
}
exports.SQuotaAdjusterSettings = SQuotaAdjusterSettings;
/** @internal */
SQuotaAdjusterSettings.__pulumiType = 'gcp:cloudquota/sQuotaAdjusterSettings:SQuotaAdjusterSettings';
