// @pulumi/gcp@8.41.1/clouddeploy/deployPolicy.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeployPolicy = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A `DeployPolicy` inhibits manual or DeployPolicy-driven actions within a Delivery Pipeline or Target.
 *
 * To get more information about DeployPolicy, see:
 *
 * * [API documentation](https://cloud.google.com/deploy/docs/api/reference/rest/v1/projects.locations.deployPolicies)
 * * How-to Guides
 *     * [Restrict deploy behavior using policies](https://cloud.google.com/deploy/docs/deploy-policy)
 *
 * ## Example Usage
 *
 * ### Clouddeploy Deploy Policy Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const b_deploy_policy = new gcp.clouddeploy.DeployPolicy("b-deploy-policy", {
 *     name: "cd-policy",
 *     location: "us-central1",
 *     selectors: [{
 *         deliveryPipeline: {
 *             id: "cd-pipeline",
 *         },
 *     }],
 *     rules: [{
 *         rolloutRestriction: {
 *             id: "rule",
 *             timeWindows: {
 *                 timeZone: "America/Los_Angeles",
 *                 weeklyWindows: [{
 *                     startTime: {
 *                         hours: 0,
 *                         minutes: 0,
 *                     },
 *                     endTime: {
 *                         hours: 24,
 *                         minutes: 0,
 *                     },
 *                 }],
 *             },
 *         },
 *     }],
 * });
 * ```
 * ### Clouddeploy Deploy Policy Full
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const f_deploy_policy = new gcp.clouddeploy.DeployPolicy("f-deploy-policy", {
 *     name: "cd-policy",
 *     location: "us-central1",
 *     annotations: {
 *         my_first_annotation: "example-annotation-1",
 *         my_second_annotation: "example-annotation-2",
 *     },
 *     labels: {
 *         my_first_label: "example-label-1",
 *         my_second_label: "example-label-2",
 *     },
 *     description: "policy resource",
 *     selectors: [
 *         {
 *             deliveryPipeline: {
 *                 id: "cd-pipeline",
 *                 labels: {
 *                     foo: "bar",
 *                 },
 *             },
 *         },
 *         {
 *             target: {
 *                 id: "dev",
 *                 labels: {
 *                     foo: "bar",
 *                 },
 *             },
 *         },
 *     ],
 *     suspended: true,
 *     rules: [
 *         {
 *             rolloutRestriction: {
 *                 id: "rule",
 *                 timeWindows: {
 *                     timeZone: "America/Los_Angeles",
 *                     weeklyWindows: [{
 *                         startTime: {
 *                             hours: 0,
 *                             minutes: 0,
 *                         },
 *                         endTime: {
 *                             hours: 13,
 *                             minutes: 0,
 *                         },
 *                     }],
 *                 },
 *             },
 *         },
 *         {
 *             rolloutRestriction: {
 *                 id: "rule2",
 *                 invokers: ["USER"],
 *                 actions: ["CREATE"],
 *                 timeWindows: {
 *                     timeZone: "America/Los_Angeles",
 *                     weeklyWindows: [{
 *                         startTime: {
 *                             hours: 13,
 *                             minutes: 0,
 *                         },
 *                         endTime: {
 *                             hours: 14,
 *                             minutes: 0,
 *                             seconds: 0,
 *                             nanos: 0,
 *                         },
 *                         daysOfWeeks: ["MONDAY"],
 *                     }],
 *                     oneTimeWindows: [{
 *                         startTime: {
 *                             hours: 0,
 *                             minutes: 0,
 *                         },
 *                         endTime: {
 *                             hours: 16,
 *                             minutes: 0,
 *                         },
 *                         startDate: {
 *                             year: 2019,
 *                             month: 1,
 *                             day: 1,
 *                         },
 *                         endDate: {
 *                             year: 2019,
 *                             month: 12,
 *                             day: 31,
 *                         },
 *                     }],
 *                 },
 *             },
 *         },
 *     ],
 * });
 * ```
 *
 * ## Import
 *
 * DeployPolicy can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/locations/{{location}}/deployPolicies/{{name}}`
 *
 * * `{{project}}/{{location}}/{{name}}`
 *
 * * `{{location}}/{{name}}`
 *
 * When using the `pulumi import` command, DeployPolicy can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:clouddeploy/deployPolicy:DeployPolicy default projects/{{project}}/locations/{{location}}/deployPolicies/{{name}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:clouddeploy/deployPolicy:DeployPolicy default {{project}}/{{location}}/{{name}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:clouddeploy/deployPolicy:DeployPolicy default {{location}}/{{name}}
 * ```
 */
class DeployPolicy extends pulumi.CustomResource {
    /**
     * Get an existing DeployPolicy resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new DeployPolicy(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of DeployPolicy.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === DeployPolicy.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["annotations"] = state ? state.annotations : undefined;
            resourceInputs["createTime"] = state ? state.createTime : undefined;
            resourceInputs["description"] = state ? state.description : undefined;
            resourceInputs["effectiveAnnotations"] = state ? state.effectiveAnnotations : undefined;
            resourceInputs["effectiveLabels"] = state ? state.effectiveLabels : undefined;
            resourceInputs["etag"] = state ? state.etag : undefined;
            resourceInputs["labels"] = state ? state.labels : undefined;
            resourceInputs["location"] = state ? state.location : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["pulumiLabels"] = state ? state.pulumiLabels : undefined;
            resourceInputs["rules"] = state ? state.rules : undefined;
            resourceInputs["selectors"] = state ? state.selectors : undefined;
            resourceInputs["suspended"] = state ? state.suspended : undefined;
            resourceInputs["uid"] = state ? state.uid : undefined;
            resourceInputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.location === undefined) && !opts.urn) {
                throw new Error("Missing required property 'location'");
            }
            if ((!args || args.rules === undefined) && !opts.urn) {
                throw new Error("Missing required property 'rules'");
            }
            if ((!args || args.selectors === undefined) && !opts.urn) {
                throw new Error("Missing required property 'selectors'");
            }
            resourceInputs["annotations"] = args ? args.annotations : undefined;
            resourceInputs["description"] = args ? args.description : undefined;
            resourceInputs["labels"] = args ? args.labels : undefined;
            resourceInputs["location"] = args ? args.location : undefined;
            resourceInputs["name"] = args ? args.name : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["rules"] = args ? args.rules : undefined;
            resourceInputs["selectors"] = args ? args.selectors : undefined;
            resourceInputs["suspended"] = args ? args.suspended : undefined;
            resourceInputs["createTime"] = undefined /*out*/;
            resourceInputs["effectiveAnnotations"] = undefined /*out*/;
            resourceInputs["effectiveLabels"] = undefined /*out*/;
            resourceInputs["etag"] = undefined /*out*/;
            resourceInputs["pulumiLabels"] = undefined /*out*/;
            resourceInputs["uid"] = undefined /*out*/;
            resourceInputs["updateTime"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const secretOpts = { additionalSecretOutputs: ["effectiveLabels", "pulumiLabels"] };
        opts = pulumi.mergeOptions(opts, secretOpts);
        super(DeployPolicy.__pulumiType, name, resourceInputs, opts);
    }
}
exports.DeployPolicy = DeployPolicy;
/** @internal */
DeployPolicy.__pulumiType = 'gcp:clouddeploy/deployPolicy:DeployPolicy';
