// @pulumi/gcp@8.41.1/apigee/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.TargetServer = exports.SyncAuthorization = exports.SharedflowDeployment = exports.Sharedflow = exports.SecurityProfileV2 = exports.SecurityMonitoringCondition = exports.SecurityAction = exports.Organization = exports.NatAddress = exports.KeystoresAliasesSelfSignedCert = exports.KeystoresAliasesPkcs12 = exports.KeystoresAliasesKeyCertFile = exports.InstanceAttachment = exports.Instance = exports.getEnvironmentIamPolicyOutput = exports.getEnvironmentIamPolicy = exports.Flowhook = exports.EnvironmentKeyvaluemapsEntries = exports.EnvironmentKeyvaluemaps = exports.EnvironmentIamPolicy = exports.EnvironmentIamMember = exports.EnvironmentIamBinding = exports.EnvironmentAddonsConfig = exports.Environment = exports.EnvReferences = exports.EnvKeystore = exports.EnvGroupAttachment = exports.EnvGroup = exports.EndpointAttachment = exports.DnsZone = exports.Developer = exports.ControlPlaneAccess = exports.AppGroup = exports.ApiProduct = exports.Api = exports.AddonsConfig = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.AddonsConfig = null;
utilities.lazyLoad(exports, ["AddonsConfig"], () => require("./addonsConfig"));
exports.Api = null;
utilities.lazyLoad(exports, ["Api"], () => require("./api"));
exports.ApiProduct = null;
utilities.lazyLoad(exports, ["ApiProduct"], () => require("./apiProduct"));
exports.AppGroup = null;
utilities.lazyLoad(exports, ["AppGroup"], () => require("./appGroup"));
exports.ControlPlaneAccess = null;
utilities.lazyLoad(exports, ["ControlPlaneAccess"], () => require("./controlPlaneAccess"));
exports.Developer = null;
utilities.lazyLoad(exports, ["Developer"], () => require("./developer"));
exports.DnsZone = null;
utilities.lazyLoad(exports, ["DnsZone"], () => require("./dnsZone"));
exports.EndpointAttachment = null;
utilities.lazyLoad(exports, ["EndpointAttachment"], () => require("./endpointAttachment"));
exports.EnvGroup = null;
utilities.lazyLoad(exports, ["EnvGroup"], () => require("./envGroup"));
exports.EnvGroupAttachment = null;
utilities.lazyLoad(exports, ["EnvGroupAttachment"], () => require("./envGroupAttachment"));
exports.EnvKeystore = null;
utilities.lazyLoad(exports, ["EnvKeystore"], () => require("./envKeystore"));
exports.EnvReferences = null;
utilities.lazyLoad(exports, ["EnvReferences"], () => require("./envReferences"));
exports.Environment = null;
utilities.lazyLoad(exports, ["Environment"], () => require("./environment"));
exports.EnvironmentAddonsConfig = null;
utilities.lazyLoad(exports, ["EnvironmentAddonsConfig"], () => require("./environmentAddonsConfig"));
exports.EnvironmentIamBinding = null;
utilities.lazyLoad(exports, ["EnvironmentIamBinding"], () => require("./environmentIamBinding"));
exports.EnvironmentIamMember = null;
utilities.lazyLoad(exports, ["EnvironmentIamMember"], () => require("./environmentIamMember"));
exports.EnvironmentIamPolicy = null;
utilities.lazyLoad(exports, ["EnvironmentIamPolicy"], () => require("./environmentIamPolicy"));
exports.EnvironmentKeyvaluemaps = null;
utilities.lazyLoad(exports, ["EnvironmentKeyvaluemaps"], () => require("./environmentKeyvaluemaps"));
exports.EnvironmentKeyvaluemapsEntries = null;
utilities.lazyLoad(exports, ["EnvironmentKeyvaluemapsEntries"], () => require("./environmentKeyvaluemapsEntries"));
exports.Flowhook = null;
utilities.lazyLoad(exports, ["Flowhook"], () => require("./flowhook"));
exports.getEnvironmentIamPolicy = null;
exports.getEnvironmentIamPolicyOutput = null;
utilities.lazyLoad(exports, ["getEnvironmentIamPolicy", "getEnvironmentIamPolicyOutput"], () => require("./getEnvironmentIamPolicy"));
exports.Instance = null;
utilities.lazyLoad(exports, ["Instance"], () => require("./instance"));
exports.InstanceAttachment = null;
utilities.lazyLoad(exports, ["InstanceAttachment"], () => require("./instanceAttachment"));
exports.KeystoresAliasesKeyCertFile = null;
utilities.lazyLoad(exports, ["KeystoresAliasesKeyCertFile"], () => require("./keystoresAliasesKeyCertFile"));
exports.KeystoresAliasesPkcs12 = null;
utilities.lazyLoad(exports, ["KeystoresAliasesPkcs12"], () => require("./keystoresAliasesPkcs12"));
exports.KeystoresAliasesSelfSignedCert = null;
utilities.lazyLoad(exports, ["KeystoresAliasesSelfSignedCert"], () => require("./keystoresAliasesSelfSignedCert"));
exports.NatAddress = null;
utilities.lazyLoad(exports, ["NatAddress"], () => require("./natAddress"));
exports.Organization = null;
utilities.lazyLoad(exports, ["Organization"], () => require("./organization"));
exports.SecurityAction = null;
utilities.lazyLoad(exports, ["SecurityAction"], () => require("./securityAction"));
exports.SecurityMonitoringCondition = null;
utilities.lazyLoad(exports, ["SecurityMonitoringCondition"], () => require("./securityMonitoringCondition"));
exports.SecurityProfileV2 = null;
utilities.lazyLoad(exports, ["SecurityProfileV2"], () => require("./securityProfileV2"));
exports.Sharedflow = null;
utilities.lazyLoad(exports, ["Sharedflow"], () => require("./sharedflow"));
exports.SharedflowDeployment = null;
utilities.lazyLoad(exports, ["SharedflowDeployment"], () => require("./sharedflowDeployment"));
exports.SyncAuthorization = null;
utilities.lazyLoad(exports, ["SyncAuthorization"], () => require("./syncAuthorization"));
exports.TargetServer = null;
utilities.lazyLoad(exports, ["TargetServer"], () => require("./targetServer"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "gcp:apigee/addonsConfig:AddonsConfig":
                return new exports.AddonsConfig(name, undefined, { urn });
            case "gcp:apigee/api:Api":
                return new exports.Api(name, undefined, { urn });
            case "gcp:apigee/apiProduct:ApiProduct":
                return new exports.ApiProduct(name, undefined, { urn });
            case "gcp:apigee/appGroup:AppGroup":
                return new exports.AppGroup(name, undefined, { urn });
            case "gcp:apigee/controlPlaneAccess:ControlPlaneAccess":
                return new exports.ControlPlaneAccess(name, undefined, { urn });
            case "gcp:apigee/developer:Developer":
                return new exports.Developer(name, undefined, { urn });
            case "gcp:apigee/dnsZone:DnsZone":
                return new exports.DnsZone(name, undefined, { urn });
            case "gcp:apigee/endpointAttachment:EndpointAttachment":
                return new exports.EndpointAttachment(name, undefined, { urn });
            case "gcp:apigee/envGroup:EnvGroup":
                return new exports.EnvGroup(name, undefined, { urn });
            case "gcp:apigee/envGroupAttachment:EnvGroupAttachment":
                return new exports.EnvGroupAttachment(name, undefined, { urn });
            case "gcp:apigee/envKeystore:EnvKeystore":
                return new exports.EnvKeystore(name, undefined, { urn });
            case "gcp:apigee/envReferences:EnvReferences":
                return new exports.EnvReferences(name, undefined, { urn });
            case "gcp:apigee/environment:Environment":
                return new exports.Environment(name, undefined, { urn });
            case "gcp:apigee/environmentAddonsConfig:EnvironmentAddonsConfig":
                return new exports.EnvironmentAddonsConfig(name, undefined, { urn });
            case "gcp:apigee/environmentIamBinding:EnvironmentIamBinding":
                return new exports.EnvironmentIamBinding(name, undefined, { urn });
            case "gcp:apigee/environmentIamMember:EnvironmentIamMember":
                return new exports.EnvironmentIamMember(name, undefined, { urn });
            case "gcp:apigee/environmentIamPolicy:EnvironmentIamPolicy":
                return new exports.EnvironmentIamPolicy(name, undefined, { urn });
            case "gcp:apigee/environmentKeyvaluemaps:EnvironmentKeyvaluemaps":
                return new exports.EnvironmentKeyvaluemaps(name, undefined, { urn });
            case "gcp:apigee/environmentKeyvaluemapsEntries:EnvironmentKeyvaluemapsEntries":
                return new exports.EnvironmentKeyvaluemapsEntries(name, undefined, { urn });
            case "gcp:apigee/flowhook:Flowhook":
                return new exports.Flowhook(name, undefined, { urn });
            case "gcp:apigee/instance:Instance":
                return new exports.Instance(name, undefined, { urn });
            case "gcp:apigee/instanceAttachment:InstanceAttachment":
                return new exports.InstanceAttachment(name, undefined, { urn });
            case "gcp:apigee/keystoresAliasesKeyCertFile:KeystoresAliasesKeyCertFile":
                return new exports.KeystoresAliasesKeyCertFile(name, undefined, { urn });
            case "gcp:apigee/keystoresAliasesPkcs12:KeystoresAliasesPkcs12":
                return new exports.KeystoresAliasesPkcs12(name, undefined, { urn });
            case "gcp:apigee/keystoresAliasesSelfSignedCert:KeystoresAliasesSelfSignedCert":
                return new exports.KeystoresAliasesSelfSignedCert(name, undefined, { urn });
            case "gcp:apigee/natAddress:NatAddress":
                return new exports.NatAddress(name, undefined, { urn });
            case "gcp:apigee/organization:Organization":
                return new exports.Organization(name, undefined, { urn });
            case "gcp:apigee/securityAction:SecurityAction":
                return new exports.SecurityAction(name, undefined, { urn });
            case "gcp:apigee/securityMonitoringCondition:SecurityMonitoringCondition":
                return new exports.SecurityMonitoringCondition(name, undefined, { urn });
            case "gcp:apigee/securityProfileV2:SecurityProfileV2":
                return new exports.SecurityProfileV2(name, undefined, { urn });
            case "gcp:apigee/sharedflow:Sharedflow":
                return new exports.Sharedflow(name, undefined, { urn });
            case "gcp:apigee/sharedflowDeployment:SharedflowDeployment":
                return new exports.SharedflowDeployment(name, undefined, { urn });
            case "gcp:apigee/syncAuthorization:SyncAuthorization":
                return new exports.SyncAuthorization(name, undefined, { urn });
            case "gcp:apigee/targetServer:TargetServer":
                return new exports.TargetServer(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("gcp", "apigee/addonsConfig", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/api", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/apiProduct", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/appGroup", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/controlPlaneAccess", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/developer", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/dnsZone", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/endpointAttachment", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/envGroup", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/envGroupAttachment", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/envKeystore", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/envReferences", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/environment", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/environmentAddonsConfig", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/environmentIamBinding", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/environmentIamMember", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/environmentIamPolicy", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/environmentKeyvaluemaps", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/environmentKeyvaluemapsEntries", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/flowhook", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/instance", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/instanceAttachment", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/keystoresAliasesKeyCertFile", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/keystoresAliasesPkcs12", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/keystoresAliasesSelfSignedCert", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/natAddress", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/organization", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/securityAction", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/securityMonitoringCondition", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/securityProfileV2", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/sharedflow", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/sharedflowDeployment", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/syncAuthorization", _module);
pulumi.runtime.registerResourceModule("gcp", "apigee/targetServer", _module);
