// @pulumi/gcp@8.41.1/apigee/securityMonitoringCondition.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityMonitoringCondition = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Security monitoring condition for risk assessment version 2 in Apigee.
 *
 * To get more information about SecurityMonitoringCondition, see:
 *
 * * [API documentation](https://cloud.google.com/apigee/docs/reference/apis/apigee/rest/v1/organizations.securityMonitoringConditions/create)
 * * How-to Guides
 *     * [Creating a security monitoring condition](https://cloud.google.com/apigee/docs/api-security/security-scores#monitoring-conditions)
 *
 * ## Example Usage
 *
 * ### Apigee Security Monitoring Condition Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const current = gcp.organizations.getClientConfig({});
 * const apigeeNetwork = new gcp.compute.Network("apigee_network", {name: "apigee-network"});
 * const apigeeRange = new gcp.compute.GlobalAddress("apigee_range", {
 *     name: "apigee-range",
 *     purpose: "VPC_PEERING",
 *     addressType: "INTERNAL",
 *     prefixLength: 16,
 *     network: apigeeNetwork.id,
 * });
 * const apigeeVpcConnection = new gcp.servicenetworking.Connection("apigee_vpc_connection", {
 *     network: apigeeNetwork.id,
 *     service: "servicenetworking.googleapis.com",
 *     reservedPeeringRanges: [apigeeRange.name],
 * });
 * const apigeeOrg = new gcp.apigee.Organization("apigee_org", {
 *     analyticsRegion: "us-central1",
 *     projectId: current.then(current => current.project),
 *     authorizedNetwork: apigeeNetwork.id,
 * }, {
 *     dependsOn: [apigeeVpcConnection],
 * });
 * const env = new gcp.apigee.Environment("env", {
 *     name: "my-environment",
 *     description: "Apigee Environment",
 *     displayName: "environment-1",
 *     orgId: apigeeOrg.id,
 * });
 * const apigeeOrgSecurityAddonsConfig = new gcp.apigee.AddonsConfig("apigee_org_security_addons_config", {
 *     org: apigeeOrg.name,
 *     addonsConfig: {
 *         apiSecurityConfig: {
 *             enabled: true,
 *         },
 *     },
 * });
 * const securityMonitoringCondition = new gcp.apigee.SecurityMonitoringCondition("security_monitoring_condition", {
 *     conditionId: "my-condition",
 *     orgId: apigeeOrg.id,
 *     profile: "google-default",
 *     scope: "my-environment",
 *     includeAllResources: {},
 * }, {
 *     dependsOn: [apigeeOrgSecurityAddonsConfig],
 * });
 * ```
 *
 * ## Import
 *
 * SecurityMonitoringCondition can be imported using any of these accepted formats:
 *
 * * `{{org_id}}/securityMonitoringConditions/{{condition_id}}`
 *
 * * `{{org_id}}/{{condition_id}}`
 *
 * When using the `pulumi import` command, SecurityMonitoringCondition can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:apigee/securityMonitoringCondition:SecurityMonitoringCondition default {{org_id}}/securityMonitoringConditions/{{condition_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:apigee/securityMonitoringCondition:SecurityMonitoringCondition default {{org_id}}/{{condition_id}}
 * ```
 */
class SecurityMonitoringCondition extends pulumi.CustomResource {
    /**
     * Get an existing SecurityMonitoringCondition resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new SecurityMonitoringCondition(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of SecurityMonitoringCondition.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === SecurityMonitoringCondition.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["conditionId"] = state ? state.conditionId : undefined;
            resourceInputs["createTime"] = state ? state.createTime : undefined;
            resourceInputs["includeAllResources"] = state ? state.includeAllResources : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["orgId"] = state ? state.orgId : undefined;
            resourceInputs["profile"] = state ? state.profile : undefined;
            resourceInputs["scope"] = state ? state.scope : undefined;
            resourceInputs["totalDeployedResources"] = state ? state.totalDeployedResources : undefined;
            resourceInputs["totalMonitoredResources"] = state ? state.totalMonitoredResources : undefined;
            resourceInputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.conditionId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'conditionId'");
            }
            if ((!args || args.orgId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'orgId'");
            }
            if ((!args || args.profile === undefined) && !opts.urn) {
                throw new Error("Missing required property 'profile'");
            }
            if ((!args || args.scope === undefined) && !opts.urn) {
                throw new Error("Missing required property 'scope'");
            }
            resourceInputs["conditionId"] = args ? args.conditionId : undefined;
            resourceInputs["includeAllResources"] = args ? args.includeAllResources : undefined;
            resourceInputs["orgId"] = args ? args.orgId : undefined;
            resourceInputs["profile"] = args ? args.profile : undefined;
            resourceInputs["scope"] = args ? args.scope : undefined;
            resourceInputs["createTime"] = undefined /*out*/;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["totalDeployedResources"] = undefined /*out*/;
            resourceInputs["totalMonitoredResources"] = undefined /*out*/;
            resourceInputs["updateTime"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(SecurityMonitoringCondition.__pulumiType, name, resourceInputs, opts);
    }
}
exports.SecurityMonitoringCondition = SecurityMonitoringCondition;
/** @internal */
SecurityMonitoringCondition.__pulumiType = 'gcp:apigee/securityMonitoringCondition:SecurityMonitoringCondition';
