// @pulumi/gcp@8.41.1/backupdisasterrecovery/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceConfig = exports.ManagementServer = exports.getManagementServerOutput = exports.getManagementServer = exports.getDataSourceOutput = exports.getDataSource = exports.getBackupVaultOutput = exports.getBackupVault = exports.getBackupPlanAssociationOutput = exports.getBackupPlanAssociation = exports.getBackupPlanOutput = exports.getBackupPlan = exports.getBackupOutput = exports.getBackup = exports.BackupVault = exports.BackupPlanAssociation = exports.BackupPlan = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.BackupPlan = null;
utilities.lazyLoad(exports, ["BackupPlan"], () => require("./backupPlan"));
exports.BackupPlanAssociation = null;
utilities.lazyLoad(exports, ["BackupPlanAssociation"], () => require("./backupPlanAssociation"));
exports.BackupVault = null;
utilities.lazyLoad(exports, ["BackupVault"], () => require("./backupVault"));
exports.getBackup = null;
exports.getBackupOutput = null;
utilities.lazyLoad(exports, ["getBackup", "getBackupOutput"], () => require("./getBackup"));
exports.getBackupPlan = null;
exports.getBackupPlanOutput = null;
utilities.lazyLoad(exports, ["getBackupPlan", "getBackupPlanOutput"], () => require("./getBackupPlan"));
exports.getBackupPlanAssociation = null;
exports.getBackupPlanAssociationOutput = null;
utilities.lazyLoad(exports, ["getBackupPlanAssociation", "getBackupPlanAssociationOutput"], () => require("./getBackupPlanAssociation"));
exports.getBackupVault = null;
exports.getBackupVaultOutput = null;
utilities.lazyLoad(exports, ["getBackupVault", "getBackupVaultOutput"], () => require("./getBackupVault"));
exports.getDataSource = null;
exports.getDataSourceOutput = null;
utilities.lazyLoad(exports, ["getDataSource", "getDataSourceOutput"], () => require("./getDataSource"));
exports.getManagementServer = null;
exports.getManagementServerOutput = null;
utilities.lazyLoad(exports, ["getManagementServer", "getManagementServerOutput"], () => require("./getManagementServer"));
exports.ManagementServer = null;
utilities.lazyLoad(exports, ["ManagementServer"], () => require("./managementServer"));
exports.ServiceConfig = null;
utilities.lazyLoad(exports, ["ServiceConfig"], () => require("./serviceConfig"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "gcp:backupdisasterrecovery/backupPlan:BackupPlan":
                return new exports.BackupPlan(name, undefined, { urn });
            case "gcp:backupdisasterrecovery/backupPlanAssociation:BackupPlanAssociation":
                return new exports.BackupPlanAssociation(name, undefined, { urn });
            case "gcp:backupdisasterrecovery/backupVault:BackupVault":
                return new exports.BackupVault(name, undefined, { urn });
            case "gcp:backupdisasterrecovery/managementServer:ManagementServer":
                return new exports.ManagementServer(name, undefined, { urn });
            case "gcp:backupdisasterrecovery/serviceConfig:ServiceConfig":
                return new exports.ServiceConfig(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("gcp", "backupdisasterrecovery/backupPlan", _module);
pulumi.runtime.registerResourceModule("gcp", "backupdisasterrecovery/backupPlanAssociation", _module);
pulumi.runtime.registerResourceModule("gcp", "backupdisasterrecovery/backupVault", _module);
pulumi.runtime.registerResourceModule("gcp", "backupdisasterrecovery/managementServer", _module);
pulumi.runtime.registerResourceModule("gcp", "backupdisasterrecovery/serviceConfig", _module);
