// @pulumi/gcp@8.41.1/compute/previewFeature.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreviewFeature = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * ## Example Usage
 *
 * ### Preview Feature Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const gcePreviewFeature = new gcp.compute.PreviewFeature("gce_preview_feature", {
 *     name: "alpha-api-access",
 *     activationStatus: "DISABLED",
 *     rolloutOperation: {
 *         rolloutInput: {
 *             predefinedRolloutPlan: "ROLLOUT_PLAN_FAST_ROLLOUT",
 *         },
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * PreviewFeature can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/global/previewFeatures/{{name}}`
 *
 * * `{{project}}/{{name}}`
 *
 * * `{{name}}`
 *
 * When using the `pulumi import` command, PreviewFeature can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:compute/previewFeature:PreviewFeature default projects/{{project}}/global/previewFeatures/{{name}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:compute/previewFeature:PreviewFeature default {{project}}/{{name}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:compute/previewFeature:PreviewFeature default {{name}}
 * ```
 */
class PreviewFeature extends pulumi.CustomResource {
    /**
     * Get an existing PreviewFeature resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new PreviewFeature(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of PreviewFeature.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === PreviewFeature.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["activationStatus"] = state ? state.activationStatus : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["rolloutOperation"] = state ? state.rolloutOperation : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.activationStatus === undefined) && !opts.urn) {
                throw new Error("Missing required property 'activationStatus'");
            }
            resourceInputs["activationStatus"] = args ? args.activationStatus : undefined;
            resourceInputs["name"] = args ? args.name : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["rolloutOperation"] = args ? args.rolloutOperation : undefined;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(PreviewFeature.__pulumiType, name, resourceInputs, opts);
    }
}
exports.PreviewFeature = PreviewFeature;
/** @internal */
PreviewFeature.__pulumiType = 'gcp:compute/previewFeature:PreviewFeature';
