// @pulumi/gcp@8.41.1/compute/regionResizeRequest.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegionResizeRequest = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * ## Example Usage
 *
 * ### Compute Rmig Resize Request
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const a3Dws = new gcp.compute.RegionInstanceTemplate("a3_dws", {
 *     name: "a3-dws",
 *     region: "us-central1",
 *     description: "This template is used to create a mig instance that is compatible with DWS resize requests.",
 *     instanceDescription: "A3 GPU",
 *     machineType: "a3-highgpu-8g",
 *     canIpForward: false,
 *     scheduling: {
 *         automaticRestart: false,
 *         onHostMaintenance: "TERMINATE",
 *     },
 *     disks: [{
 *         sourceImage: "cos-cloud/cos-121-lts",
 *         autoDelete: true,
 *         boot: true,
 *         diskType: "pd-ssd",
 *         diskSizeGb: 960,
 *         mode: "READ_WRITE",
 *     }],
 *     guestAccelerators: [{
 *         type: "nvidia-h100-80gb",
 *         count: 8,
 *     }],
 *     reservationAffinity: {
 *         type: "NO_RESERVATION",
 *     },
 *     shieldedInstanceConfig: {
 *         enableVtpm: true,
 *         enableIntegrityMonitoring: true,
 *     },
 *     networkInterfaces: [{
 *         network: "default",
 *     }],
 * });
 * const a3DwsRegionInstanceGroupManager = new gcp.compute.RegionInstanceGroupManager("a3_dws", {
 *     name: "a3-dws",
 *     baseInstanceName: "a3-dws",
 *     region: "us-central1",
 *     versions: [{
 *         instanceTemplate: a3Dws.selfLink,
 *     }],
 *     instanceLifecyclePolicy: {
 *         defaultActionOnFailure: "DO_NOTHING",
 *     },
 *     distributionPolicyTargetShape: "ANY_SINGLE_ZONE",
 *     distributionPolicyZones: [
 *         "us-central1-a",
 *         "us-central1-b",
 *         "us-central1-c",
 *         "us-central1-f",
 *     ],
 *     updatePolicy: {
 *         instanceRedistributionType: "NONE",
 *         type: "OPPORTUNISTIC",
 *         minimalAction: "REPLACE",
 *         maxSurgeFixed: 0,
 *         maxUnavailableFixed: 6,
 *     },
 *     waitForInstances: false,
 * });
 * const a3ResizeRequest = new gcp.compute.RegionResizeRequest("a3_resize_request", {
 *     name: "a3-dws",
 *     instanceGroupManager: a3DwsRegionInstanceGroupManager.name,
 *     region: "us-central1",
 *     description: "Test resize request resource",
 *     resizeBy: 2,
 *     requestedRunDuration: {
 *         seconds: "14400",
 *         nanos: 0,
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * RegionResizeRequest can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/regions/{{region}}/instanceGroupManagers/{{instance_group_manager}}/resizeRequests/{{name}}`
 *
 * * `{{project}}/{{region}}/{{instance_group_manager}}/{{name}}`
 *
 * * `{{region}}/{{instance_group_manager}}/{{name}}`
 *
 * * `{{instance_group_manager}}/{{name}}`
 *
 * When using the `pulumi import` command, RegionResizeRequest can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:compute/regionResizeRequest:RegionResizeRequest default projects/{{project}}/regions/{{region}}/instanceGroupManagers/{{instance_group_manager}}/resizeRequests/{{name}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:compute/regionResizeRequest:RegionResizeRequest default {{project}}/{{region}}/{{instance_group_manager}}/{{name}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:compute/regionResizeRequest:RegionResizeRequest default {{region}}/{{instance_group_manager}}/{{name}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:compute/regionResizeRequest:RegionResizeRequest default {{instance_group_manager}}/{{name}}
 * ```
 */
class RegionResizeRequest extends pulumi.CustomResource {
    /**
     * Get an existing RegionResizeRequest resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new RegionResizeRequest(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of RegionResizeRequest.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === RegionResizeRequest.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["creationTimestamp"] = state ? state.creationTimestamp : undefined;
            resourceInputs["description"] = state ? state.description : undefined;
            resourceInputs["instanceGroupManager"] = state ? state.instanceGroupManager : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["region"] = state ? state.region : undefined;
            resourceInputs["requestedRunDuration"] = state ? state.requestedRunDuration : undefined;
            resourceInputs["resizeBy"] = state ? state.resizeBy : undefined;
            resourceInputs["state"] = state ? state.state : undefined;
            resourceInputs["statuses"] = state ? state.statuses : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.instanceGroupManager === undefined) && !opts.urn) {
                throw new Error("Missing required property 'instanceGroupManager'");
            }
            if ((!args || args.resizeBy === undefined) && !opts.urn) {
                throw new Error("Missing required property 'resizeBy'");
            }
            resourceInputs["description"] = args ? args.description : undefined;
            resourceInputs["instanceGroupManager"] = args ? args.instanceGroupManager : undefined;
            resourceInputs["name"] = args ? args.name : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["region"] = args ? args.region : undefined;
            resourceInputs["requestedRunDuration"] = args ? args.requestedRunDuration : undefined;
            resourceInputs["resizeBy"] = args ? args.resizeBy : undefined;
            resourceInputs["creationTimestamp"] = undefined /*out*/;
            resourceInputs["state"] = undefined /*out*/;
            resourceInputs["statuses"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(RegionResizeRequest.__pulumiType, name, resourceInputs, opts);
    }
}
exports.RegionResizeRequest = RegionResizeRequest;
/** @internal */
RegionResizeRequest.__pulumiType = 'gcp:compute/regionResizeRequest:RegionResizeRequest';
