// @pulumi/gcp@8.41.1/spanner/backupSchedule.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupSchedule = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A backup schedule for a Cloud Spanner Database.
 * This resource is owned by the database it is backing up, and is deleted along with the database.
 * The actual backups are not though.
 *
 * To get more information about BackupSchedule, see:
 *
 * * [API documentation](https://cloud.google.com/spanner/docs/reference/rest/v1/projects.instances.databases.backupSchedules)
 * * How-to Guides
 *     * [Official Documentation](https://cloud.google.com/spanner/docs/backup)
 *
 * > **Warning:** This resource creates a Spanner Backup Schedule on a project that already has
 * a Spanner database.
 * This resource is owned by the database it is backing up, and is deleted along
 * with the database. The actual backups are not though.
 *
 * ## Example Usage
 *
 * ### Spanner Backup Schedule Daily Full
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const main = new gcp.spanner.Instance("main", {
 *     name: "instance-id",
 *     config: "regional-europe-west1",
 *     displayName: "main-instance",
 *     numNodes: 1,
 * });
 * const database = new gcp.spanner.Database("database", {
 *     instance: main.name,
 *     name: "database-id",
 *     versionRetentionPeriod: "3d",
 *     ddls: [
 *         "CREATE TABLE t1 (t1 INT64 NOT NULL,) PRIMARY KEY(t1)",
 *         "CREATE TABLE t2 (t2 INT64 NOT NULL,) PRIMARY KEY(t2)",
 *     ],
 *     deletionProtection: true,
 * });
 * const full_backup = new gcp.spanner.BackupSchedule("full-backup", {
 *     instance: main.name,
 *     database: database.name,
 *     name: "backup-schedule-id",
 *     retentionDuration: "31620000s",
 *     spec: {
 *         cronSpec: {
 *             text: "0 12 * * *",
 *         },
 *     },
 *     fullBackupSpec: {},
 *     encryptionConfig: {
 *         encryptionType: "USE_DATABASE_ENCRYPTION",
 *     },
 * });
 * ```
 * ### Spanner Backup Schedule Daily Incremental
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const main = new gcp.spanner.Instance("main", {
 *     name: "instance-id",
 *     config: "regional-europe-west1",
 *     displayName: "main-instance",
 *     numNodes: 1,
 *     edition: "ENTERPRISE",
 * });
 * const database = new gcp.spanner.Database("database", {
 *     instance: main.name,
 *     name: "database-id",
 *     versionRetentionPeriod: "3d",
 *     ddls: [
 *         "CREATE TABLE t1 (t1 INT64 NOT NULL,) PRIMARY KEY(t1)",
 *         "CREATE TABLE t2 (t2 INT64 NOT NULL,) PRIMARY KEY(t2)",
 *     ],
 *     deletionProtection: true,
 * });
 * const incremental_backup = new gcp.spanner.BackupSchedule("incremental-backup", {
 *     instance: main.name,
 *     database: database.name,
 *     name: "backup-schedule-id",
 *     retentionDuration: "31620000s",
 *     spec: {
 *         cronSpec: {
 *             text: "0 12 * * *",
 *         },
 *     },
 *     incrementalBackupSpec: {},
 *     encryptionConfig: {
 *         encryptionType: "GOOGLE_DEFAULT_ENCRYPTION",
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * BackupSchedule can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/instances/{{instance}}/databases/{{database}}/backupSchedules/{{name}}`
 *
 * * `{{project}}/{{instance}}/{{database}}/{{name}}`
 *
 * * `{{instance}}/{{database}}/{{name}}`
 *
 * When using the `pulumi import` command, BackupSchedule can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:spanner/backupSchedule:BackupSchedule default projects/{{project}}/instances/{{instance}}/databases/{{database}}/backupSchedules/{{name}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:spanner/backupSchedule:BackupSchedule default {{project}}/{{instance}}/{{database}}/{{name}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:spanner/backupSchedule:BackupSchedule default {{instance}}/{{database}}/{{name}}
 * ```
 */
class BackupSchedule extends pulumi.CustomResource {
    /**
     * Get an existing BackupSchedule resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new BackupSchedule(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of BackupSchedule.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === BackupSchedule.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["database"] = state ? state.database : undefined;
            resourceInputs["encryptionConfig"] = state ? state.encryptionConfig : undefined;
            resourceInputs["fullBackupSpec"] = state ? state.fullBackupSpec : undefined;
            resourceInputs["incrementalBackupSpec"] = state ? state.incrementalBackupSpec : undefined;
            resourceInputs["instance"] = state ? state.instance : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["retentionDuration"] = state ? state.retentionDuration : undefined;
            resourceInputs["spec"] = state ? state.spec : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.database === undefined) && !opts.urn) {
                throw new Error("Missing required property 'database'");
            }
            if ((!args || args.instance === undefined) && !opts.urn) {
                throw new Error("Missing required property 'instance'");
            }
            if ((!args || args.retentionDuration === undefined) && !opts.urn) {
                throw new Error("Missing required property 'retentionDuration'");
            }
            resourceInputs["database"] = args ? args.database : undefined;
            resourceInputs["encryptionConfig"] = args ? args.encryptionConfig : undefined;
            resourceInputs["fullBackupSpec"] = args ? args.fullBackupSpec : undefined;
            resourceInputs["incrementalBackupSpec"] = args ? args.incrementalBackupSpec : undefined;
            resourceInputs["instance"] = args ? args.instance : undefined;
            resourceInputs["name"] = args ? args.name : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["retentionDuration"] = args ? args.retentionDuration : undefined;
            resourceInputs["spec"] = args ? args.spec : undefined;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(BackupSchedule.__pulumiType, name, resourceInputs, opts);
    }
}
exports.BackupSchedule = BackupSchedule;
/** @internal */
BackupSchedule.__pulumiType = 'gcp:spanner/backupSchedule:BackupSchedule';
