// @pulumi/gcp@8.41.1/securitycenter/folderSccBigQueryExport.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.FolderSccBigQueryExport = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A Cloud Security Command Center (Cloud SCC) Big Query Export Config.
 * It represents exporting Security Command Center data, including assets, findings, and security marks
 * to a BigQuery instance.
 *
 * > **Note:** In order to use Cloud SCC resources, your organization must be enrolled
 * in [SCC Standard/Premium](https://cloud.google.com/security-command-center/docs/quickstart-security-command-center).
 * Without doing so, you may run into errors during resource creation.
 *
 * To get more information about FolderSccBigQueryExport, see:
 *
 * * [API documentation](https://cloud.google.com/security-command-center/docs/reference/rest/v1/folders.bigQueryExports)
 * * How-to Guides
 *     * [Official Documentation](https://cloud.google.com/security-command-center/docs/how-to-analyze-findings-in-big-query)
 *
 * ## Example Usage
 *
 * ### Scc Folder Big Query Export Config Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const folder = new gcp.organizations.Folder("folder", {
 *     parent: "organizations/123456789",
 *     displayName: "folder-name",
 *     deletionProtection: false,
 * });
 * const _default = new gcp.bigquery.Dataset("default", {
 *     datasetId: "my_dataset_id",
 *     friendlyName: "test",
 *     description: "This is a test description",
 *     location: "US",
 *     defaultTableExpirationMs: 3600000,
 *     defaultPartitionExpirationMs: null,
 *     labels: {
 *         env: "default",
 *     },
 * });
 * const customBigQueryExportConfig = new gcp.securitycenter.FolderSccBigQueryExport("custom_big_query_export_config", {
 *     bigQueryExportId: "my-export",
 *     folder: folder.folderId,
 *     dataset: _default.id,
 *     description: "Cloud Security Command Center Findings Big Query Export Config",
 *     filter: "state=\"ACTIVE\" AND NOT mute=\"MUTED\"",
 * });
 * ```
 *
 * ## Import
 *
 * FolderSccBigQueryExport can be imported using any of these accepted formats:
 *
 * * `folders/{{folder}}/bigQueryExports/{{big_query_export_id}}`
 *
 * * `{{folder}}/{{big_query_export_id}}`
 *
 * When using the `pulumi import` command, FolderSccBigQueryExport can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:securitycenter/folderSccBigQueryExport:FolderSccBigQueryExport default folders/{{folder}}/bigQueryExports/{{big_query_export_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:securitycenter/folderSccBigQueryExport:FolderSccBigQueryExport default {{folder}}/{{big_query_export_id}}
 * ```
 */
class FolderSccBigQueryExport extends pulumi.CustomResource {
    /**
     * Get an existing FolderSccBigQueryExport resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new FolderSccBigQueryExport(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of FolderSccBigQueryExport.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === FolderSccBigQueryExport.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["bigQueryExportId"] = state ? state.bigQueryExportId : undefined;
            resourceInputs["createTime"] = state ? state.createTime : undefined;
            resourceInputs["dataset"] = state ? state.dataset : undefined;
            resourceInputs["description"] = state ? state.description : undefined;
            resourceInputs["filter"] = state ? state.filter : undefined;
            resourceInputs["folder"] = state ? state.folder : undefined;
            resourceInputs["mostRecentEditor"] = state ? state.mostRecentEditor : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["principal"] = state ? state.principal : undefined;
            resourceInputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.bigQueryExportId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'bigQueryExportId'");
            }
            if ((!args || args.dataset === undefined) && !opts.urn) {
                throw new Error("Missing required property 'dataset'");
            }
            if ((!args || args.description === undefined) && !opts.urn) {
                throw new Error("Missing required property 'description'");
            }
            if ((!args || args.filter === undefined) && !opts.urn) {
                throw new Error("Missing required property 'filter'");
            }
            if ((!args || args.folder === undefined) && !opts.urn) {
                throw new Error("Missing required property 'folder'");
            }
            resourceInputs["bigQueryExportId"] = args ? args.bigQueryExportId : undefined;
            resourceInputs["dataset"] = args ? args.dataset : undefined;
            resourceInputs["description"] = args ? args.description : undefined;
            resourceInputs["filter"] = args ? args.filter : undefined;
            resourceInputs["folder"] = args ? args.folder : undefined;
            resourceInputs["createTime"] = undefined /*out*/;
            resourceInputs["mostRecentEditor"] = undefined /*out*/;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["principal"] = undefined /*out*/;
            resourceInputs["updateTime"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(FolderSccBigQueryExport.__pulumiType, name, resourceInputs, opts);
    }
}
exports.FolderSccBigQueryExport = FolderSccBigQueryExport;
/** @internal */
FolderSccBigQueryExport.__pulumiType = 'gcp:securitycenter/folderSccBigQueryExport:FolderSccBigQueryExport';
