// @pulumi/gcp@8.41.1/parametermanager/parameter.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Parameter = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A Parameter resource is a logical parameter.
 *
 * To get more information about Parameter, see:
 *
 * * [API documentation](https://cloud.google.com/secret-manager/parameter-manager/docs/reference/rest/v1/projects.locations.parameters)
 *
 * ## Example Usage
 *
 * ### Parameter Config Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const parameter_basic = new gcp.parametermanager.Parameter("parameter-basic", {parameterId: "parameter"});
 * ```
 * ### Parameter With Format
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const parameter_with_format = new gcp.parametermanager.Parameter("parameter-with-format", {
 *     parameterId: "parameter",
 *     format: "JSON",
 * });
 * ```
 * ### Parameter With Labels
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const parameter_with_labels = new gcp.parametermanager.Parameter("parameter-with-labels", {
 *     parameterId: "parameter",
 *     labels: {
 *         key1: "val1",
 *         key2: "val2",
 *         key3: "val3",
 *         key4: "val4",
 *         key5: "val5",
 *     },
 * });
 * ```
 * ### Parameter With Kms Key
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const project = gcp.organizations.getProject({});
 * const parameter_with_kms_key = new gcp.parametermanager.Parameter("parameter-with-kms-key", {
 *     parameterId: "parameter",
 *     kmsKey: "kms-key",
 * });
 * ```
 *
 * ## Import
 *
 * Parameter can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/locations/global/parameters/{{parameter_id}}`
 *
 * * `{{project}}/{{parameter_id}}`
 *
 * * `{{parameter_id}}`
 *
 * When using the `pulumi import` command, Parameter can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:parametermanager/parameter:Parameter default projects/{{project}}/locations/global/parameters/{{parameter_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:parametermanager/parameter:Parameter default {{project}}/{{parameter_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:parametermanager/parameter:Parameter default {{parameter_id}}
 * ```
 */
class Parameter extends pulumi.CustomResource {
    /**
     * Get an existing Parameter resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Parameter(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Parameter.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Parameter.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["createTime"] = state ? state.createTime : undefined;
            resourceInputs["effectiveLabels"] = state ? state.effectiveLabels : undefined;
            resourceInputs["format"] = state ? state.format : undefined;
            resourceInputs["kmsKey"] = state ? state.kmsKey : undefined;
            resourceInputs["labels"] = state ? state.labels : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["parameterId"] = state ? state.parameterId : undefined;
            resourceInputs["policyMembers"] = state ? state.policyMembers : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["pulumiLabels"] = state ? state.pulumiLabels : undefined;
            resourceInputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.parameterId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'parameterId'");
            }
            resourceInputs["format"] = args ? args.format : undefined;
            resourceInputs["kmsKey"] = args ? args.kmsKey : undefined;
            resourceInputs["labels"] = args ? args.labels : undefined;
            resourceInputs["parameterId"] = args ? args.parameterId : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["createTime"] = undefined /*out*/;
            resourceInputs["effectiveLabels"] = undefined /*out*/;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["policyMembers"] = undefined /*out*/;
            resourceInputs["pulumiLabels"] = undefined /*out*/;
            resourceInputs["updateTime"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const secretOpts = { additionalSecretOutputs: ["effectiveLabels", "pulumiLabels"] };
        opts = pulumi.mergeOptions(opts, secretOpts);
        super(Parameter.__pulumiType, name, resourceInputs, opts);
    }
}
exports.Parameter = Parameter;
/** @internal */
Parameter.__pulumiType = 'gcp:parametermanager/parameter:Parameter';
