// @pulumi/gcp@8.41.1/parametermanager/parameterVersion.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParameterVersion = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A Parameter Version resource that stores the actual value of the parameter.
 *
 * To get more information about ParameterVersion, see:
 *
 * * [API documentation](https://cloud.google.com/secret-manager/parameter-manager/docs/reference/rest/v1/projects.locations.parameters.versions)
 *
 * ## Example Usage
 *
 * ### Parameter Version Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const parameter_basic = new gcp.parametermanager.Parameter("parameter-basic", {parameterId: "parameter"});
 * const parameter_version_basic = new gcp.parametermanager.ParameterVersion("parameter-version-basic", {
 *     parameter: parameter_basic.id,
 *     parameterVersionId: "parameter_version",
 *     parameterData: "app-parameter-version-data",
 * });
 * ```
 * ### Parameter Version With Json Format
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const parameter_basic = new gcp.parametermanager.Parameter("parameter-basic", {
 *     parameterId: "parameter",
 *     format: "JSON",
 * });
 * const parameter_version_with_json_format = new gcp.parametermanager.ParameterVersion("parameter-version-with-json-format", {
 *     parameter: parameter_basic.id,
 *     parameterVersionId: "parameter_version",
 *     parameterData: JSON.stringify({
 *         key1: "val1",
 *         key2: "val2",
 *     }),
 * });
 * ```
 * ### Parameter Version With Kms Key
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const project = gcp.organizations.getProject({});
 * const parameter_basic = new gcp.parametermanager.Parameter("parameter-basic", {
 *     parameterId: "parameter",
 *     kmsKey: "kms-key",
 * });
 * const parameter_version_with_kms_key = new gcp.parametermanager.ParameterVersion("parameter-version-with-kms-key", {
 *     parameter: parameter_basic.id,
 *     parameterVersionId: "parameter_version",
 *     parameterData: "app-parameter-version-data",
 * });
 * ```
 * ### Parameter Version With Json Format With File
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 * import * as std from "@pulumi/std";
 *
 * const parameter_basic = new gcp.parametermanager.Parameter("parameter-basic", {
 *     parameterId: "parameter",
 *     format: "JSON",
 * });
 * const parameter_version_with_json_format_with_file = new gcp.parametermanager.ParameterVersion("parameter-version-with-json-format-with-file", {
 *     parameter: parameter_basic.id,
 *     parameterVersionId: "parameter_version",
 *     parameterData: std.file({
 *         input: "parameter-json-data.json",
 *     }).then(invoke => invoke.result),
 * });
 * ```
 * ### Parameter Version With Yaml Format With File
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 * import * as std from "@pulumi/std";
 *
 * const parameter_basic = new gcp.parametermanager.Parameter("parameter-basic", {
 *     parameterId: "parameter",
 *     format: "YAML",
 * });
 * const parameter_version_with_yaml_format_with_file = new gcp.parametermanager.ParameterVersion("parameter-version-with-yaml-format-with-file", {
 *     parameter: parameter_basic.id,
 *     parameterVersionId: "parameter_version",
 *     parameterData: std.file({
 *         input: "parameter-yaml-data.yaml",
 *     }).then(invoke => invoke.result),
 * });
 * ```
 *
 * ## Import
 *
 * ParameterVersion can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/locations/global/parameters/{{parameter_id}}/versions/{{parameter_version_id}}`
 *
 * When using the `pulumi import` command, ParameterVersion can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:parametermanager/parameterVersion:ParameterVersion default projects/{{project}}/locations/global/parameters/{{parameter_id}}/versions/{{parameter_version_id}}
 * ```
 */
class ParameterVersion extends pulumi.CustomResource {
    /**
     * Get an existing ParameterVersion resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ParameterVersion(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ParameterVersion.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ParameterVersion.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["createTime"] = state ? state.createTime : undefined;
            resourceInputs["disabled"] = state ? state.disabled : undefined;
            resourceInputs["kmsKeyVersion"] = state ? state.kmsKeyVersion : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["parameter"] = state ? state.parameter : undefined;
            resourceInputs["parameterData"] = state ? state.parameterData : undefined;
            resourceInputs["parameterVersionId"] = state ? state.parameterVersionId : undefined;
            resourceInputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.parameter === undefined) && !opts.urn) {
                throw new Error("Missing required property 'parameter'");
            }
            if ((!args || args.parameterData === undefined) && !opts.urn) {
                throw new Error("Missing required property 'parameterData'");
            }
            if ((!args || args.parameterVersionId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'parameterVersionId'");
            }
            resourceInputs["disabled"] = args ? args.disabled : undefined;
            resourceInputs["parameter"] = args ? args.parameter : undefined;
            resourceInputs["parameterData"] = (args === null || args === void 0 ? void 0 : args.parameterData) ? pulumi.secret(args.parameterData) : undefined;
            resourceInputs["parameterVersionId"] = args ? args.parameterVersionId : undefined;
            resourceInputs["createTime"] = undefined /*out*/;
            resourceInputs["kmsKeyVersion"] = undefined /*out*/;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["updateTime"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const secretOpts = { additionalSecretOutputs: ["parameterData"] };
        opts = pulumi.mergeOptions(opts, secretOpts);
        super(ParameterVersion.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ParameterVersion = ParameterVersion;
/** @internal */
ParameterVersion.__pulumiType = 'gcp:parametermanager/parameterVersion:ParameterVersion';
