// @pulumi/gcp@8.41.1/tpu/v2queuedResource.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.V2QueuedResource = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * ## Example Usage
 *
 * ### Tpu V2 Queued Resource Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const qr = new gcp.tpu.V2QueuedResource("qr", {
 *     name: "test-qr",
 *     zone: "us-central1-c",
 *     project: "my-project-name",
 *     tpu: {
 *         nodeSpecs: [{
 *             parent: "projects/my-project-name/locations/us-central1-c",
 *             nodeId: "test-tpu",
 *             node: {
 *                 runtimeVersion: "tpu-vm-tf-2.13.0",
 *                 acceleratorType: "v2-8",
 *                 description: "Text description of the TPU.",
 *             },
 *         }],
 *     },
 * });
 * ```
 * ### Tpu V2 Queued Resource Full
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const network = new gcp.compute.Network("network", {
 *     name: "tpu-net",
 *     autoCreateSubnetworks: false,
 * });
 * const subnet = new gcp.compute.Subnetwork("subnet", {
 *     name: "tpu-subnet",
 *     ipCidrRange: "10.0.0.0/16",
 *     region: "us-central1",
 *     network: network.id,
 * });
 * const qr = new gcp.tpu.V2QueuedResource("qr", {
 *     name: "test-qr",
 *     zone: "us-central1-c",
 *     project: "my-project-name",
 *     tpu: {
 *         nodeSpecs: [{
 *             parent: "projects/my-project-name/locations/us-central1-c",
 *             nodeId: "test-tpu",
 *             node: {
 *                 runtimeVersion: "tpu-vm-tf-2.13.0",
 *                 acceleratorType: "v2-8",
 *                 description: "Text description of the TPU.",
 *                 networkConfig: {
 *                     canIpForward: true,
 *                     enableExternalIps: true,
 *                     network: network.id,
 *                     subnetwork: subnet.id,
 *                     queueCount: 32,
 *                 },
 *             },
 *         }],
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * QueuedResource can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/locations/{{zone}}/queuedResources/{{name}}`
 *
 * * `{{project}}/{{zone}}/{{name}}`
 *
 * * `{{zone}}/{{name}}`
 *
 * * `{{name}}`
 *
 * When using the `pulumi import` command, QueuedResource can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:tpu/v2QueuedResource:V2QueuedResource default projects/{{project}}/locations/{{zone}}/queuedResources/{{name}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:tpu/v2QueuedResource:V2QueuedResource default {{project}}/{{zone}}/{{name}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:tpu/v2QueuedResource:V2QueuedResource default {{zone}}/{{name}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:tpu/v2QueuedResource:V2QueuedResource default {{name}}
 * ```
 */
class V2QueuedResource extends pulumi.CustomResource {
    /**
     * Get an existing V2QueuedResource resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new V2QueuedResource(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of V2QueuedResource.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === V2QueuedResource.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["tpu"] = state ? state.tpu : undefined;
            resourceInputs["zone"] = state ? state.zone : undefined;
        }
        else {
            const args = argsOrState;
            resourceInputs["name"] = args ? args.name : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["tpu"] = args ? args.tpu : undefined;
            resourceInputs["zone"] = args ? args.zone : undefined;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(V2QueuedResource.__pulumiType, name, resourceInputs, opts);
    }
}
exports.V2QueuedResource = V2QueuedResource;
/** @internal */
V2QueuedResource.__pulumiType = 'gcp:tpu/v2QueuedResource:V2QueuedResource';
