// @pulumi/gcp@8.41.1/logging/logScope.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogScope = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Describes a group of resources to read log entries from
 *
 * To get more information about LogScope, see:
 *
 * * [API documentation](https://cloud.google.com/logging/docs/reference/v2/rest/v2/projects.locations.logScopes)
 * * How-to Guides
 *     * [Official Documentation](https://cloud.google.com/logging/docs/apis)
 *
 * ## Example Usage
 *
 * ### Logging Log Scope Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const loggingLogScope = new gcp.logging.LogScope("logging_log_scope", {
 *     parent: "projects/my-project-name",
 *     location: "global",
 *     name: "projects/my-project-name/locations/global/logScopes/my-log-scope",
 *     resourceNames: [
 *         "projects/my-project-name",
 *         "projects/my-project-name/locations/global/buckets/_Default/views/view1",
 *         "projects/my-project-name/locations/global/buckets/_Default/views/view2",
 *     ],
 *     description: "A log scope configured with Terraform",
 * });
 * ```
 *
 * ## Import
 *
 * LogScope can be imported using any of these accepted formats:
 *
 * * `{{parent}}/locations/{{location}}/logScopes/{{name}}`
 *
 * When using the `pulumi import` command, LogScope can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:logging/logScope:LogScope default {{parent}}/locations/{{location}}/logScopes/{{name}}
 * ```
 */
class LogScope extends pulumi.CustomResource {
    /**
     * Get an existing LogScope resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new LogScope(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of LogScope.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === LogScope.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["createTime"] = state ? state.createTime : undefined;
            resourceInputs["description"] = state ? state.description : undefined;
            resourceInputs["location"] = state ? state.location : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["parent"] = state ? state.parent : undefined;
            resourceInputs["resourceNames"] = state ? state.resourceNames : undefined;
            resourceInputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.resourceNames === undefined) && !opts.urn) {
                throw new Error("Missing required property 'resourceNames'");
            }
            resourceInputs["description"] = args ? args.description : undefined;
            resourceInputs["location"] = args ? args.location : undefined;
            resourceInputs["name"] = args ? args.name : undefined;
            resourceInputs["parent"] = args ? args.parent : undefined;
            resourceInputs["resourceNames"] = args ? args.resourceNames : undefined;
            resourceInputs["createTime"] = undefined /*out*/;
            resourceInputs["updateTime"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(LogScope.__pulumiType, name, resourceInputs, opts);
    }
}
exports.LogScope = LogScope;
/** @internal */
LogScope.__pulumiType = 'gcp:logging/logScope:LogScope';
