// @pulumi/gcp@8.41.1/discoveryengine/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.TargetSite = exports.Sitemap = exports.SearchEngine = exports.Schema = exports.RecommendationEngine = exports.DataStore = exports.CmekConfig = exports.ChatEngine = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.ChatEngine = null;
utilities.lazyLoad(exports, ["ChatEngine"], () => require("./chatEngine"));
exports.CmekConfig = null;
utilities.lazyLoad(exports, ["CmekConfig"], () => require("./cmekConfig"));
exports.DataStore = null;
utilities.lazyLoad(exports, ["DataStore"], () => require("./dataStore"));
exports.RecommendationEngine = null;
utilities.lazyLoad(exports, ["RecommendationEngine"], () => require("./recommendationEngine"));
exports.Schema = null;
utilities.lazyLoad(exports, ["Schema"], () => require("./schema"));
exports.SearchEngine = null;
utilities.lazyLoad(exports, ["SearchEngine"], () => require("./searchEngine"));
exports.Sitemap = null;
utilities.lazyLoad(exports, ["Sitemap"], () => require("./sitemap"));
exports.TargetSite = null;
utilities.lazyLoad(exports, ["TargetSite"], () => require("./targetSite"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "gcp:discoveryengine/chatEngine:ChatEngine":
                return new exports.ChatEngine(name, undefined, { urn });
            case "gcp:discoveryengine/cmekConfig:CmekConfig":
                return new exports.CmekConfig(name, undefined, { urn });
            case "gcp:discoveryengine/dataStore:DataStore":
                return new exports.DataStore(name, undefined, { urn });
            case "gcp:discoveryengine/recommendationEngine:RecommendationEngine":
                return new exports.RecommendationEngine(name, undefined, { urn });
            case "gcp:discoveryengine/schema:Schema":
                return new exports.Schema(name, undefined, { urn });
            case "gcp:discoveryengine/searchEngine:SearchEngine":
                return new exports.SearchEngine(name, undefined, { urn });
            case "gcp:discoveryengine/sitemap:Sitemap":
                return new exports.Sitemap(name, undefined, { urn });
            case "gcp:discoveryengine/targetSite:TargetSite":
                return new exports.TargetSite(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("gcp", "discoveryengine/chatEngine", _module);
pulumi.runtime.registerResourceModule("gcp", "discoveryengine/cmekConfig", _module);
pulumi.runtime.registerResourceModule("gcp", "discoveryengine/dataStore", _module);
pulumi.runtime.registerResourceModule("gcp", "discoveryengine/recommendationEngine", _module);
pulumi.runtime.registerResourceModule("gcp", "discoveryengine/schema", _module);
pulumi.runtime.registerResourceModule("gcp", "discoveryengine/searchEngine", _module);
pulumi.runtime.registerResourceModule("gcp", "discoveryengine/sitemap", _module);
pulumi.runtime.registerResourceModule("gcp", "discoveryengine/targetSite", _module);
