// @pulumi/gcp@8.41.1/colab/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Schedule = exports.RuntimeTemplateIamPolicy = exports.RuntimeTemplateIamMember = exports.RuntimeTemplateIamBinding = exports.RuntimeTemplate = exports.Runtime = exports.NotebookExecution = exports.getRuntimeTemplateIamPolicyOutput = exports.getRuntimeTemplateIamPolicy = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.getRuntimeTemplateIamPolicy = null;
exports.getRuntimeTemplateIamPolicyOutput = null;
utilities.lazyLoad(exports, ["getRuntimeTemplateIamPolicy", "getRuntimeTemplateIamPolicyOutput"], () => require("./getRuntimeTemplateIamPolicy"));
exports.NotebookExecution = null;
utilities.lazyLoad(exports, ["NotebookExecution"], () => require("./notebookExecution"));
exports.Runtime = null;
utilities.lazyLoad(exports, ["Runtime"], () => require("./runtime"));
exports.RuntimeTemplate = null;
utilities.lazyLoad(exports, ["RuntimeTemplate"], () => require("./runtimeTemplate"));
exports.RuntimeTemplateIamBinding = null;
utilities.lazyLoad(exports, ["RuntimeTemplateIamBinding"], () => require("./runtimeTemplateIamBinding"));
exports.RuntimeTemplateIamMember = null;
utilities.lazyLoad(exports, ["RuntimeTemplateIamMember"], () => require("./runtimeTemplateIamMember"));
exports.RuntimeTemplateIamPolicy = null;
utilities.lazyLoad(exports, ["RuntimeTemplateIamPolicy"], () => require("./runtimeTemplateIamPolicy"));
exports.Schedule = null;
utilities.lazyLoad(exports, ["Schedule"], () => require("./schedule"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "gcp:colab/notebookExecution:NotebookExecution":
                return new exports.NotebookExecution(name, undefined, { urn });
            case "gcp:colab/runtime:Runtime":
                return new exports.Runtime(name, undefined, { urn });
            case "gcp:colab/runtimeTemplate:RuntimeTemplate":
                return new exports.RuntimeTemplate(name, undefined, { urn });
            case "gcp:colab/runtimeTemplateIamBinding:RuntimeTemplateIamBinding":
                return new exports.RuntimeTemplateIamBinding(name, undefined, { urn });
            case "gcp:colab/runtimeTemplateIamMember:RuntimeTemplateIamMember":
                return new exports.RuntimeTemplateIamMember(name, undefined, { urn });
            case "gcp:colab/runtimeTemplateIamPolicy:RuntimeTemplateIamPolicy":
                return new exports.RuntimeTemplateIamPolicy(name, undefined, { urn });
            case "gcp:colab/schedule:Schedule":
                return new exports.Schedule(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("gcp", "colab/notebookExecution", _module);
pulumi.runtime.registerResourceModule("gcp", "colab/runtime", _module);
pulumi.runtime.registerResourceModule("gcp", "colab/runtimeTemplate", _module);
pulumi.runtime.registerResourceModule("gcp", "colab/runtimeTemplateIamBinding", _module);
pulumi.runtime.registerResourceModule("gcp", "colab/runtimeTemplateIamMember", _module);
pulumi.runtime.registerResourceModule("gcp", "colab/runtimeTemplateIamPolicy", _module);
pulumi.runtime.registerResourceModule("gcp", "colab/schedule", _module);
