// @pulumi/gcp@8.41.1/iam/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkloadIdentityPoolProvider = exports.WorkloadIdentityPoolNamespace = exports.WorkloadIdentityPoolManagedIdentity = exports.WorkloadIdentityPoolIamPolicy = exports.WorkloadIdentityPoolIamMember = exports.WorkloadIdentityPoolIamBinding = exports.WorkloadIdentityPool = exports.WorkforcePoolProviderKey = exports.WorkforcePoolProvider = exports.WorkforcePoolIamPolicy = exports.WorkforcePoolIamMember = exports.WorkforcePoolIamBinding = exports.WorkforcePool = exports.ProjectsPolicyBinding = exports.PrincipalAccessBoundaryPolicy = exports.OrganizationsPolicyBinding = exports.OauthClientCredential = exports.OauthClient = exports.getWorkloadIdentityPoolProviderOutput = exports.getWorkloadIdentityPoolProvider = exports.getWorkloadIdentityPoolIamPolicyOutput = exports.getWorkloadIdentityPoolIamPolicy = exports.getWorkloadIdentityPoolOutput = exports.getWorkloadIdentityPool = exports.getWorkforcePoolIamPolicyOutput = exports.getWorkforcePoolIamPolicy = exports.getTestablePermissionsOutput = exports.getTestablePermissions = exports.getRuleOutput = exports.getRule = exports.FoldersPolicyBinding = exports.DenyPolicy = exports.AccessBoundaryPolicy = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.AccessBoundaryPolicy = null;
utilities.lazyLoad(exports, ["AccessBoundaryPolicy"], () => require("./accessBoundaryPolicy"));
exports.DenyPolicy = null;
utilities.lazyLoad(exports, ["DenyPolicy"], () => require("./denyPolicy"));
exports.FoldersPolicyBinding = null;
utilities.lazyLoad(exports, ["FoldersPolicyBinding"], () => require("./foldersPolicyBinding"));
exports.getRule = null;
exports.getRuleOutput = null;
utilities.lazyLoad(exports, ["getRule", "getRuleOutput"], () => require("./getRule"));
exports.getTestablePermissions = null;
exports.getTestablePermissionsOutput = null;
utilities.lazyLoad(exports, ["getTestablePermissions", "getTestablePermissionsOutput"], () => require("./getTestablePermissions"));
exports.getWorkforcePoolIamPolicy = null;
exports.getWorkforcePoolIamPolicyOutput = null;
utilities.lazyLoad(exports, ["getWorkforcePoolIamPolicy", "getWorkforcePoolIamPolicyOutput"], () => require("./getWorkforcePoolIamPolicy"));
exports.getWorkloadIdentityPool = null;
exports.getWorkloadIdentityPoolOutput = null;
utilities.lazyLoad(exports, ["getWorkloadIdentityPool", "getWorkloadIdentityPoolOutput"], () => require("./getWorkloadIdentityPool"));
exports.getWorkloadIdentityPoolIamPolicy = null;
exports.getWorkloadIdentityPoolIamPolicyOutput = null;
utilities.lazyLoad(exports, ["getWorkloadIdentityPoolIamPolicy", "getWorkloadIdentityPoolIamPolicyOutput"], () => require("./getWorkloadIdentityPoolIamPolicy"));
exports.getWorkloadIdentityPoolProvider = null;
exports.getWorkloadIdentityPoolProviderOutput = null;
utilities.lazyLoad(exports, ["getWorkloadIdentityPoolProvider", "getWorkloadIdentityPoolProviderOutput"], () => require("./getWorkloadIdentityPoolProvider"));
exports.OauthClient = null;
utilities.lazyLoad(exports, ["OauthClient"], () => require("./oauthClient"));
exports.OauthClientCredential = null;
utilities.lazyLoad(exports, ["OauthClientCredential"], () => require("./oauthClientCredential"));
exports.OrganizationsPolicyBinding = null;
utilities.lazyLoad(exports, ["OrganizationsPolicyBinding"], () => require("./organizationsPolicyBinding"));
exports.PrincipalAccessBoundaryPolicy = null;
utilities.lazyLoad(exports, ["PrincipalAccessBoundaryPolicy"], () => require("./principalAccessBoundaryPolicy"));
exports.ProjectsPolicyBinding = null;
utilities.lazyLoad(exports, ["ProjectsPolicyBinding"], () => require("./projectsPolicyBinding"));
exports.WorkforcePool = null;
utilities.lazyLoad(exports, ["WorkforcePool"], () => require("./workforcePool"));
exports.WorkforcePoolIamBinding = null;
utilities.lazyLoad(exports, ["WorkforcePoolIamBinding"], () => require("./workforcePoolIamBinding"));
exports.WorkforcePoolIamMember = null;
utilities.lazyLoad(exports, ["WorkforcePoolIamMember"], () => require("./workforcePoolIamMember"));
exports.WorkforcePoolIamPolicy = null;
utilities.lazyLoad(exports, ["WorkforcePoolIamPolicy"], () => require("./workforcePoolIamPolicy"));
exports.WorkforcePoolProvider = null;
utilities.lazyLoad(exports, ["WorkforcePoolProvider"], () => require("./workforcePoolProvider"));
exports.WorkforcePoolProviderKey = null;
utilities.lazyLoad(exports, ["WorkforcePoolProviderKey"], () => require("./workforcePoolProviderKey"));
exports.WorkloadIdentityPool = null;
utilities.lazyLoad(exports, ["WorkloadIdentityPool"], () => require("./workloadIdentityPool"));
exports.WorkloadIdentityPoolIamBinding = null;
utilities.lazyLoad(exports, ["WorkloadIdentityPoolIamBinding"], () => require("./workloadIdentityPoolIamBinding"));
exports.WorkloadIdentityPoolIamMember = null;
utilities.lazyLoad(exports, ["WorkloadIdentityPoolIamMember"], () => require("./workloadIdentityPoolIamMember"));
exports.WorkloadIdentityPoolIamPolicy = null;
utilities.lazyLoad(exports, ["WorkloadIdentityPoolIamPolicy"], () => require("./workloadIdentityPoolIamPolicy"));
exports.WorkloadIdentityPoolManagedIdentity = null;
utilities.lazyLoad(exports, ["WorkloadIdentityPoolManagedIdentity"], () => require("./workloadIdentityPoolManagedIdentity"));
exports.WorkloadIdentityPoolNamespace = null;
utilities.lazyLoad(exports, ["WorkloadIdentityPoolNamespace"], () => require("./workloadIdentityPoolNamespace"));
exports.WorkloadIdentityPoolProvider = null;
utilities.lazyLoad(exports, ["WorkloadIdentityPoolProvider"], () => require("./workloadIdentityPoolProvider"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "gcp:iam/accessBoundaryPolicy:AccessBoundaryPolicy":
                return new exports.AccessBoundaryPolicy(name, undefined, { urn });
            case "gcp:iam/denyPolicy:DenyPolicy":
                return new exports.DenyPolicy(name, undefined, { urn });
            case "gcp:iam/foldersPolicyBinding:FoldersPolicyBinding":
                return new exports.FoldersPolicyBinding(name, undefined, { urn });
            case "gcp:iam/oauthClient:OauthClient":
                return new exports.OauthClient(name, undefined, { urn });
            case "gcp:iam/oauthClientCredential:OauthClientCredential":
                return new exports.OauthClientCredential(name, undefined, { urn });
            case "gcp:iam/organizationsPolicyBinding:OrganizationsPolicyBinding":
                return new exports.OrganizationsPolicyBinding(name, undefined, { urn });
            case "gcp:iam/principalAccessBoundaryPolicy:PrincipalAccessBoundaryPolicy":
                return new exports.PrincipalAccessBoundaryPolicy(name, undefined, { urn });
            case "gcp:iam/projectsPolicyBinding:ProjectsPolicyBinding":
                return new exports.ProjectsPolicyBinding(name, undefined, { urn });
            case "gcp:iam/workforcePool:WorkforcePool":
                return new exports.WorkforcePool(name, undefined, { urn });
            case "gcp:iam/workforcePoolIamBinding:WorkforcePoolIamBinding":
                return new exports.WorkforcePoolIamBinding(name, undefined, { urn });
            case "gcp:iam/workforcePoolIamMember:WorkforcePoolIamMember":
                return new exports.WorkforcePoolIamMember(name, undefined, { urn });
            case "gcp:iam/workforcePoolIamPolicy:WorkforcePoolIamPolicy":
                return new exports.WorkforcePoolIamPolicy(name, undefined, { urn });
            case "gcp:iam/workforcePoolProvider:WorkforcePoolProvider":
                return new exports.WorkforcePoolProvider(name, undefined, { urn });
            case "gcp:iam/workforcePoolProviderKey:WorkforcePoolProviderKey":
                return new exports.WorkforcePoolProviderKey(name, undefined, { urn });
            case "gcp:iam/workloadIdentityPool:WorkloadIdentityPool":
                return new exports.WorkloadIdentityPool(name, undefined, { urn });
            case "gcp:iam/workloadIdentityPoolIamBinding:WorkloadIdentityPoolIamBinding":
                return new exports.WorkloadIdentityPoolIamBinding(name, undefined, { urn });
            case "gcp:iam/workloadIdentityPoolIamMember:WorkloadIdentityPoolIamMember":
                return new exports.WorkloadIdentityPoolIamMember(name, undefined, { urn });
            case "gcp:iam/workloadIdentityPoolIamPolicy:WorkloadIdentityPoolIamPolicy":
                return new exports.WorkloadIdentityPoolIamPolicy(name, undefined, { urn });
            case "gcp:iam/workloadIdentityPoolManagedIdentity:WorkloadIdentityPoolManagedIdentity":
                return new exports.WorkloadIdentityPoolManagedIdentity(name, undefined, { urn });
            case "gcp:iam/workloadIdentityPoolNamespace:WorkloadIdentityPoolNamespace":
                return new exports.WorkloadIdentityPoolNamespace(name, undefined, { urn });
            case "gcp:iam/workloadIdentityPoolProvider:WorkloadIdentityPoolProvider":
                return new exports.WorkloadIdentityPoolProvider(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("gcp", "iam/accessBoundaryPolicy", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/denyPolicy", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/foldersPolicyBinding", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/oauthClient", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/oauthClientCredential", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/organizationsPolicyBinding", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/principalAccessBoundaryPolicy", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/projectsPolicyBinding", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/workforcePool", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/workforcePoolIamBinding", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/workforcePoolIamMember", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/workforcePoolIamPolicy", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/workforcePoolProvider", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/workforcePoolProviderKey", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/workloadIdentityPool", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/workloadIdentityPoolIamBinding", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/workloadIdentityPoolIamMember", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/workloadIdentityPoolIamPolicy", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/workloadIdentityPoolManagedIdentity", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/workloadIdentityPoolNamespace", _module);
pulumi.runtime.registerResourceModule("gcp", "iam/workloadIdentityPoolProvider", _module);
