// @pulumi/gcp@8.41.1/networkmanagement/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcFlowLogsConfig = exports.getConnectivityTestRunOutput = exports.getConnectivityTestRun = exports.ConnectivityTest = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.ConnectivityTest = null;
utilities.lazyLoad(exports, ["ConnectivityTest"], () => require("./connectivityTest"));
exports.getConnectivityTestRun = null;
exports.getConnectivityTestRunOutput = null;
utilities.lazyLoad(exports, ["getConnectivityTestRun", "getConnectivityTestRunOutput"], () => require("./getConnectivityTestRun"));
exports.VpcFlowLogsConfig = null;
utilities.lazyLoad(exports, ["VpcFlowLogsConfig"], () => require("./vpcFlowLogsConfig"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "gcp:networkmanagement/connectivityTest:ConnectivityTest":
                return new exports.ConnectivityTest(name, undefined, { urn });
            case "gcp:networkmanagement/vpcFlowLogsConfig:VpcFlowLogsConfig":
                return new exports.VpcFlowLogsConfig(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("gcp", "networkmanagement/connectivityTest", _module);
pulumi.runtime.registerResourceModule("gcp", "networkmanagement/vpcFlowLogsConfig", _module);
