// @pulumi/gcp@8.41.1/dataproc/metastoreTableIamMember.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetastoreTableIamMember = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Three different resources help you manage your IAM policy for Dataproc Metastore Table. Each of these resources serves a different use case:
 *
 * * `gcp.dataproc.MetastoreTableIamPolicy`: Authoritative. Sets the IAM policy for the table and replaces any existing policy already attached.
 * * `gcp.dataproc.MetastoreTableIamBinding`: Authoritative for a given role. Updates the IAM policy to grant a role to a list of members. Other roles within the IAM policy for the table are preserved.
 * * `gcp.dataproc.MetastoreTableIamMember`: Non-authoritative. Updates the IAM policy to grant a role to a new member. Other members for the role for the table are preserved.
 *
 * A data source can be used to retrieve policy data in advent you do not need creation
 *
 * * `gcp.dataproc.MetastoreTableIamPolicy`: Retrieves the IAM policy for the table
 *
 * > **Note:** `gcp.dataproc.MetastoreTableIamPolicy` **cannot** be used in conjunction with `gcp.dataproc.MetastoreTableIamBinding` and `gcp.dataproc.MetastoreTableIamMember` or they will fight over what your policy should be.
 *
 * > **Note:** `gcp.dataproc.MetastoreTableIamBinding` resources **can be** used in conjunction with `gcp.dataproc.MetastoreTableIamMember` resources **only if** they do not grant privilege to the same role.
 *
 * ## gcp.dataproc.MetastoreTableIamPolicy
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const admin = gcp.organizations.getIAMPolicy({
 *     bindings: [{
 *         role: "roles/viewer",
 *         members: ["user:jane@example.com"],
 *     }],
 * });
 * const policy = new gcp.dataproc.MetastoreTableIamPolicy("policy", {
 *     project: dpmsService.project,
 *     location: dpmsService.location,
 *     serviceId: dpmsService.serviceId,
 *     databaseId: hive.hiveConfig[0].properties.database,
 *     table: hive.hiveConfig[0].properties.table,
 *     policyData: admin.then(admin => admin.policyData),
 * });
 * ```
 *
 * ## gcp.dataproc.MetastoreTableIamBinding
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const binding = new gcp.dataproc.MetastoreTableIamBinding("binding", {
 *     project: dpmsService.project,
 *     location: dpmsService.location,
 *     serviceId: dpmsService.serviceId,
 *     databaseId: hive.hiveConfig[0].properties.database,
 *     table: hive.hiveConfig[0].properties.table,
 *     role: "roles/viewer",
 *     members: ["user:jane@example.com"],
 * });
 * ```
 *
 * ## gcp.dataproc.MetastoreTableIamMember
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const member = new gcp.dataproc.MetastoreTableIamMember("member", {
 *     project: dpmsService.project,
 *     location: dpmsService.location,
 *     serviceId: dpmsService.serviceId,
 *     databaseId: hive.hiveConfig[0].properties.database,
 *     table: hive.hiveConfig[0].properties.table,
 *     role: "roles/viewer",
 *     member: "user:jane@example.com",
 * });
 * ```
 *
 * ## This resource supports User Project Overrides.
 *
 * -
 *
 * # IAM policy for Dataproc Metastore Table
 *
 * Three different resources help you manage your IAM policy for Dataproc Metastore Table. Each of these resources serves a different use case:
 *
 * * `gcp.dataproc.MetastoreTableIamPolicy`: Authoritative. Sets the IAM policy for the table and replaces any existing policy already attached.
 * * `gcp.dataproc.MetastoreTableIamBinding`: Authoritative for a given role. Updates the IAM policy to grant a role to a list of members. Other roles within the IAM policy for the table are preserved.
 * * `gcp.dataproc.MetastoreTableIamMember`: Non-authoritative. Updates the IAM policy to grant a role to a new member. Other members for the role for the table are preserved.
 *
 * A data source can be used to retrieve policy data in advent you do not need creation
 *
 * * `gcp.dataproc.MetastoreTableIamPolicy`: Retrieves the IAM policy for the table
 *
 * > **Note:** `gcp.dataproc.MetastoreTableIamPolicy` **cannot** be used in conjunction with `gcp.dataproc.MetastoreTableIamBinding` and `gcp.dataproc.MetastoreTableIamMember` or they will fight over what your policy should be.
 *
 * > **Note:** `gcp.dataproc.MetastoreTableIamBinding` resources **can be** used in conjunction with `gcp.dataproc.MetastoreTableIamMember` resources **only if** they do not grant privilege to the same role.
 *
 * ## gcp.dataproc.MetastoreTableIamPolicy
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const admin = gcp.organizations.getIAMPolicy({
 *     bindings: [{
 *         role: "roles/viewer",
 *         members: ["user:jane@example.com"],
 *     }],
 * });
 * const policy = new gcp.dataproc.MetastoreTableIamPolicy("policy", {
 *     project: dpmsService.project,
 *     location: dpmsService.location,
 *     serviceId: dpmsService.serviceId,
 *     databaseId: hive.hiveConfig[0].properties.database,
 *     table: hive.hiveConfig[0].properties.table,
 *     policyData: admin.then(admin => admin.policyData),
 * });
 * ```
 *
 * ## gcp.dataproc.MetastoreTableIamBinding
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const binding = new gcp.dataproc.MetastoreTableIamBinding("binding", {
 *     project: dpmsService.project,
 *     location: dpmsService.location,
 *     serviceId: dpmsService.serviceId,
 *     databaseId: hive.hiveConfig[0].properties.database,
 *     table: hive.hiveConfig[0].properties.table,
 *     role: "roles/viewer",
 *     members: ["user:jane@example.com"],
 * });
 * ```
 *
 * ## gcp.dataproc.MetastoreTableIamMember
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const member = new gcp.dataproc.MetastoreTableIamMember("member", {
 *     project: dpmsService.project,
 *     location: dpmsService.location,
 *     serviceId: dpmsService.serviceId,
 *     databaseId: hive.hiveConfig[0].properties.database,
 *     table: hive.hiveConfig[0].properties.table,
 *     role: "roles/viewer",
 *     member: "user:jane@example.com",
 * });
 * ```
 *
 * ## Import
 *
 * For all import syntaxes, the "resource in question" can take any of the following forms:
 *
 * * projects/{{project}}/locations/{{location}}/services/{{serviceId}}/databases/{{databaseId}}/tables/{{name}}
 *
 * * {{project}}/{{location}}/{{serviceId}}/{{databaseId}}/{{name}}
 *
 * * {{location}}/{{serviceId}}/{{databaseId}}/{{name}}
 *
 * * {{name}}
 *
 * Any variables not passed in the import command will be taken from the provider configuration.
 *
 * Dataproc Metastore table IAM resources can be imported using the resource identifiers, role, and member.
 *
 * IAM member imports use space-delimited identifiers: the resource in question, the role, and the member identity, e.g.
 *
 * ```sh
 * $ pulumi import gcp:dataproc/metastoreTableIamMember:MetastoreTableIamMember editor "projects/{{project}}/locations/{{location}}/services/{{serviceId}}/databases/{{databaseId}}/tables/{{table}} roles/viewer user:jane@example.com"
 * ```
 *
 * IAM binding imports use space-delimited identifiers: the resource in question and the role, e.g.
 *
 * ```sh
 * $ pulumi import gcp:dataproc/metastoreTableIamMember:MetastoreTableIamMember editor "projects/{{project}}/locations/{{location}}/services/{{serviceId}}/databases/{{databaseId}}/tables/{{table}} roles/viewer"
 * ```
 *
 * IAM policy imports use the identifier of the resource in question, e.g.
 *
 * ```sh
 * $ pulumi import gcp:dataproc/metastoreTableIamMember:MetastoreTableIamMember editor projects/{{project}}/locations/{{location}}/services/{{serviceId}}/databases/{{databaseId}}/tables/{{table}}
 * ```
 *
 * -> **Custom Roles** If you're importing a IAM resource with a custom role, make sure to use the
 *
 *  full name of the custom role, e.g. `[projects/my-project|organizations/my-org]/roles/my-custom-role`.
 */
class MetastoreTableIamMember extends pulumi.CustomResource {
    /**
     * Get an existing MetastoreTableIamMember resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new MetastoreTableIamMember(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of MetastoreTableIamMember.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === MetastoreTableIamMember.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["condition"] = state ? state.condition : undefined;
            resourceInputs["databaseId"] = state ? state.databaseId : undefined;
            resourceInputs["etag"] = state ? state.etag : undefined;
            resourceInputs["location"] = state ? state.location : undefined;
            resourceInputs["member"] = state ? state.member : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["role"] = state ? state.role : undefined;
            resourceInputs["serviceId"] = state ? state.serviceId : undefined;
            resourceInputs["table"] = state ? state.table : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.databaseId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'databaseId'");
            }
            if ((!args || args.member === undefined) && !opts.urn) {
                throw new Error("Missing required property 'member'");
            }
            if ((!args || args.role === undefined) && !opts.urn) {
                throw new Error("Missing required property 'role'");
            }
            if ((!args || args.serviceId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'serviceId'");
            }
            if ((!args || args.table === undefined) && !opts.urn) {
                throw new Error("Missing required property 'table'");
            }
            resourceInputs["condition"] = args ? args.condition : undefined;
            resourceInputs["databaseId"] = args ? args.databaseId : undefined;
            resourceInputs["location"] = args ? args.location : undefined;
            resourceInputs["member"] = args ? args.member : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["role"] = args ? args.role : undefined;
            resourceInputs["serviceId"] = args ? args.serviceId : undefined;
            resourceInputs["table"] = args ? args.table : undefined;
            resourceInputs["etag"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(MetastoreTableIamMember.__pulumiType, name, resourceInputs, opts);
    }
}
exports.MetastoreTableIamMember = MetastoreTableIamMember;
/** @internal */
MetastoreTableIamMember.__pulumiType = 'gcp:dataproc/metastoreTableIamMember:MetastoreTableIamMember';
