// @pulumi/gcp@8.41.1/dataproc/metastoreDatabaseIamMember.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetastoreDatabaseIamMember = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Three different resources help you manage your IAM policy for Dataproc Metastore Database. Each of these resources serves a different use case:
 *
 * * `gcp.dataproc.MetastoreDatabaseIamPolicy`: Authoritative. Sets the IAM policy for the database and replaces any existing policy already attached.
 * * `gcp.dataproc.MetastoreDatabaseIamBinding`: Authoritative for a given role. Updates the IAM policy to grant a role to a list of members. Other roles within the IAM policy for the database are preserved.
 * * `gcp.dataproc.MetastoreDatabaseIamMember`: Non-authoritative. Updates the IAM policy to grant a role to a new member. Other members for the role for the database are preserved.
 *
 * A data source can be used to retrieve policy data in advent you do not need creation
 *
 * * `gcp.dataproc.MetastoreDatabaseIamPolicy`: Retrieves the IAM policy for the database
 *
 * > **Note:** `gcp.dataproc.MetastoreDatabaseIamPolicy` **cannot** be used in conjunction with `gcp.dataproc.MetastoreDatabaseIamBinding` and `gcp.dataproc.MetastoreDatabaseIamMember` or they will fight over what your policy should be.
 *
 * > **Note:** `gcp.dataproc.MetastoreDatabaseIamBinding` resources **can be** used in conjunction with `gcp.dataproc.MetastoreDatabaseIamMember` resources **only if** they do not grant privilege to the same role.
 *
 * ## gcp.dataproc.MetastoreDatabaseIamPolicy
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const admin = gcp.organizations.getIAMPolicy({
 *     bindings: [{
 *         role: "roles/viewer",
 *         members: ["user:jane@example.com"],
 *     }],
 * });
 * const policy = new gcp.dataproc.MetastoreDatabaseIamPolicy("policy", {
 *     project: dpmsService.project,
 *     location: dpmsService.location,
 *     serviceId: dpmsService.serviceId,
 *     database: hive.hiveConfig[0].properties.database,
 *     policyData: admin.then(admin => admin.policyData),
 * });
 * ```
 *
 * ## gcp.dataproc.MetastoreDatabaseIamBinding
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const binding = new gcp.dataproc.MetastoreDatabaseIamBinding("binding", {
 *     project: dpmsService.project,
 *     location: dpmsService.location,
 *     serviceId: dpmsService.serviceId,
 *     database: hive.hiveConfig[0].properties.database,
 *     role: "roles/viewer",
 *     members: ["user:jane@example.com"],
 * });
 * ```
 *
 * ## gcp.dataproc.MetastoreDatabaseIamMember
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const member = new gcp.dataproc.MetastoreDatabaseIamMember("member", {
 *     project: dpmsService.project,
 *     location: dpmsService.location,
 *     serviceId: dpmsService.serviceId,
 *     database: hive.hiveConfig[0].properties.database,
 *     role: "roles/viewer",
 *     member: "user:jane@example.com",
 * });
 * ```
 *
 * ## This resource supports User Project Overrides.
 *
 * -
 *
 * # IAM policy for Dataproc Metastore Database
 *
 * Three different resources help you manage your IAM policy for Dataproc Metastore Database. Each of these resources serves a different use case:
 *
 * * `gcp.dataproc.MetastoreDatabaseIamPolicy`: Authoritative. Sets the IAM policy for the database and replaces any existing policy already attached.
 * * `gcp.dataproc.MetastoreDatabaseIamBinding`: Authoritative for a given role. Updates the IAM policy to grant a role to a list of members. Other roles within the IAM policy for the database are preserved.
 * * `gcp.dataproc.MetastoreDatabaseIamMember`: Non-authoritative. Updates the IAM policy to grant a role to a new member. Other members for the role for the database are preserved.
 *
 * A data source can be used to retrieve policy data in advent you do not need creation
 *
 * * `gcp.dataproc.MetastoreDatabaseIamPolicy`: Retrieves the IAM policy for the database
 *
 * > **Note:** `gcp.dataproc.MetastoreDatabaseIamPolicy` **cannot** be used in conjunction with `gcp.dataproc.MetastoreDatabaseIamBinding` and `gcp.dataproc.MetastoreDatabaseIamMember` or they will fight over what your policy should be.
 *
 * > **Note:** `gcp.dataproc.MetastoreDatabaseIamBinding` resources **can be** used in conjunction with `gcp.dataproc.MetastoreDatabaseIamMember` resources **only if** they do not grant privilege to the same role.
 *
 * ## gcp.dataproc.MetastoreDatabaseIamPolicy
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const admin = gcp.organizations.getIAMPolicy({
 *     bindings: [{
 *         role: "roles/viewer",
 *         members: ["user:jane@example.com"],
 *     }],
 * });
 * const policy = new gcp.dataproc.MetastoreDatabaseIamPolicy("policy", {
 *     project: dpmsService.project,
 *     location: dpmsService.location,
 *     serviceId: dpmsService.serviceId,
 *     database: hive.hiveConfig[0].properties.database,
 *     policyData: admin.then(admin => admin.policyData),
 * });
 * ```
 *
 * ## gcp.dataproc.MetastoreDatabaseIamBinding
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const binding = new gcp.dataproc.MetastoreDatabaseIamBinding("binding", {
 *     project: dpmsService.project,
 *     location: dpmsService.location,
 *     serviceId: dpmsService.serviceId,
 *     database: hive.hiveConfig[0].properties.database,
 *     role: "roles/viewer",
 *     members: ["user:jane@example.com"],
 * });
 * ```
 *
 * ## gcp.dataproc.MetastoreDatabaseIamMember
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const member = new gcp.dataproc.MetastoreDatabaseIamMember("member", {
 *     project: dpmsService.project,
 *     location: dpmsService.location,
 *     serviceId: dpmsService.serviceId,
 *     database: hive.hiveConfig[0].properties.database,
 *     role: "roles/viewer",
 *     member: "user:jane@example.com",
 * });
 * ```
 *
 * ## Import
 *
 * For all import syntaxes, the "resource in question" can take any of the following forms:
 *
 * * projects/{{project}}/locations/{{location}}/services/{{serviceId}}/databases/{{name}}
 *
 * * {{project}}/{{location}}/{{serviceId}}/{{name}}
 *
 * * {{location}}/{{serviceId}}/{{name}}
 *
 * * {{name}}
 *
 * Any variables not passed in the import command will be taken from the provider configuration.
 *
 * Dataproc Metastore database IAM resources can be imported using the resource identifiers, role, and member.
 *
 * IAM member imports use space-delimited identifiers: the resource in question, the role, and the member identity, e.g.
 *
 * ```sh
 * $ pulumi import gcp:dataproc/metastoreDatabaseIamMember:MetastoreDatabaseIamMember editor "projects/{{project}}/locations/{{location}}/services/{{serviceId}}/databases/{{database}} roles/viewer user:jane@example.com"
 * ```
 *
 * IAM binding imports use space-delimited identifiers: the resource in question and the role, e.g.
 *
 * ```sh
 * $ pulumi import gcp:dataproc/metastoreDatabaseIamMember:MetastoreDatabaseIamMember editor "projects/{{project}}/locations/{{location}}/services/{{serviceId}}/databases/{{database}} roles/viewer"
 * ```
 *
 * IAM policy imports use the identifier of the resource in question, e.g.
 *
 * ```sh
 * $ pulumi import gcp:dataproc/metastoreDatabaseIamMember:MetastoreDatabaseIamMember editor projects/{{project}}/locations/{{location}}/services/{{serviceId}}/databases/{{database}}
 * ```
 *
 * -> **Custom Roles** If you're importing a IAM resource with a custom role, make sure to use the
 *
 *  full name of the custom role, e.g. `[projects/my-project|organizations/my-org]/roles/my-custom-role`.
 */
class MetastoreDatabaseIamMember extends pulumi.CustomResource {
    /**
     * Get an existing MetastoreDatabaseIamMember resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new MetastoreDatabaseIamMember(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of MetastoreDatabaseIamMember.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === MetastoreDatabaseIamMember.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["condition"] = state ? state.condition : undefined;
            resourceInputs["database"] = state ? state.database : undefined;
            resourceInputs["etag"] = state ? state.etag : undefined;
            resourceInputs["location"] = state ? state.location : undefined;
            resourceInputs["member"] = state ? state.member : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["role"] = state ? state.role : undefined;
            resourceInputs["serviceId"] = state ? state.serviceId : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.database === undefined) && !opts.urn) {
                throw new Error("Missing required property 'database'");
            }
            if ((!args || args.member === undefined) && !opts.urn) {
                throw new Error("Missing required property 'member'");
            }
            if ((!args || args.role === undefined) && !opts.urn) {
                throw new Error("Missing required property 'role'");
            }
            if ((!args || args.serviceId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'serviceId'");
            }
            resourceInputs["condition"] = args ? args.condition : undefined;
            resourceInputs["database"] = args ? args.database : undefined;
            resourceInputs["location"] = args ? args.location : undefined;
            resourceInputs["member"] = args ? args.member : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["role"] = args ? args.role : undefined;
            resourceInputs["serviceId"] = args ? args.serviceId : undefined;
            resourceInputs["etag"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(MetastoreDatabaseIamMember.__pulumiType, name, resourceInputs, opts);
    }
}
exports.MetastoreDatabaseIamMember = MetastoreDatabaseIamMember;
/** @internal */
MetastoreDatabaseIamMember.__pulumiType = 'gcp:dataproc/metastoreDatabaseIamMember:MetastoreDatabaseIamMember';
