// @pulumi/gcp@8.41.1/modelarmor/template.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Template = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A `Template` is a resource of Model Armor that lets you configure how Model Armor screens prompts and responses.
 * It functions as sets of customized filters and thresholds for different safety and security confidence levels, allowing control over what content is flagged.
 *
 * ## Example Usage
 *
 * ### Modelarmor Template Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const template_basic = new gcp.modelarmor.Template("template-basic", {
 *     location: "<no value>",
 *     templateId: "<no value>",
 *     filterConfig: {},
 *     templateMetadata: {},
 * });
 * ```
 * ## Import
 *
 * Template can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/locations/{{location}}/templates/{{template_id}}`
 *
 * * `{{project}}/{{location}}/{{template_id}}`
 *
 * * `{{location}}/{{template_id}}`
 *
 * When using the `pulumi import` command, Template can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:modelarmor/template:Template default projects/{{project}}/locations/{{location}}/templates/{{template_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:modelarmor/template:Template default {{project}}/{{location}}/{{template_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:modelarmor/template:Template default {{location}}/{{template_id}}
 * ```
 */
class Template extends pulumi.CustomResource {
    /**
     * Get an existing Template resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Template(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Template.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Template.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["createTime"] = state ? state.createTime : undefined;
            resourceInputs["effectiveLabels"] = state ? state.effectiveLabels : undefined;
            resourceInputs["filterConfig"] = state ? state.filterConfig : undefined;
            resourceInputs["labels"] = state ? state.labels : undefined;
            resourceInputs["location"] = state ? state.location : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["pulumiLabels"] = state ? state.pulumiLabels : undefined;
            resourceInputs["templateId"] = state ? state.templateId : undefined;
            resourceInputs["templateMetadata"] = state ? state.templateMetadata : undefined;
            resourceInputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.filterConfig === undefined) && !opts.urn) {
                throw new Error("Missing required property 'filterConfig'");
            }
            if ((!args || args.location === undefined) && !opts.urn) {
                throw new Error("Missing required property 'location'");
            }
            if ((!args || args.templateId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'templateId'");
            }
            resourceInputs["filterConfig"] = args ? args.filterConfig : undefined;
            resourceInputs["labels"] = args ? args.labels : undefined;
            resourceInputs["location"] = args ? args.location : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["templateId"] = args ? args.templateId : undefined;
            resourceInputs["templateMetadata"] = args ? args.templateMetadata : undefined;
            resourceInputs["createTime"] = undefined /*out*/;
            resourceInputs["effectiveLabels"] = undefined /*out*/;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["pulumiLabels"] = undefined /*out*/;
            resourceInputs["updateTime"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const secretOpts = { additionalSecretOutputs: ["effectiveLabels", "pulumiLabels"] };
        opts = pulumi.mergeOptions(opts, secretOpts);
        super(Template.__pulumiType, name, resourceInputs, opts);
    }
}
exports.Template = Template;
/** @internal */
Template.__pulumiType = 'gcp:modelarmor/template:Template';
