// @pulumi/gcp@8.41.1/diagflow/conversationProfile.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConversationProfile = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A conversation profile configures a set of parameters that control the suggestions made to an agent. These parameters control the suggestions that are surfaced during runtime. Each profile configures either a Dialogflow virtual agent or a human agent for a conversation.
 *
 * To get more information about ConversationProfile, see:
 *
 * * [API documentation](https://cloud.google.com/dialogflow/docs/reference/rest/v2beta1/projects/conversationProfiles)
 * * How-to Guides
 *     * [Official Documentation](https://cloud.google.com/dialogflow/docs/)
 *
 * ## Example Usage
 *
 * ### Dialogflow Conversation Profile Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const basicAgent = new gcp.diagflow.Agent("basic_agent", {
 *     displayName: "example_agent",
 *     defaultLanguageCode: "en-us",
 *     timeZone: "America/New_York",
 * });
 * const basicProfile = new gcp.diagflow.ConversationProfile("basic_profile", {
 *     displayName: "dialogflow-profile",
 *     location: "global",
 *     automatedAgentConfig: {
 *         agent: pulumi.interpolate`projects/${basicAgent.id}/locations/global/agent/environments/draft`,
 *     },
 *     humanAgentAssistantConfig: {
 *         messageAnalysisConfig: {
 *             enableEntityExtraction: true,
 *             enableSentimentAnalysis: true,
 *         },
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * ConversationProfile can be imported using any of these accepted formats:
 *
 * * `{{name}}`
 *
 * When using the `pulumi import` command, ConversationProfile can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:diagflow/conversationProfile:ConversationProfile default {{name}}
 * ```
 */
class ConversationProfile extends pulumi.CustomResource {
    /**
     * Get an existing ConversationProfile resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ConversationProfile(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ConversationProfile.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ConversationProfile.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["automatedAgentConfig"] = state ? state.automatedAgentConfig : undefined;
            resourceInputs["displayName"] = state ? state.displayName : undefined;
            resourceInputs["humanAgentAssistantConfig"] = state ? state.humanAgentAssistantConfig : undefined;
            resourceInputs["humanAgentHandoffConfig"] = state ? state.humanAgentHandoffConfig : undefined;
            resourceInputs["languageCode"] = state ? state.languageCode : undefined;
            resourceInputs["location"] = state ? state.location : undefined;
            resourceInputs["loggingConfig"] = state ? state.loggingConfig : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["newMessageEventNotificationConfig"] = state ? state.newMessageEventNotificationConfig : undefined;
            resourceInputs["notificationConfig"] = state ? state.notificationConfig : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["securitySettings"] = state ? state.securitySettings : undefined;
            resourceInputs["sttConfig"] = state ? state.sttConfig : undefined;
            resourceInputs["timeZone"] = state ? state.timeZone : undefined;
            resourceInputs["ttsConfig"] = state ? state.ttsConfig : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.displayName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'displayName'");
            }
            if ((!args || args.location === undefined) && !opts.urn) {
                throw new Error("Missing required property 'location'");
            }
            resourceInputs["automatedAgentConfig"] = args ? args.automatedAgentConfig : undefined;
            resourceInputs["displayName"] = args ? args.displayName : undefined;
            resourceInputs["humanAgentAssistantConfig"] = args ? args.humanAgentAssistantConfig : undefined;
            resourceInputs["humanAgentHandoffConfig"] = args ? args.humanAgentHandoffConfig : undefined;
            resourceInputs["languageCode"] = args ? args.languageCode : undefined;
            resourceInputs["location"] = args ? args.location : undefined;
            resourceInputs["loggingConfig"] = args ? args.loggingConfig : undefined;
            resourceInputs["newMessageEventNotificationConfig"] = args ? args.newMessageEventNotificationConfig : undefined;
            resourceInputs["notificationConfig"] = args ? args.notificationConfig : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["securitySettings"] = args ? args.securitySettings : undefined;
            resourceInputs["sttConfig"] = args ? args.sttConfig : undefined;
            resourceInputs["timeZone"] = args ? args.timeZone : undefined;
            resourceInputs["ttsConfig"] = args ? args.ttsConfig : undefined;
            resourceInputs["name"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(ConversationProfile.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ConversationProfile = ConversationProfile;
/** @internal */
ConversationProfile.__pulumiType = 'gcp:diagflow/conversationProfile:ConversationProfile';
