// @pulumi/gcp@8.41.1/diagflow/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Intent = exports.Fulfillment = exports.EntityType = exports.EncryptionSpec = exports.CxWebhook = exports.CxVersion = exports.CxTool = exports.CxTestCase = exports.CxSecuritySettings = exports.CxPlaybook = exports.CxPage = exports.CxIntent = exports.CxGenerator = exports.CxGenerativeSettings = exports.CxFlow = exports.CxEnvironment = exports.CxEntityType = exports.CxAgent = exports.ConversationProfile = exports.Agent = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.Agent = null;
utilities.lazyLoad(exports, ["Agent"], () => require("./agent"));
exports.ConversationProfile = null;
utilities.lazyLoad(exports, ["ConversationProfile"], () => require("./conversationProfile"));
exports.CxAgent = null;
utilities.lazyLoad(exports, ["CxAgent"], () => require("./cxAgent"));
exports.CxEntityType = null;
utilities.lazyLoad(exports, ["CxEntityType"], () => require("./cxEntityType"));
exports.CxEnvironment = null;
utilities.lazyLoad(exports, ["CxEnvironment"], () => require("./cxEnvironment"));
exports.CxFlow = null;
utilities.lazyLoad(exports, ["CxFlow"], () => require("./cxFlow"));
exports.CxGenerativeSettings = null;
utilities.lazyLoad(exports, ["CxGenerativeSettings"], () => require("./cxGenerativeSettings"));
exports.CxGenerator = null;
utilities.lazyLoad(exports, ["CxGenerator"], () => require("./cxGenerator"));
exports.CxIntent = null;
utilities.lazyLoad(exports, ["CxIntent"], () => require("./cxIntent"));
exports.CxPage = null;
utilities.lazyLoad(exports, ["CxPage"], () => require("./cxPage"));
exports.CxPlaybook = null;
utilities.lazyLoad(exports, ["CxPlaybook"], () => require("./cxPlaybook"));
exports.CxSecuritySettings = null;
utilities.lazyLoad(exports, ["CxSecuritySettings"], () => require("./cxSecuritySettings"));
exports.CxTestCase = null;
utilities.lazyLoad(exports, ["CxTestCase"], () => require("./cxTestCase"));
exports.CxTool = null;
utilities.lazyLoad(exports, ["CxTool"], () => require("./cxTool"));
exports.CxVersion = null;
utilities.lazyLoad(exports, ["CxVersion"], () => require("./cxVersion"));
exports.CxWebhook = null;
utilities.lazyLoad(exports, ["CxWebhook"], () => require("./cxWebhook"));
exports.EncryptionSpec = null;
utilities.lazyLoad(exports, ["EncryptionSpec"], () => require("./encryptionSpec"));
exports.EntityType = null;
utilities.lazyLoad(exports, ["EntityType"], () => require("./entityType"));
exports.Fulfillment = null;
utilities.lazyLoad(exports, ["Fulfillment"], () => require("./fulfillment"));
exports.Intent = null;
utilities.lazyLoad(exports, ["Intent"], () => require("./intent"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "gcp:diagflow/agent:Agent":
                return new exports.Agent(name, undefined, { urn });
            case "gcp:diagflow/conversationProfile:ConversationProfile":
                return new exports.ConversationProfile(name, undefined, { urn });
            case "gcp:diagflow/cxAgent:CxAgent":
                return new exports.CxAgent(name, undefined, { urn });
            case "gcp:diagflow/cxEntityType:CxEntityType":
                return new exports.CxEntityType(name, undefined, { urn });
            case "gcp:diagflow/cxEnvironment:CxEnvironment":
                return new exports.CxEnvironment(name, undefined, { urn });
            case "gcp:diagflow/cxFlow:CxFlow":
                return new exports.CxFlow(name, undefined, { urn });
            case "gcp:diagflow/cxGenerativeSettings:CxGenerativeSettings":
                return new exports.CxGenerativeSettings(name, undefined, { urn });
            case "gcp:diagflow/cxGenerator:CxGenerator":
                return new exports.CxGenerator(name, undefined, { urn });
            case "gcp:diagflow/cxIntent:CxIntent":
                return new exports.CxIntent(name, undefined, { urn });
            case "gcp:diagflow/cxPage:CxPage":
                return new exports.CxPage(name, undefined, { urn });
            case "gcp:diagflow/cxPlaybook:CxPlaybook":
                return new exports.CxPlaybook(name, undefined, { urn });
            case "gcp:diagflow/cxSecuritySettings:CxSecuritySettings":
                return new exports.CxSecuritySettings(name, undefined, { urn });
            case "gcp:diagflow/cxTestCase:CxTestCase":
                return new exports.CxTestCase(name, undefined, { urn });
            case "gcp:diagflow/cxTool:CxTool":
                return new exports.CxTool(name, undefined, { urn });
            case "gcp:diagflow/cxVersion:CxVersion":
                return new exports.CxVersion(name, undefined, { urn });
            case "gcp:diagflow/cxWebhook:CxWebhook":
                return new exports.CxWebhook(name, undefined, { urn });
            case "gcp:diagflow/encryptionSpec:EncryptionSpec":
                return new exports.EncryptionSpec(name, undefined, { urn });
            case "gcp:diagflow/entityType:EntityType":
                return new exports.EntityType(name, undefined, { urn });
            case "gcp:diagflow/fulfillment:Fulfillment":
                return new exports.Fulfillment(name, undefined, { urn });
            case "gcp:diagflow/intent:Intent":
                return new exports.Intent(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("gcp", "diagflow/agent", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/conversationProfile", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/cxAgent", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/cxEntityType", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/cxEnvironment", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/cxFlow", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/cxGenerativeSettings", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/cxGenerator", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/cxIntent", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/cxPage", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/cxPlaybook", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/cxSecuritySettings", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/cxTestCase", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/cxTool", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/cxVersion", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/cxWebhook", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/encryptionSpec", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/entityType", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/fulfillment", _module);
pulumi.runtime.registerResourceModule("gcp", "diagflow/intent", _module);
