// @pulumi/gcp@8.41.1/apihub/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginInstance = exports.Plugin = exports.HostProjectRegistration = exports.Curation = exports.ApiHubInstance = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.ApiHubInstance = null;
utilities.lazyLoad(exports, ["ApiHubInstance"], () => require("./apiHubInstance"));
exports.Curation = null;
utilities.lazyLoad(exports, ["Curation"], () => require("./curation"));
exports.HostProjectRegistration = null;
utilities.lazyLoad(exports, ["HostProjectRegistration"], () => require("./hostProjectRegistration"));
exports.Plugin = null;
utilities.lazyLoad(exports, ["Plugin"], () => require("./plugin"));
exports.PluginInstance = null;
utilities.lazyLoad(exports, ["PluginInstance"], () => require("./pluginInstance"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "gcp:apihub/apiHubInstance:ApiHubInstance":
                return new exports.ApiHubInstance(name, undefined, { urn });
            case "gcp:apihub/curation:Curation":
                return new exports.Curation(name, undefined, { urn });
            case "gcp:apihub/hostProjectRegistration:HostProjectRegistration":
                return new exports.HostProjectRegistration(name, undefined, { urn });
            case "gcp:apihub/plugin:Plugin":
                return new exports.Plugin(name, undefined, { urn });
            case "gcp:apihub/pluginInstance:PluginInstance":
                return new exports.PluginInstance(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("gcp", "apihub/apiHubInstance", _module);
pulumi.runtime.registerResourceModule("gcp", "apihub/curation", _module);
pulumi.runtime.registerResourceModule("gcp", "apihub/hostProjectRegistration", _module);
pulumi.runtime.registerResourceModule("gcp", "apihub/plugin", _module);
pulumi.runtime.registerResourceModule("gcp", "apihub/pluginInstance", _module);
