// @pulumi/gcp@8.41.1/apihub/pluginInstance.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginInstance = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Description
 *
 * ## Example Usage
 *
 * ### Apihub Plugin Instance Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const apihubPluginInstanceBasic = new gcp.apihub.PluginInstance("apihub_plugin_instance_basic", {
 *     location: "us-central1",
 *     plugin: "existing-plugin-id",
 *     pluginInstanceId: "test",
 *     displayName: "Sample Plugin Instance Display Name",
 *     disable: false,
 *     actions: [{
 *         actionId: "existing-action-id",
 *     }],
 * });
 * ```
 *
 * ## Import
 *
 * PluginInstance can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/locations/{{location}}/plugins/{{plugin}}/instances/{{plugin_instance_id}}`
 *
 * * `{{project}}/{{location}}/{{plugin}}/{{plugin_instance_id}}`
 *
 * * `{{location}}/{{plugin}}/{{plugin_instance_id}}`
 *
 * When using the `pulumi import` command, PluginInstance can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:apihub/pluginInstance:PluginInstance default projects/{{project}}/locations/{{location}}/plugins/{{plugin}}/instances/{{plugin_instance_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:apihub/pluginInstance:PluginInstance default {{project}}/{{location}}/{{plugin}}/{{plugin_instance_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:apihub/pluginInstance:PluginInstance default {{location}}/{{plugin}}/{{plugin_instance_id}}
 * ```
 */
class PluginInstance extends pulumi.CustomResource {
    /**
     * Get an existing PluginInstance resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new PluginInstance(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of PluginInstance.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === PluginInstance.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["actions"] = state ? state.actions : undefined;
            resourceInputs["authConfig"] = state ? state.authConfig : undefined;
            resourceInputs["createTime"] = state ? state.createTime : undefined;
            resourceInputs["disable"] = state ? state.disable : undefined;
            resourceInputs["displayName"] = state ? state.displayName : undefined;
            resourceInputs["errorMessage"] = state ? state.errorMessage : undefined;
            resourceInputs["location"] = state ? state.location : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["plugin"] = state ? state.plugin : undefined;
            resourceInputs["pluginInstanceId"] = state ? state.pluginInstanceId : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["state"] = state ? state.state : undefined;
            resourceInputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.displayName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'displayName'");
            }
            if ((!args || args.location === undefined) && !opts.urn) {
                throw new Error("Missing required property 'location'");
            }
            if ((!args || args.plugin === undefined) && !opts.urn) {
                throw new Error("Missing required property 'plugin'");
            }
            if ((!args || args.pluginInstanceId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'pluginInstanceId'");
            }
            resourceInputs["actions"] = args ? args.actions : undefined;
            resourceInputs["authConfig"] = args ? args.authConfig : undefined;
            resourceInputs["disable"] = args ? args.disable : undefined;
            resourceInputs["displayName"] = args ? args.displayName : undefined;
            resourceInputs["location"] = args ? args.location : undefined;
            resourceInputs["plugin"] = args ? args.plugin : undefined;
            resourceInputs["pluginInstanceId"] = args ? args.pluginInstanceId : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["createTime"] = undefined /*out*/;
            resourceInputs["errorMessage"] = undefined /*out*/;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["state"] = undefined /*out*/;
            resourceInputs["updateTime"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(PluginInstance.__pulumiType, name, resourceInputs, opts);
    }
}
exports.PluginInstance = PluginInstance;
/** @internal */
PluginInstance.__pulumiType = 'gcp:apihub/pluginInstance:PluginInstance';
