// @pulumi/gcp@8.41.1/oracledatabase/odbSubnet.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.OdbSubnet = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * An OdbSubnet resource which represents a subnet under an OdbNetwork.
 *
 * To get more information about OdbSubnet, see:
 * * How-to Guides
 *     * [OracleDatabase@Google Cloud](https://cloud.google.com/oracle/database/docs/overview')
 *
 * ## Example Usage
 *
 * ### Oracledatabase Odbsubnet
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const my_odbsubnet = new gcp.oracledatabase.OdbSubnet("my-odbsubnet", {
 *     odbSubnetId: "my-odbsubnet",
 *     location: "europe-west2",
 *     project: "my-project",
 *     odbnetwork: "my-odbnetwork",
 *     cidrRange: "10.1.1.0/24",
 *     purpose: "CLIENT_SUBNET",
 *     labels: {
 *         terraform_created: "true",
 *     },
 *     deletionProtection: true,
 * });
 * ```
 *
 * ## Import
 *
 * OdbSubnet can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/locations/{{location}}/odbNetworks/{{odbnetwork}}/odbSubnets/{{odb_subnet_id}}`
 *
 * * `{{project}}/{{location}}/{{odbnetwork}}/{{odb_subnet_id}}`
 *
 * * `{{location}}/{{odbnetwork}}/{{odb_subnet_id}}`
 *
 * When using the `pulumi import` command, OdbSubnet can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:oracledatabase/odbSubnet:OdbSubnet default projects/{{project}}/locations/{{location}}/odbNetworks/{{odbnetwork}}/odbSubnets/{{odb_subnet_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:oracledatabase/odbSubnet:OdbSubnet default {{project}}/{{location}}/{{odbnetwork}}/{{odb_subnet_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:oracledatabase/odbSubnet:OdbSubnet default {{location}}/{{odbnetwork}}/{{odb_subnet_id}}
 * ```
 */
class OdbSubnet extends pulumi.CustomResource {
    /**
     * Get an existing OdbSubnet resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new OdbSubnet(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of OdbSubnet.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === OdbSubnet.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["cidrRange"] = state ? state.cidrRange : undefined;
            resourceInputs["createTime"] = state ? state.createTime : undefined;
            resourceInputs["deletionProtection"] = state ? state.deletionProtection : undefined;
            resourceInputs["effectiveLabels"] = state ? state.effectiveLabels : undefined;
            resourceInputs["labels"] = state ? state.labels : undefined;
            resourceInputs["location"] = state ? state.location : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["odbSubnetId"] = state ? state.odbSubnetId : undefined;
            resourceInputs["odbnetwork"] = state ? state.odbnetwork : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["pulumiLabels"] = state ? state.pulumiLabels : undefined;
            resourceInputs["purpose"] = state ? state.purpose : undefined;
            resourceInputs["state"] = state ? state.state : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.cidrRange === undefined) && !opts.urn) {
                throw new Error("Missing required property 'cidrRange'");
            }
            if ((!args || args.location === undefined) && !opts.urn) {
                throw new Error("Missing required property 'location'");
            }
            if ((!args || args.odbSubnetId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'odbSubnetId'");
            }
            if ((!args || args.odbnetwork === undefined) && !opts.urn) {
                throw new Error("Missing required property 'odbnetwork'");
            }
            if ((!args || args.purpose === undefined) && !opts.urn) {
                throw new Error("Missing required property 'purpose'");
            }
            resourceInputs["cidrRange"] = args ? args.cidrRange : undefined;
            resourceInputs["deletionProtection"] = args ? args.deletionProtection : undefined;
            resourceInputs["labels"] = args ? args.labels : undefined;
            resourceInputs["location"] = args ? args.location : undefined;
            resourceInputs["odbSubnetId"] = args ? args.odbSubnetId : undefined;
            resourceInputs["odbnetwork"] = args ? args.odbnetwork : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["purpose"] = args ? args.purpose : undefined;
            resourceInputs["createTime"] = undefined /*out*/;
            resourceInputs["effectiveLabels"] = undefined /*out*/;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["pulumiLabels"] = undefined /*out*/;
            resourceInputs["state"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const secretOpts = { additionalSecretOutputs: ["effectiveLabels", "pulumiLabels"] };
        opts = pulumi.mergeOptions(opts, secretOpts);
        super(OdbSubnet.__pulumiType, name, resourceInputs, opts);
    }
}
exports.OdbSubnet = OdbSubnet;
/** @internal */
OdbSubnet.__pulumiType = 'gcp:oracledatabase/odbSubnet:OdbSubnet';
