// @pulumi/gcp@8.41.1/beyondcorp/applicationIamPolicy.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationIamPolicy = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * > **Warning:** `gcp.beyondcorp.Application` and associated IAM resources are deprecated. Use `gcp.beyondcorp.SecurityGatewayApplication` instead.
 *
 * Three different resources help you manage your IAM policy for BeyondCorp Application. Each of these resources serves a different use case:
 *
 * * `gcp.beyondcorp.ApplicationIamPolicy`: Authoritative. Sets the IAM policy for the application and replaces any existing policy already attached.
 * * `gcp.beyondcorp.ApplicationIamBinding`: Authoritative for a given role. Updates the IAM policy to grant a role to a list of members. Other roles within the IAM policy for the application are preserved.
 * * `gcp.beyondcorp.ApplicationIamMember`: Non-authoritative. Updates the IAM policy to grant a role to a new member. Other members for the role for the application are preserved.
 *
 * A data source can be used to retrieve policy data in advent you do not need creation
 *
 * * `gcp.beyondcorp.ApplicationIamPolicy`: Retrieves the IAM policy for the application
 *
 * > **Note:** `gcp.beyondcorp.ApplicationIamPolicy` **cannot** be used in conjunction with `gcp.beyondcorp.ApplicationIamBinding` and `gcp.beyondcorp.ApplicationIamMember` or they will fight over what your policy should be.
 *
 * > **Note:** `gcp.beyondcorp.ApplicationIamBinding` resources **can be** used in conjunction with `gcp.beyondcorp.ApplicationIamMember` resources **only if** they do not grant privilege to the same role.
 *
 * > **Note:**  This resource supports IAM Conditions but they have some known limitations which can be found [here](https://cloud.google.com/iam/docs/conditions-overview#limitations). Please review this article if you are having issues with IAM Conditions.
 *
 * ## gcp.beyondcorp.ApplicationIamPolicy
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const admin = gcp.organizations.getIAMPolicy({
 *     bindings: [{
 *         role: "roles/beyondcorp.securityGatewayUser",
 *         members: ["user:jane@example.com"],
 *     }],
 * });
 * const policy = new gcp.beyondcorp.ApplicationIamPolicy("policy", {
 *     project: example.project,
 *     securityGatewaysId: example.securityGatewaysId,
 *     applicationId: example.applicationId,
 *     policyData: admin.then(admin => admin.policyData),
 * });
 * ```
 *
 * With IAM Conditions:
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const admin = gcp.organizations.getIAMPolicy({
 *     bindings: [{
 *         role: "roles/beyondcorp.securityGatewayUser",
 *         members: ["user:jane@example.com"],
 *         condition: {
 *             title: "expires_after_2019_12_31",
 *             description: "Expiring at midnight of 2019-12-31",
 *             expression: "request.time < timestamp(\"2020-01-01T00:00:00Z\")",
 *         },
 *     }],
 * });
 * const policy = new gcp.beyondcorp.ApplicationIamPolicy("policy", {
 *     project: example.project,
 *     securityGatewaysId: example.securityGatewaysId,
 *     applicationId: example.applicationId,
 *     policyData: admin.then(admin => admin.policyData),
 * });
 * ```
 * ## gcp.beyondcorp.ApplicationIamBinding
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const binding = new gcp.beyondcorp.ApplicationIamBinding("binding", {
 *     project: example.project,
 *     securityGatewaysId: example.securityGatewaysId,
 *     applicationId: example.applicationId,
 *     role: "roles/beyondcorp.securityGatewayUser",
 *     members: ["user:jane@example.com"],
 * });
 * ```
 *
 * With IAM Conditions:
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const binding = new gcp.beyondcorp.ApplicationIamBinding("binding", {
 *     project: example.project,
 *     securityGatewaysId: example.securityGatewaysId,
 *     applicationId: example.applicationId,
 *     role: "roles/beyondcorp.securityGatewayUser",
 *     members: ["user:jane@example.com"],
 *     condition: {
 *         title: "expires_after_2019_12_31",
 *         description: "Expiring at midnight of 2019-12-31",
 *         expression: "request.time < timestamp(\"2020-01-01T00:00:00Z\")",
 *     },
 * });
 * ```
 * ## gcp.beyondcorp.ApplicationIamMember
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const member = new gcp.beyondcorp.ApplicationIamMember("member", {
 *     project: example.project,
 *     securityGatewaysId: example.securityGatewaysId,
 *     applicationId: example.applicationId,
 *     role: "roles/beyondcorp.securityGatewayUser",
 *     member: "user:jane@example.com",
 * });
 * ```
 *
 * With IAM Conditions:
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const member = new gcp.beyondcorp.ApplicationIamMember("member", {
 *     project: example.project,
 *     securityGatewaysId: example.securityGatewaysId,
 *     applicationId: example.applicationId,
 *     role: "roles/beyondcorp.securityGatewayUser",
 *     member: "user:jane@example.com",
 *     condition: {
 *         title: "expires_after_2019_12_31",
 *         description: "Expiring at midnight of 2019-12-31",
 *         expression: "request.time < timestamp(\"2020-01-01T00:00:00Z\")",
 *     },
 * });
 * ```
 *
 * ## This resource supports User Project Overrides.
 *
 * -
 *
 * # IAM policy for BeyondCorp Application
 * > **Warning:** `gcp.beyondcorp.Application` and associated IAM resources are deprecated. Use `gcp.beyondcorp.SecurityGatewayApplication` instead.
 *
 * Three different resources help you manage your IAM policy for BeyondCorp Application. Each of these resources serves a different use case:
 *
 * * `gcp.beyondcorp.ApplicationIamPolicy`: Authoritative. Sets the IAM policy for the application and replaces any existing policy already attached.
 * * `gcp.beyondcorp.ApplicationIamBinding`: Authoritative for a given role. Updates the IAM policy to grant a role to a list of members. Other roles within the IAM policy for the application are preserved.
 * * `gcp.beyondcorp.ApplicationIamMember`: Non-authoritative. Updates the IAM policy to grant a role to a new member. Other members for the role for the application are preserved.
 *
 * A data source can be used to retrieve policy data in advent you do not need creation
 *
 * * `gcp.beyondcorp.ApplicationIamPolicy`: Retrieves the IAM policy for the application
 *
 * > **Note:** `gcp.beyondcorp.ApplicationIamPolicy` **cannot** be used in conjunction with `gcp.beyondcorp.ApplicationIamBinding` and `gcp.beyondcorp.ApplicationIamMember` or they will fight over what your policy should be.
 *
 * > **Note:** `gcp.beyondcorp.ApplicationIamBinding` resources **can be** used in conjunction with `gcp.beyondcorp.ApplicationIamMember` resources **only if** they do not grant privilege to the same role.
 *
 * > **Note:**  This resource supports IAM Conditions but they have some known limitations which can be found [here](https://cloud.google.com/iam/docs/conditions-overview#limitations). Please review this article if you are having issues with IAM Conditions.
 *
 * ## gcp.beyondcorp.ApplicationIamPolicy
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const admin = gcp.organizations.getIAMPolicy({
 *     bindings: [{
 *         role: "roles/beyondcorp.securityGatewayUser",
 *         members: ["user:jane@example.com"],
 *     }],
 * });
 * const policy = new gcp.beyondcorp.ApplicationIamPolicy("policy", {
 *     project: example.project,
 *     securityGatewaysId: example.securityGatewaysId,
 *     applicationId: example.applicationId,
 *     policyData: admin.then(admin => admin.policyData),
 * });
 * ```
 *
 * With IAM Conditions:
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const admin = gcp.organizations.getIAMPolicy({
 *     bindings: [{
 *         role: "roles/beyondcorp.securityGatewayUser",
 *         members: ["user:jane@example.com"],
 *         condition: {
 *             title: "expires_after_2019_12_31",
 *             description: "Expiring at midnight of 2019-12-31",
 *             expression: "request.time < timestamp(\"2020-01-01T00:00:00Z\")",
 *         },
 *     }],
 * });
 * const policy = new gcp.beyondcorp.ApplicationIamPolicy("policy", {
 *     project: example.project,
 *     securityGatewaysId: example.securityGatewaysId,
 *     applicationId: example.applicationId,
 *     policyData: admin.then(admin => admin.policyData),
 * });
 * ```
 * ## gcp.beyondcorp.ApplicationIamBinding
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const binding = new gcp.beyondcorp.ApplicationIamBinding("binding", {
 *     project: example.project,
 *     securityGatewaysId: example.securityGatewaysId,
 *     applicationId: example.applicationId,
 *     role: "roles/beyondcorp.securityGatewayUser",
 *     members: ["user:jane@example.com"],
 * });
 * ```
 *
 * With IAM Conditions:
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const binding = new gcp.beyondcorp.ApplicationIamBinding("binding", {
 *     project: example.project,
 *     securityGatewaysId: example.securityGatewaysId,
 *     applicationId: example.applicationId,
 *     role: "roles/beyondcorp.securityGatewayUser",
 *     members: ["user:jane@example.com"],
 *     condition: {
 *         title: "expires_after_2019_12_31",
 *         description: "Expiring at midnight of 2019-12-31",
 *         expression: "request.time < timestamp(\"2020-01-01T00:00:00Z\")",
 *     },
 * });
 * ```
 * ## gcp.beyondcorp.ApplicationIamMember
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const member = new gcp.beyondcorp.ApplicationIamMember("member", {
 *     project: example.project,
 *     securityGatewaysId: example.securityGatewaysId,
 *     applicationId: example.applicationId,
 *     role: "roles/beyondcorp.securityGatewayUser",
 *     member: "user:jane@example.com",
 * });
 * ```
 *
 * With IAM Conditions:
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const member = new gcp.beyondcorp.ApplicationIamMember("member", {
 *     project: example.project,
 *     securityGatewaysId: example.securityGatewaysId,
 *     applicationId: example.applicationId,
 *     role: "roles/beyondcorp.securityGatewayUser",
 *     member: "user:jane@example.com",
 *     condition: {
 *         title: "expires_after_2019_12_31",
 *         description: "Expiring at midnight of 2019-12-31",
 *         expression: "request.time < timestamp(\"2020-01-01T00:00:00Z\")",
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * For all import syntaxes, the "resource in question" can take any of the following forms:
 *
 * * projects/{{project}}/locations/global/securityGateways/{{security_gateways_id}}/applications/{{application_id}}
 *
 * * {{project}}/{{security_gateways_id}}/{{application_id}}
 *
 * * {{security_gateways_id}}/{{application_id}}
 *
 * * {{application_id}}
 *
 * Any variables not passed in the import command will be taken from the provider configuration.
 *
 * BeyondCorp application IAM resources can be imported using the resource identifiers, role, and member.
 *
 * IAM member imports use space-delimited identifiers: the resource in question, the role, and the member identity, e.g.
 *
 * ```sh
 * $ pulumi import gcp:beyondcorp/applicationIamPolicy:ApplicationIamPolicy editor "projects/{{project}}/locations/global/securityGateways/{{security_gateways_id}}/applications/{{application_id}} roles/beyondcorp.securityGatewayUser user:jane@example.com"
 * ```
 *
 * IAM binding imports use space-delimited identifiers: the resource in question and the role, e.g.
 *
 * ```sh
 * $ pulumi import gcp:beyondcorp/applicationIamPolicy:ApplicationIamPolicy editor "projects/{{project}}/locations/global/securityGateways/{{security_gateways_id}}/applications/{{application_id}} roles/beyondcorp.securityGatewayUser"
 * ```
 *
 * IAM policy imports use the identifier of the resource in question, e.g.
 *
 * ```sh
 * $ pulumi import gcp:beyondcorp/applicationIamPolicy:ApplicationIamPolicy editor projects/{{project}}/locations/global/securityGateways/{{security_gateways_id}}/applications/{{application_id}}
 * ```
 *
 * -> **Custom Roles** If you're importing a IAM resource with a custom role, make sure to use the
 *
 *  full name of the custom role, e.g. `[projects/my-project|organizations/my-org]/roles/my-custom-role`.
 */
class ApplicationIamPolicy extends pulumi.CustomResource {
    /**
     * Get an existing ApplicationIamPolicy resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ApplicationIamPolicy(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ApplicationIamPolicy.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ApplicationIamPolicy.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["applicationId"] = state ? state.applicationId : undefined;
            resourceInputs["etag"] = state ? state.etag : undefined;
            resourceInputs["policyData"] = state ? state.policyData : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["securityGatewaysId"] = state ? state.securityGatewaysId : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.applicationId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'applicationId'");
            }
            if ((!args || args.policyData === undefined) && !opts.urn) {
                throw new Error("Missing required property 'policyData'");
            }
            if ((!args || args.securityGatewaysId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'securityGatewaysId'");
            }
            resourceInputs["applicationId"] = args ? args.applicationId : undefined;
            resourceInputs["policyData"] = args ? args.policyData : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["securityGatewaysId"] = args ? args.securityGatewaysId : undefined;
            resourceInputs["etag"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(ApplicationIamPolicy.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ApplicationIamPolicy = ApplicationIamPolicy;
/** @internal */
ApplicationIamPolicy.__pulumiType = 'gcp:beyondcorp/applicationIamPolicy:ApplicationIamPolicy';
