// @pulumi/gcp@8.41.1/memorystore/instance.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Instance = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A Google Cloud Memorystore instance.
 *
 * To get more information about Instance, see:
 *
 * * [API documentation](https://cloud.google.com/memorystore/docs/valkey/reference/rest/v1/projects.locations.instances)
 * * How-to Guides
 *     * [Official Documentation](https://cloud.google.com/memorystore/docs/valkey/create-instances)
 *
 * ## Example Usage
 *
 * ### Memorystore Instance Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const producerNet = new gcp.compute.Network("producer_net", {
 *     name: "my-network",
 *     autoCreateSubnetworks: false,
 * });
 * const producerSubnet = new gcp.compute.Subnetwork("producer_subnet", {
 *     name: "my-subnet",
 *     ipCidrRange: "10.0.0.248/29",
 *     region: "us-central1",
 *     network: producerNet.id,
 * });
 * const _default = new gcp.networkconnectivity.ServiceConnectionPolicy("default", {
 *     name: "my-policy",
 *     location: "us-central1",
 *     serviceClass: "gcp-memorystore",
 *     description: "my basic service connection policy",
 *     network: producerNet.id,
 *     pscConfig: {
 *         subnetworks: [producerSubnet.id],
 *     },
 * });
 * const project = gcp.organizations.getProject({});
 * const instance_basic = new gcp.memorystore.Instance("instance-basic", {
 *     instanceId: "basic-instance",
 *     shardCount: 1,
 *     desiredAutoCreatedEndpoints: [{
 *         network: producerNet.id,
 *         projectId: project.then(project => project.projectId),
 *     }],
 *     location: "us-central1",
 *     deletionProtectionEnabled: false,
 *     maintenancePolicy: {
 *         weeklyMaintenanceWindows: [{
 *             day: "MONDAY",
 *             startTime: {
 *                 hours: 1,
 *                 minutes: 0,
 *                 seconds: 0,
 *                 nanos: 0,
 *             },
 *         }],
 *     },
 * }, {
 *     dependsOn: [_default],
 * });
 * ```
 * ### Memorystore Instance Full
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const producerNet = new gcp.compute.Network("producer_net", {
 *     name: "my-network",
 *     autoCreateSubnetworks: false,
 * });
 * const producerSubnet = new gcp.compute.Subnetwork("producer_subnet", {
 *     name: "my-subnet",
 *     ipCidrRange: "10.0.0.248/29",
 *     region: "us-central1",
 *     network: producerNet.id,
 * });
 * const _default = new gcp.networkconnectivity.ServiceConnectionPolicy("default", {
 *     name: "my-policy",
 *     location: "us-central1",
 *     serviceClass: "gcp-memorystore",
 *     description: "my basic service connection policy",
 *     network: producerNet.id,
 *     pscConfig: {
 *         subnetworks: [producerSubnet.id],
 *     },
 * });
 * const project = gcp.organizations.getProject({});
 * const instance_full = new gcp.memorystore.Instance("instance-full", {
 *     instanceId: "full-instance",
 *     shardCount: 1,
 *     desiredAutoCreatedEndpoints: [{
 *         network: producerNet.id,
 *         projectId: project.then(project => project.projectId),
 *     }],
 *     location: "us-central1",
 *     replicaCount: 1,
 *     nodeType: "SHARED_CORE_NANO",
 *     transitEncryptionMode: "TRANSIT_ENCRYPTION_DISABLED",
 *     authorizationMode: "AUTH_DISABLED",
 *     kmsKey: "my-key",
 *     engineConfigs: {
 *         "maxmemory-policy": "volatile-ttl",
 *     },
 *     allowFewerZonesDeployment: true,
 *     zoneDistributionConfig: {
 *         mode: "SINGLE_ZONE",
 *         zone: "us-central1-b",
 *     },
 *     maintenancePolicy: {
 *         weeklyMaintenanceWindows: [{
 *             day: "MONDAY",
 *             startTime: {
 *                 hours: 1,
 *                 minutes: 0,
 *                 seconds: 0,
 *                 nanos: 0,
 *             },
 *         }],
 *     },
 *     engineVersion: "VALKEY_7_2",
 *     deletionProtectionEnabled: false,
 *     mode: "CLUSTER",
 *     persistenceConfig: {
 *         mode: "RDB",
 *         rdbConfig: {
 *             rdbSnapshotPeriod: "ONE_HOUR",
 *             rdbSnapshotStartTime: "2024-10-02T15:01:23Z",
 *         },
 *     },
 *     labels: {
 *         abc: "xyz",
 *     },
 * }, {
 *     dependsOn: [_default],
 * });
 * ```
 * ### Memorystore Instance Persistence Aof
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const producerNet = new gcp.compute.Network("producer_net", {
 *     name: "my-network",
 *     autoCreateSubnetworks: false,
 * });
 * const producerSubnet = new gcp.compute.Subnetwork("producer_subnet", {
 *     name: "my-subnet",
 *     ipCidrRange: "10.0.0.248/29",
 *     region: "us-central1",
 *     network: producerNet.id,
 * });
 * const _default = new gcp.networkconnectivity.ServiceConnectionPolicy("default", {
 *     name: "my-policy",
 *     location: "us-central1",
 *     serviceClass: "gcp-memorystore",
 *     description: "my basic service connection policy",
 *     network: producerNet.id,
 *     pscConfig: {
 *         subnetworks: [producerSubnet.id],
 *     },
 * });
 * const project = gcp.organizations.getProject({});
 * const instance_persistence_aof = new gcp.memorystore.Instance("instance-persistence-aof", {
 *     instanceId: "aof-instance",
 *     shardCount: 1,
 *     desiredAutoCreatedEndpoints: [{
 *         network: producerNet.id,
 *         projectId: project.then(project => project.projectId),
 *     }],
 *     location: "us-central1",
 *     persistenceConfig: {
 *         mode: "AOF",
 *         aofConfig: {
 *             appendFsync: "EVERY_SEC",
 *         },
 *     },
 *     deletionProtectionEnabled: false,
 * }, {
 *     dependsOn: [_default],
 * });
 * ```
 * ### Memorystore Instance Secondary Instance
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const primaryProducerNet = new gcp.compute.Network("primary_producer_net", {
 *     name: "my-network-primary-instance",
 *     autoCreateSubnetworks: false,
 * });
 * const primaryProducerSubnet = new gcp.compute.Subnetwork("primary_producer_subnet", {
 *     name: "my-subnet-primary-instance",
 *     ipCidrRange: "10.0.1.0/29",
 *     region: "asia-east1",
 *     network: primaryProducerNet.id,
 * });
 * const primaryPolicy = new gcp.networkconnectivity.ServiceConnectionPolicy("primary_policy", {
 *     name: "my-policy-primary-instance",
 *     location: "asia-east1",
 *     serviceClass: "gcp-memorystore",
 *     description: "my basic service connection policy",
 *     network: primaryProducerNet.id,
 *     pscConfig: {
 *         subnetworks: [primaryProducerSubnet.id],
 *     },
 * });
 * const project = gcp.organizations.getProject({});
 * // Primary instance
 * const primaryInstance = new gcp.memorystore.Instance("primary_instance", {
 *     instanceId: "primary-instance",
 *     shardCount: 1,
 *     desiredAutoCreatedEndpoints: [{
 *         network: primaryProducerNet.id,
 *         projectId: project.then(project => project.projectId),
 *     }],
 *     location: "asia-east1",
 *     replicaCount: 1,
 *     nodeType: "SHARED_CORE_NANO",
 *     transitEncryptionMode: "TRANSIT_ENCRYPTION_DISABLED",
 *     authorizationMode: "AUTH_DISABLED",
 *     engineConfigs: {
 *         "maxmemory-policy": "volatile-ttl",
 *     },
 *     zoneDistributionConfig: {
 *         mode: "SINGLE_ZONE",
 *         zone: "asia-east1-c",
 *     },
 *     deletionProtectionEnabled: true,
 *     persistenceConfig: {
 *         mode: "RDB",
 *         rdbConfig: {
 *             rdbSnapshotPeriod: "ONE_HOUR",
 *             rdbSnapshotStartTime: "2024-10-02T15:01:23Z",
 *         },
 *     },
 *     labels: {
 *         abc: "xyz",
 *     },
 * }, {
 *     dependsOn: [primaryPolicy],
 * });
 * const secondaryProducerNet = new gcp.compute.Network("secondary_producer_net", {
 *     name: "my-network-secondary-instance",
 *     autoCreateSubnetworks: false,
 * });
 * const secondaryProducerSubnet = new gcp.compute.Subnetwork("secondary_producer_subnet", {
 *     name: "my-subnet-secondary-instance",
 *     ipCidrRange: "10.0.2.0/29",
 *     region: "europe-north1",
 *     network: secondaryProducerNet.id,
 * });
 * const secondaryPolicy = new gcp.networkconnectivity.ServiceConnectionPolicy("secondary_policy", {
 *     name: "my-policy-secondary-instance",
 *     location: "europe-north1",
 *     serviceClass: "gcp-memorystore",
 *     description: "my basic service connection policy",
 *     network: secondaryProducerNet.id,
 *     pscConfig: {
 *         subnetworks: [secondaryProducerSubnet.id],
 *     },
 * });
 * // Secondary instance
 * const secondaryInstance = new gcp.memorystore.Instance("secondary_instance", {
 *     instanceId: "secondary-instance",
 *     shardCount: 1,
 *     desiredAutoCreatedEndpoints: [{
 *         network: secondaryProducerNet.id,
 *         projectId: project.then(project => project.projectId),
 *     }],
 *     location: "europe-north1",
 *     replicaCount: 1,
 *     nodeType: "SHARED_CORE_NANO",
 *     transitEncryptionMode: "TRANSIT_ENCRYPTION_DISABLED",
 *     authorizationMode: "AUTH_DISABLED",
 *     engineConfigs: {
 *         "maxmemory-policy": "volatile-ttl",
 *     },
 *     zoneDistributionConfig: {
 *         mode: "SINGLE_ZONE",
 *         zone: "europe-north1-c",
 *     },
 *     deletionProtectionEnabled: true,
 *     crossInstanceReplicationConfig: {
 *         instanceRole: "SECONDARY",
 *         primaryInstance: {
 *             instance: primaryInstance.id,
 *         },
 *     },
 *     persistenceConfig: {
 *         mode: "RDB",
 *         rdbConfig: {
 *             rdbSnapshotPeriod: "ONE_HOUR",
 *             rdbSnapshotStartTime: "2024-10-02T15:01:23Z",
 *         },
 *     },
 *     labels: {
 *         abc: "xyz",
 *     },
 * }, {
 *     dependsOn: [secondaryPolicy],
 * });
 * ```
 *
 * ## Import
 *
 * Instance can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/locations/{{location}}/instances/{{instance_id}}`
 *
 * * `{{project}}/{{location}}/{{instance_id}}`
 *
 * * `{{location}}/{{instance_id}}`
 *
 * When using the `pulumi import` command, Instance can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:memorystore/instance:Instance default projects/{{project}}/locations/{{location}}/instances/{{instance_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:memorystore/instance:Instance default {{project}}/{{location}}/{{instance_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:memorystore/instance:Instance default {{location}}/{{instance_id}}
 * ```
 */
class Instance extends pulumi.CustomResource {
    /**
     * Get an existing Instance resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Instance(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Instance.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Instance.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["allowFewerZonesDeployment"] = state ? state.allowFewerZonesDeployment : undefined;
            resourceInputs["authorizationMode"] = state ? state.authorizationMode : undefined;
            resourceInputs["automatedBackupConfig"] = state ? state.automatedBackupConfig : undefined;
            resourceInputs["backupCollection"] = state ? state.backupCollection : undefined;
            resourceInputs["createTime"] = state ? state.createTime : undefined;
            resourceInputs["crossInstanceReplicationConfig"] = state ? state.crossInstanceReplicationConfig : undefined;
            resourceInputs["deletionProtectionEnabled"] = state ? state.deletionProtectionEnabled : undefined;
            resourceInputs["desiredAutoCreatedEndpoints"] = state ? state.desiredAutoCreatedEndpoints : undefined;
            resourceInputs["desiredPscAutoConnections"] = state ? state.desiredPscAutoConnections : undefined;
            resourceInputs["discoveryEndpoints"] = state ? state.discoveryEndpoints : undefined;
            resourceInputs["effectiveLabels"] = state ? state.effectiveLabels : undefined;
            resourceInputs["endpoints"] = state ? state.endpoints : undefined;
            resourceInputs["engineConfigs"] = state ? state.engineConfigs : undefined;
            resourceInputs["engineVersion"] = state ? state.engineVersion : undefined;
            resourceInputs["gcsSource"] = state ? state.gcsSource : undefined;
            resourceInputs["instanceId"] = state ? state.instanceId : undefined;
            resourceInputs["kmsKey"] = state ? state.kmsKey : undefined;
            resourceInputs["labels"] = state ? state.labels : undefined;
            resourceInputs["location"] = state ? state.location : undefined;
            resourceInputs["maintenancePolicy"] = state ? state.maintenancePolicy : undefined;
            resourceInputs["maintenanceSchedules"] = state ? state.maintenanceSchedules : undefined;
            resourceInputs["managedBackupSource"] = state ? state.managedBackupSource : undefined;
            resourceInputs["managedServerCas"] = state ? state.managedServerCas : undefined;
            resourceInputs["mode"] = state ? state.mode : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["nodeConfigs"] = state ? state.nodeConfigs : undefined;
            resourceInputs["nodeType"] = state ? state.nodeType : undefined;
            resourceInputs["persistenceConfig"] = state ? state.persistenceConfig : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["pscAttachmentDetails"] = state ? state.pscAttachmentDetails : undefined;
            resourceInputs["pscAutoConnections"] = state ? state.pscAutoConnections : undefined;
            resourceInputs["pulumiLabels"] = state ? state.pulumiLabels : undefined;
            resourceInputs["replicaCount"] = state ? state.replicaCount : undefined;
            resourceInputs["shardCount"] = state ? state.shardCount : undefined;
            resourceInputs["state"] = state ? state.state : undefined;
            resourceInputs["stateInfos"] = state ? state.stateInfos : undefined;
            resourceInputs["transitEncryptionMode"] = state ? state.transitEncryptionMode : undefined;
            resourceInputs["uid"] = state ? state.uid : undefined;
            resourceInputs["updateTime"] = state ? state.updateTime : undefined;
            resourceInputs["zoneDistributionConfig"] = state ? state.zoneDistributionConfig : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.instanceId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'instanceId'");
            }
            if ((!args || args.location === undefined) && !opts.urn) {
                throw new Error("Missing required property 'location'");
            }
            if ((!args || args.shardCount === undefined) && !opts.urn) {
                throw new Error("Missing required property 'shardCount'");
            }
            resourceInputs["allowFewerZonesDeployment"] = args ? args.allowFewerZonesDeployment : undefined;
            resourceInputs["authorizationMode"] = args ? args.authorizationMode : undefined;
            resourceInputs["automatedBackupConfig"] = args ? args.automatedBackupConfig : undefined;
            resourceInputs["crossInstanceReplicationConfig"] = args ? args.crossInstanceReplicationConfig : undefined;
            resourceInputs["deletionProtectionEnabled"] = args ? args.deletionProtectionEnabled : undefined;
            resourceInputs["desiredAutoCreatedEndpoints"] = args ? args.desiredAutoCreatedEndpoints : undefined;
            resourceInputs["desiredPscAutoConnections"] = args ? args.desiredPscAutoConnections : undefined;
            resourceInputs["engineConfigs"] = args ? args.engineConfigs : undefined;
            resourceInputs["engineVersion"] = args ? args.engineVersion : undefined;
            resourceInputs["gcsSource"] = args ? args.gcsSource : undefined;
            resourceInputs["instanceId"] = args ? args.instanceId : undefined;
            resourceInputs["kmsKey"] = args ? args.kmsKey : undefined;
            resourceInputs["labels"] = args ? args.labels : undefined;
            resourceInputs["location"] = args ? args.location : undefined;
            resourceInputs["maintenancePolicy"] = args ? args.maintenancePolicy : undefined;
            resourceInputs["managedBackupSource"] = args ? args.managedBackupSource : undefined;
            resourceInputs["mode"] = args ? args.mode : undefined;
            resourceInputs["nodeType"] = args ? args.nodeType : undefined;
            resourceInputs["persistenceConfig"] = args ? args.persistenceConfig : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["replicaCount"] = args ? args.replicaCount : undefined;
            resourceInputs["shardCount"] = args ? args.shardCount : undefined;
            resourceInputs["transitEncryptionMode"] = args ? args.transitEncryptionMode : undefined;
            resourceInputs["zoneDistributionConfig"] = args ? args.zoneDistributionConfig : undefined;
            resourceInputs["backupCollection"] = undefined /*out*/;
            resourceInputs["createTime"] = undefined /*out*/;
            resourceInputs["discoveryEndpoints"] = undefined /*out*/;
            resourceInputs["effectiveLabels"] = undefined /*out*/;
            resourceInputs["endpoints"] = undefined /*out*/;
            resourceInputs["maintenanceSchedules"] = undefined /*out*/;
            resourceInputs["managedServerCas"] = undefined /*out*/;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["nodeConfigs"] = undefined /*out*/;
            resourceInputs["pscAttachmentDetails"] = undefined /*out*/;
            resourceInputs["pscAutoConnections"] = undefined /*out*/;
            resourceInputs["pulumiLabels"] = undefined /*out*/;
            resourceInputs["state"] = undefined /*out*/;
            resourceInputs["stateInfos"] = undefined /*out*/;
            resourceInputs["uid"] = undefined /*out*/;
            resourceInputs["updateTime"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const secretOpts = { additionalSecretOutputs: ["effectiveLabels", "pulumiLabels"] };
        opts = pulumi.mergeOptions(opts, secretOpts);
        super(Instance.__pulumiType, name, resourceInputs, opts);
    }
}
exports.Instance = Instance;
/** @internal */
Instance.__pulumiType = 'gcp:memorystore/instance:Instance';
