// @pulumi/gcp@8.41.1/bigqueryanalyticshub/listing.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Listing = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A Bigquery Analytics Hub data exchange listing
 *
 * To get more information about Listing, see:
 *
 * * [API documentation](https://cloud.google.com/bigquery/docs/reference/analytics-hub/rest/v1/projects.locations.dataExchanges.listings)
 * * How-to Guides
 *     * [Official Documentation](https://cloud.google.com/bigquery/docs/analytics-hub-introduction)
 *
 * ## Example Usage
 *
 * ### Bigquery Analyticshub Listing Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const listing = new gcp.bigqueryanalyticshub.DataExchange("listing", {
 *     location: "US",
 *     dataExchangeId: "my_data_exchange",
 *     displayName: "my_data_exchange",
 *     description: "example data exchange",
 * });
 * const listingDataset = new gcp.bigquery.Dataset("listing", {
 *     datasetId: "my_listing",
 *     friendlyName: "my_listing",
 *     description: "example data exchange",
 *     location: "US",
 * });
 * const listingListing = new gcp.bigqueryanalyticshub.Listing("listing", {
 *     location: "US",
 *     dataExchangeId: listing.dataExchangeId,
 *     listingId: "my_listing",
 *     displayName: "my_listing",
 *     description: "example data exchange",
 *     bigqueryDataset: {
 *         dataset: listingDataset.id,
 *     },
 * });
 * ```
 * ### Bigquery Analyticshub Listing Restricted
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const listing = new gcp.bigqueryanalyticshub.DataExchange("listing", {
 *     location: "US",
 *     dataExchangeId: "my_data_exchange",
 *     displayName: "my_data_exchange",
 *     description: "example data exchange",
 * });
 * const listingDataset = new gcp.bigquery.Dataset("listing", {
 *     datasetId: "my_listing",
 *     friendlyName: "my_listing",
 *     description: "example data exchange",
 *     location: "US",
 * });
 * const listingListing = new gcp.bigqueryanalyticshub.Listing("listing", {
 *     location: "US",
 *     dataExchangeId: listing.dataExchangeId,
 *     listingId: "my_listing",
 *     displayName: "my_listing",
 *     description: "example data exchange",
 *     bigqueryDataset: {
 *         dataset: listingDataset.id,
 *     },
 *     restrictedExportConfig: {
 *         enabled: true,
 *         restrictQueryResult: true,
 *     },
 * });
 * ```
 * ### Bigquery Analyticshub Listing Dcr
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const listing = new gcp.bigqueryanalyticshub.DataExchange("listing", {
 *     location: "US",
 *     dataExchangeId: "dcr_data_exchange",
 *     displayName: "dcr_data_exchange",
 *     description: "example dcr data exchange",
 *     sharingEnvironmentConfig: {
 *         dcrExchangeConfig: {},
 *     },
 * });
 * const listingDataset = new gcp.bigquery.Dataset("listing", {
 *     datasetId: "dcr_listing",
 *     friendlyName: "dcr_listing",
 *     description: "example dcr data exchange",
 *     location: "US",
 * });
 * const listingTable = new gcp.bigquery.Table("listing", {
 *     deletionProtection: false,
 *     tableId: "dcr_listing",
 *     datasetId: listingDataset.datasetId,
 *     schema: `[
 *   {
 *     "name": "name",
 *     "type": "STRING",
 *     "mode": "NULLABLE"
 *   },
 *   {
 *     "name": "post_abbr",
 *     "type": "STRING",
 *     "mode": "NULLABLE"
 *   },
 *   {
 *     "name": "date",
 *     "type": "DATE",
 *     "mode": "NULLABLE"
 *   }
 * ]
 * `,
 * });
 * const listingListing = new gcp.bigqueryanalyticshub.Listing("listing", {
 *     location: "US",
 *     dataExchangeId: listing.dataExchangeId,
 *     listingId: "dcr_listing",
 *     displayName: "dcr_listing",
 *     description: "example dcr data exchange",
 *     bigqueryDataset: {
 *         dataset: listingDataset.id,
 *         selectedResources: [{
 *             table: listingTable.id,
 *         }],
 *     },
 *     restrictedExportConfig: {
 *         enabled: true,
 *     },
 * });
 * ```
 * ### Bigquery Analyticshub Listing Log Linked Dataset Query User
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const listingLogEmail = new gcp.bigqueryanalyticshub.DataExchange("listing_log_email", {
 *     location: "US",
 *     dataExchangeId: "tf_test_log_email_de",
 *     displayName: "tf_test_log_email_de",
 *     description: "Example for log email test",
 * });
 * const listingLogEmailDataset = new gcp.bigquery.Dataset("listing_log_email", {
 *     datasetId: "tf_test_log_email_ds",
 *     friendlyName: "tf_test_log_email_ds",
 *     description: "Example for log email test",
 *     location: "US",
 * });
 * const listing = new gcp.bigqueryanalyticshub.Listing("listing", {
 *     location: "US",
 *     dataExchangeId: listingLogEmail.dataExchangeId,
 *     listingId: "tf_test_log_email_listing",
 *     displayName: "tf_test_log_email_listing",
 *     description: "Example for log email test",
 *     logLinkedDatasetQueryUserEmail: true,
 *     bigqueryDataset: {
 *         dataset: listingLogEmailDataset.id,
 *     },
 * });
 * ```
 * ### Bigquery Analyticshub Listing Pubsub
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const listing = new gcp.bigqueryanalyticshub.DataExchange("listing", {
 *     location: "US",
 *     dataExchangeId: "tf_test_pubsub_data_exchange",
 *     displayName: "tf_test_pubsub_data_exchange",
 *     description: "Example for pubsub topic source",
 * });
 * const tfTestPubsubTopic = new gcp.pubsub.Topic("tf_test_pubsub_topic", {name: "test_pubsub"});
 * const listingListing = new gcp.bigqueryanalyticshub.Listing("listing", {
 *     location: "US",
 *     dataExchangeId: listing.dataExchangeId,
 *     listingId: "tf_test_pubsub_listing",
 *     displayName: "tf_test_pubsub_listing",
 *     description: "Example for pubsub topic source",
 *     pubsubTopic: {
 *         topic: tfTestPubsubTopic.id,
 *         dataAffinityRegions: [
 *             "us-central1",
 *             "europe-west1",
 *         ],
 *     },
 * });
 * ```
 * ### Bigquery Analyticshub Listing Dcr Routine
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const dcrDataExchangeExample = new gcp.bigqueryanalyticshub.DataExchange("dcr_data_exchange_example", {
 *     location: "us",
 *     dataExchangeId: "tf_test_data_exchange",
 *     displayName: "tf_test_data_exchange",
 *     description: "Example for listing with routine",
 *     sharingEnvironmentConfig: {
 *         dcrExchangeConfig: {},
 *     },
 * });
 * const listing = new gcp.bigquery.Dataset("listing", {
 *     datasetId: "tf_test_dataset",
 *     friendlyName: "tf_test_dataset",
 *     description: "Example for listing with routine",
 *     location: "us",
 * });
 * const listingRoutine = new gcp.bigquery.Routine("listing", {
 *     datasetId: listing.datasetId,
 *     routineId: "tf_test_routine",
 *     routineType: "TABLE_VALUED_FUNCTION",
 *     language: "SQL",
 *     description: "A DCR routine example.",
 *     definitionBody: "SELECT 1 + value AS value\n",
 *     arguments: [{
 *         name: "value",
 *         argumentKind: "FIXED_TYPE",
 *         dataType: JSON.stringify({
 *             typeKind: "INT64",
 *         }),
 *     }],
 *     returnTableType: JSON.stringify({
 *         columns: [{
 *             name: "value",
 *             type: {
 *                 typeKind: "INT64",
 *             },
 *         }],
 *     }),
 * });
 * const listingListing = new gcp.bigqueryanalyticshub.Listing("listing", {
 *     location: "US",
 *     dataExchangeId: dcrDataExchangeExample.dataExchangeId,
 *     listingId: "tf_test_listing_routine",
 *     displayName: "tf_test_listing_routine",
 *     description: "Example for listing with routine",
 *     bigqueryDataset: {
 *         dataset: listing.id,
 *         selectedResources: [{
 *             routine: listingRoutine.id,
 *         }],
 *     },
 *     restrictedExportConfig: {
 *         enabled: true,
 *     },
 * });
 * ```
 * ### Bigquery Analyticshub Public Listing
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const listing = new gcp.bigqueryanalyticshub.DataExchange("listing", {
 *     location: "US",
 *     dataExchangeId: "my_data_exchange",
 *     displayName: "my_data_exchange",
 *     description: "example public listing",
 *     discoveryType: "DISCOVERY_TYPE_PUBLIC",
 * });
 * const listingDataset = new gcp.bigquery.Dataset("listing", {
 *     datasetId: "my_listing",
 *     friendlyName: "my_listing",
 *     description: "example public listing",
 *     location: "US",
 * });
 * const listingListing = new gcp.bigqueryanalyticshub.Listing("listing", {
 *     location: "US",
 *     dataExchangeId: listing.dataExchangeId,
 *     listingId: "my_listing",
 *     displayName: "my_listing",
 *     description: "example public listing",
 *     discoveryType: "DISCOVERY_TYPE_PUBLIC",
 *     allowOnlyMetadataSharing: false,
 *     bigqueryDataset: {
 *         dataset: listingDataset.id,
 *     },
 * });
 * ```
 * ### Bigquery Analyticshub Listing Marketplace
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const listing = new gcp.bigqueryanalyticshub.DataExchange("listing", {
 *     location: "US",
 *     dataExchangeId: "my_data_exchange",
 *     displayName: "my_data_exchange",
 *     description: "example data exchange",
 * });
 * const listingDataset = new gcp.bigquery.Dataset("listing", {
 *     datasetId: "my_listing",
 *     friendlyName: "my_listing",
 *     description: "example data exchange",
 *     location: "US",
 * });
 * const listingListing = new gcp.bigqueryanalyticshub.Listing("listing", {
 *     location: "US",
 *     dataExchangeId: listing.dataExchangeId,
 *     listingId: "my_listing",
 *     displayName: "my_listing",
 *     description: "example data exchange",
 *     deleteCommercial: true,
 *     bigqueryDataset: {
 *         dataset: listingDataset.id,
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * Listing can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/locations/{{location}}/dataExchanges/{{data_exchange_id}}/listings/{{listing_id}}`
 *
 * * `{{project}}/{{location}}/{{data_exchange_id}}/{{listing_id}}`
 *
 * * `{{location}}/{{data_exchange_id}}/{{listing_id}}`
 *
 * When using the `pulumi import` command, Listing can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:bigqueryanalyticshub/listing:Listing default projects/{{project}}/locations/{{location}}/dataExchanges/{{data_exchange_id}}/listings/{{listing_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:bigqueryanalyticshub/listing:Listing default {{project}}/{{location}}/{{data_exchange_id}}/{{listing_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:bigqueryanalyticshub/listing:Listing default {{location}}/{{data_exchange_id}}/{{listing_id}}
 * ```
 */
class Listing extends pulumi.CustomResource {
    /**
     * Get an existing Listing resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Listing(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Listing.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Listing.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["allowOnlyMetadataSharing"] = state ? state.allowOnlyMetadataSharing : undefined;
            resourceInputs["bigqueryDataset"] = state ? state.bigqueryDataset : undefined;
            resourceInputs["categories"] = state ? state.categories : undefined;
            resourceInputs["commercialInfos"] = state ? state.commercialInfos : undefined;
            resourceInputs["dataExchangeId"] = state ? state.dataExchangeId : undefined;
            resourceInputs["dataProvider"] = state ? state.dataProvider : undefined;
            resourceInputs["deleteCommercial"] = state ? state.deleteCommercial : undefined;
            resourceInputs["description"] = state ? state.description : undefined;
            resourceInputs["discoveryType"] = state ? state.discoveryType : undefined;
            resourceInputs["displayName"] = state ? state.displayName : undefined;
            resourceInputs["documentation"] = state ? state.documentation : undefined;
            resourceInputs["icon"] = state ? state.icon : undefined;
            resourceInputs["listingId"] = state ? state.listingId : undefined;
            resourceInputs["location"] = state ? state.location : undefined;
            resourceInputs["logLinkedDatasetQueryUserEmail"] = state ? state.logLinkedDatasetQueryUserEmail : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["primaryContact"] = state ? state.primaryContact : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["publisher"] = state ? state.publisher : undefined;
            resourceInputs["pubsubTopic"] = state ? state.pubsubTopic : undefined;
            resourceInputs["requestAccess"] = state ? state.requestAccess : undefined;
            resourceInputs["restrictedExportConfig"] = state ? state.restrictedExportConfig : undefined;
            resourceInputs["state"] = state ? state.state : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.dataExchangeId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'dataExchangeId'");
            }
            if ((!args || args.displayName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'displayName'");
            }
            if ((!args || args.listingId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'listingId'");
            }
            if ((!args || args.location === undefined) && !opts.urn) {
                throw new Error("Missing required property 'location'");
            }
            resourceInputs["allowOnlyMetadataSharing"] = args ? args.allowOnlyMetadataSharing : undefined;
            resourceInputs["bigqueryDataset"] = args ? args.bigqueryDataset : undefined;
            resourceInputs["categories"] = args ? args.categories : undefined;
            resourceInputs["dataExchangeId"] = args ? args.dataExchangeId : undefined;
            resourceInputs["dataProvider"] = args ? args.dataProvider : undefined;
            resourceInputs["deleteCommercial"] = args ? args.deleteCommercial : undefined;
            resourceInputs["description"] = args ? args.description : undefined;
            resourceInputs["discoveryType"] = args ? args.discoveryType : undefined;
            resourceInputs["displayName"] = args ? args.displayName : undefined;
            resourceInputs["documentation"] = args ? args.documentation : undefined;
            resourceInputs["icon"] = args ? args.icon : undefined;
            resourceInputs["listingId"] = args ? args.listingId : undefined;
            resourceInputs["location"] = args ? args.location : undefined;
            resourceInputs["logLinkedDatasetQueryUserEmail"] = args ? args.logLinkedDatasetQueryUserEmail : undefined;
            resourceInputs["primaryContact"] = args ? args.primaryContact : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["publisher"] = args ? args.publisher : undefined;
            resourceInputs["pubsubTopic"] = args ? args.pubsubTopic : undefined;
            resourceInputs["requestAccess"] = args ? args.requestAccess : undefined;
            resourceInputs["restrictedExportConfig"] = args ? args.restrictedExportConfig : undefined;
            resourceInputs["commercialInfos"] = undefined /*out*/;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["state"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(Listing.__pulumiType, name, resourceInputs, opts);
    }
}
exports.Listing = Listing;
/** @internal */
Listing.__pulumiType = 'gcp:bigqueryanalyticshub/listing:Listing';
