// @pulumi/gcp@8.41.1/eventarc/enrollment.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Enrollment = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * The Eventarc Enrollment resource
 *
 * To get more information about Enrollment, see:
 *
 * * [API documentation](https://cloud.google.com/eventarc/docs/reference/rest/v1/projects.locations.enrollments)
 * * How-to Guides
 *     * [Official Documentation](https://cloud.google.com/eventarc/advanced/docs/receive-events/create-enrollment)
 *
 * ## Example Usage
 *
 * ### Eventarc Enrollment With Pipeline Destination
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const pipeline = new gcp.eventarc.Pipeline("pipeline", {
 *     location: "us-central1",
 *     pipelineId: "some-pipeline",
 *     destinations: [{
 *         httpEndpoint: {
 *             uri: "https://10.77.0.0:80/route",
 *         },
 *         networkConfig: {
 *             networkAttachment: "projects/my-project-name/regions/us-central1/networkAttachments/some-network-attachment",
 *         },
 *     }],
 * });
 * const primary = new gcp.eventarc.Enrollment("primary", {
 *     location: "us-central1",
 *     enrollmentId: "some-enrollment",
 *     messageBus: primaryGoogleEventarcMessageBus.id,
 *     destination: pipeline.id,
 *     celMatch: "message.type == 'google.cloud.dataflow.job.v1beta3.statusChanged'",
 * });
 * const messageBus = new gcp.eventarc.MessageBus("message_bus", {
 *     location: "us-central1",
 *     messageBusId: "some-message-bus",
 * });
 * ```
 *
 * ## Import
 *
 * Enrollment can be imported using any of these accepted formats:
 *
 * * `projects/{{project}}/locations/{{location}}/enrollments/{{enrollment_id}}`
 *
 * * `{{project}}/{{location}}/{{enrollment_id}}`
 *
 * * `{{location}}/{{enrollment_id}}`
 *
 * When using the `pulumi import` command, Enrollment can be imported using one of the formats above. For example:
 *
 * ```sh
 * $ pulumi import gcp:eventarc/enrollment:Enrollment default projects/{{project}}/locations/{{location}}/enrollments/{{enrollment_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:eventarc/enrollment:Enrollment default {{project}}/{{location}}/{{enrollment_id}}
 * ```
 *
 * ```sh
 * $ pulumi import gcp:eventarc/enrollment:Enrollment default {{location}}/{{enrollment_id}}
 * ```
 */
class Enrollment extends pulumi.CustomResource {
    /**
     * Get an existing Enrollment resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Enrollment(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Enrollment.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Enrollment.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["annotations"] = state ? state.annotations : undefined;
            resourceInputs["celMatch"] = state ? state.celMatch : undefined;
            resourceInputs["createTime"] = state ? state.createTime : undefined;
            resourceInputs["destination"] = state ? state.destination : undefined;
            resourceInputs["displayName"] = state ? state.displayName : undefined;
            resourceInputs["effectiveAnnotations"] = state ? state.effectiveAnnotations : undefined;
            resourceInputs["effectiveLabels"] = state ? state.effectiveLabels : undefined;
            resourceInputs["enrollmentId"] = state ? state.enrollmentId : undefined;
            resourceInputs["etag"] = state ? state.etag : undefined;
            resourceInputs["labels"] = state ? state.labels : undefined;
            resourceInputs["location"] = state ? state.location : undefined;
            resourceInputs["messageBus"] = state ? state.messageBus : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["project"] = state ? state.project : undefined;
            resourceInputs["pulumiLabels"] = state ? state.pulumiLabels : undefined;
            resourceInputs["uid"] = state ? state.uid : undefined;
            resourceInputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.celMatch === undefined) && !opts.urn) {
                throw new Error("Missing required property 'celMatch'");
            }
            if ((!args || args.destination === undefined) && !opts.urn) {
                throw new Error("Missing required property 'destination'");
            }
            if ((!args || args.enrollmentId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'enrollmentId'");
            }
            if ((!args || args.location === undefined) && !opts.urn) {
                throw new Error("Missing required property 'location'");
            }
            if ((!args || args.messageBus === undefined) && !opts.urn) {
                throw new Error("Missing required property 'messageBus'");
            }
            resourceInputs["annotations"] = args ? args.annotations : undefined;
            resourceInputs["celMatch"] = args ? args.celMatch : undefined;
            resourceInputs["destination"] = args ? args.destination : undefined;
            resourceInputs["displayName"] = args ? args.displayName : undefined;
            resourceInputs["enrollmentId"] = args ? args.enrollmentId : undefined;
            resourceInputs["labels"] = args ? args.labels : undefined;
            resourceInputs["location"] = args ? args.location : undefined;
            resourceInputs["messageBus"] = args ? args.messageBus : undefined;
            resourceInputs["project"] = args ? args.project : undefined;
            resourceInputs["createTime"] = undefined /*out*/;
            resourceInputs["effectiveAnnotations"] = undefined /*out*/;
            resourceInputs["effectiveLabels"] = undefined /*out*/;
            resourceInputs["etag"] = undefined /*out*/;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["pulumiLabels"] = undefined /*out*/;
            resourceInputs["uid"] = undefined /*out*/;
            resourceInputs["updateTime"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const secretOpts = { additionalSecretOutputs: ["effectiveLabels", "pulumiLabels"] };
        opts = pulumi.mergeOptions(opts, secretOpts);
        super(Enrollment.__pulumiType, name, resourceInputs, opts);
    }
}
exports.Enrollment = Enrollment;
/** @internal */
Enrollment.__pulumiType = 'gcp:eventarc/enrollment:Enrollment';
