// @pulumi/gcp@8.41.1/eventarc/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Trigger = exports.Pipeline = exports.MessageBus = exports.GoogleChannelConfig = exports.GoogleApiSource = exports.Enrollment = exports.Channel = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.Channel = null;
utilities.lazyLoad(exports, ["Channel"], () => require("./channel"));
exports.Enrollment = null;
utilities.lazyLoad(exports, ["Enrollment"], () => require("./enrollment"));
exports.GoogleApiSource = null;
utilities.lazyLoad(exports, ["GoogleApiSource"], () => require("./googleApiSource"));
exports.GoogleChannelConfig = null;
utilities.lazyLoad(exports, ["GoogleChannelConfig"], () => require("./googleChannelConfig"));
exports.MessageBus = null;
utilities.lazyLoad(exports, ["MessageBus"], () => require("./messageBus"));
exports.Pipeline = null;
utilities.lazyLoad(exports, ["Pipeline"], () => require("./pipeline"));
exports.Trigger = null;
utilities.lazyLoad(exports, ["Trigger"], () => require("./trigger"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "gcp:eventarc/channel:Channel":
                return new exports.Channel(name, undefined, { urn });
            case "gcp:eventarc/enrollment:Enrollment":
                return new exports.Enrollment(name, undefined, { urn });
            case "gcp:eventarc/googleApiSource:GoogleApiSource":
                return new exports.GoogleApiSource(name, undefined, { urn });
            case "gcp:eventarc/googleChannelConfig:GoogleChannelConfig":
                return new exports.GoogleChannelConfig(name, undefined, { urn });
            case "gcp:eventarc/messageBus:MessageBus":
                return new exports.MessageBus(name, undefined, { urn });
            case "gcp:eventarc/pipeline:Pipeline":
                return new exports.Pipeline(name, undefined, { urn });
            case "gcp:eventarc/trigger:Trigger":
                return new exports.Trigger(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("gcp", "eventarc/channel", _module);
pulumi.runtime.registerResourceModule("gcp", "eventarc/enrollment", _module);
pulumi.runtime.registerResourceModule("gcp", "eventarc/googleApiSource", _module);
pulumi.runtime.registerResourceModule("gcp", "eventarc/googleChannelConfig", _module);
pulumi.runtime.registerResourceModule("gcp", "eventarc/messageBus", _module);
pulumi.runtime.registerResourceModule("gcp", "eventarc/pipeline", _module);
pulumi.runtime.registerResourceModule("gcp", "eventarc/trigger", _module);
